<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />


<meta name="author" content="UM Bioinformatics Core" />

<meta name="date" content="2023-09-29" />

<title>Differential Expression (DE) Introduction</title>

<script src="site_libs/header-attrs-2.21/header-attrs.js"></script>
<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/paper.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<script src="site_libs/navigation-1.1/codefolding.js"></script>
<script src="site_libs/navigation-1.1/sourceembed.js"></script>
<link href="site_libs/highlightjs-9.12.0/default.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link rel="shortcut icon" href="favicon-16x16.png" />
<!--
Favicon from
https://twemoji.twitter.com/
https://favicon.io/emoji-favicons/dna/
-->

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>




<style type="text/css">
#rmd-source-code {
  display: none;
}
</style>





<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
details > summary > p:only-child {
  display: inline;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark the anchor link active (and if it's in a dropdown, also mark that active)
  var dropdown = menuAnchor.closest('li.dropdown');
  if (window.bootstrap) { // Bootstrap 4+
    menuAnchor.addClass('active');
    dropdown.find('> .dropdown-toggle').addClass('active');
  } else { // Bootstrap 3
    menuAnchor.parent().addClass('active');
    dropdown.addClass('active');
  }

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before, .tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "\e259";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "\e258";
  font-family: 'Glyphicons Halflings';
  border: none;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->
<style type="text/css">
.code-folding-btn { margin-bottom: 4px; }
</style>



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-inverse  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-bs-toggle="collapse" data-target="#navbar" data-bs-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">RNA-Seq Demystified</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="Module00_Introduction.html">Intro</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Sequence QC
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="Module01_Warming_Up.html">Warming up</a>
    </li>
    <li>
      <a href="Module02a_Sequence_QC.html">Sequence QC</a>
    </li>
    <li>
      <a href="Module02b_Trimming.html">Trimming</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Alignment
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="Module03a_Reference_Genomes.html">Reference Genomes</a>
    </li>
    <li>
      <a href="Module03b_Alignment.html">Alignment</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Quantification
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="Module04_Alignment_QC_and_Quantification.html">Alignment QC and Quantification</a>
    </li>
    <li>
      <a href="Module05_Additional_Details.html">Optional: Additional details</a>
    </li>
  </ul>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Groups/Model
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="Module06_DEAnalysisSetup.html">DE Introduction</a>
    </li>
    <li>
      <a href="Module07_DESeq2Init.html">DE Initialization</a>
    </li>
    <li>
      <a href="Module08_DESeq2DE.html">DE Normalization and Modeling</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Expression plots
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="Module09_SampleQCViz.html">Sample QC</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Diff Expression
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="Module10_DEComparisons.html">DE Testing</a>
    </li>
    <li>
      <a href="Module11_DEVisualizations.html">DE Visualization</a>
    </li>
    <li>
      <a href="Module12_DEAnnotations.html">DE Annotation</a>
    </li>
    <li>
      <a href="R_bonus_content.html">Bonus content</a>
    </li>
  </ul>
</li>
<li>
  <a href="Module99_Wrap_up.html">Wrap up</a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">

<div class="btn-group pull-right float-right">
<button type="button" class="btn btn-default btn-xs btn-secondary btn-sm dropdown-toggle" data-toggle="dropdown" data-bs-toggle="dropdown" aria-haspopup="true" aria-expanded="false"><span>Code</span> <span class="caret"></span></button>
<ul class="dropdown-menu dropdown-menu-right" style="min-width: 50px;">
<li><a id="rmd-download-source" href="#">Download Rmd</a></li>
</ul>
</div>



<h1 class="title toc-ignore">Differential Expression (DE)
Introduction</h1>
<h4 class="author">UM Bioinformatics Core</h4>
<h4 class="date">2023-09-29</h4>

</div>


<style type="text/css">
body, td {
   font-size: 18px;
}
code.r{
  font-size: 12px;
}
pre {
  font-size: 12px
}
</style>
<p>In this module, we will learn:</p>
<ul>
<li>How to set up a differential expression analysis, using reproducible
research principles<br />
</li>
<li>What is DESeq2 and why it is widely used for differential expression
comparisons</li>
<li>How to load and review gene count data</li>
</ul>
<p><br></p>
<div id="differential-expression-workflow"
class="section level1 unlisted unnumbered">
<h1 class="unlisted unnumbered">Differential Expression Workflow</h1>
<p>Today we will proceed through key steps in a differential expression
analysis, starting from a count table that’s similar to what you
generated in the first half of the workshop and one of the outputs
included in the <a
href="https://brcf.medicine.umich.edu/cores/advanced-genomics/data-delivery/">data
that the Advanced Genomics Core delivers</a> for RNA-seq libraries.</p>
<p><img src="images/wayfinder/wayfinder-DEAnalysisSetup.png" alt="wayfinder" style="width: 1000px;"/></p>
</div>
<div id="getting-started-in-rstudio" class="section level1">
<h1>Getting Started in RStudio</h1>
<div id="log-on-to-rstudio-server" class="section level2">
<h2>Log on to RStudio Server</h2>
<p>Open a web browser to the following URL: <a
href="http://bfx-workshop01.med.umich.edu">http://bfx-workshop01.med.umich.edu</a></p>
<p>You should now be looking at a page that will allow you to login to
the RStudio server:</p>
<p><img src="images/rstudio_login_screen.png" alt="rstudio default session" style="width: 1000px;"/></p>
<p>Enter your user credentials and click <kbd>Sign In</kbd>. The
credentials are the same used earlier in the workshop and were provided
via an email entitled “UM BioinfCore Workshop Login”.</p>
<p>However, if you forget yours, a helper can retrieve it for you if you
ask in Slack. You should now see the RStudio interface:</p>
<p><img src="images/rstudio_session_default.png" alt="rstudio default session" style="width:1000px;"/></p>
<p><br></p>
</div>
<div id="what-are-we-working-towards" class="section level2">
<h2>What are we working towards?</h2>
<p>To refresh ourselves on the interface of RStudio, look in the lower
right pane, that has a list of files and folders. Click on the
<code>RSD_R</code> folder and then click on the
<code>WelcomeToRSD.R</code> script to open it.</p>
<p>There should now be four panes in your RStudio window, with the
<code>WelcomeToRSD</code> script in the upper left pane, this is the
“Source” or “Scripts” pane. It displays the code that we will write to
perform our analysis.</p>
<p><img src="images/WelcomeScriptOpened.png" alt="rstudio default session" style="width:1000px;"/></p>
<p>In the Scripts pane, there is a line of icons, towards the right side
of the pane there is a <kbd>Run</kbd> button. First highlight all the
code in the Source pane, and then click <kbd>Run</kbd>.</p>
<p><br></p>
<p><strong>Checkpoint</strong></p>
<p>What just happened?</p>
<ol style="list-style-type: decimal">
<li><p>We see in that our code has been run in the Console panel,
including some warning messages.</p></li>
<li><p>We see that a number of objects were created in our Environment
panel including an object named <code>DE_results</code>.</p></li>
<li><p>We see that the Plot tab now has a volcano plot displayed - Note:
if the plot is not displayed, you may need to expand the right panel
boundary before re-running the last line of the script</p></li>
</ol>
<p><strong>What does this volcano plot show?</strong></p>
<p><em>Now that we have an idea of what we are working towards, any
questions?</em></p>
<p><br> <br></p>
</div>
<div id="create-an-rstudio-project" class="section level2">
<h2>Create an RStudio Project</h2>
<p>Just like in Computational Foundation workshop, let’s start our
analysis by creating a project where we can store our data, scripts and
outputs.</p>
<ol style="list-style-type: decimal">
<li>To create a project, go to the <kbd>File</kbd> menu, and click
<kbd>New Project…</kbd>. The following window will appear:</li>
</ol>
<p><img src="images/new_project_window.png" alt="new project window" style="width: 600px;"/></p>
<ol start="2" style="list-style-type: decimal">
<li><p>In this window, select <kbd>Existing Directory</kbd>. For
“Project working directory”, click <kbd>Browse…</kbd>, select the
“RSD_R” folder, and click <kbd>Choose</kbd>. This will use the
<code>/home/workshop/user/RSD_R</code> folder as the project
directory.</p></li>
<li><p>Finally click <kbd>Create Project</kbd>. In the “Files” tab of
your output pane (more about the RStudio layout in a moment), you should
see an RStudio project file, <strong>RSD_R.Rproj</strong>. All RStudio
projects end with the “<strong>.Rproj</strong>” file extension.</p></li>
</ol>
<p>Note that there is already a <code>data/</code> folder which contains
the data we will use for these lessons.</p>
</div>
<div id="creating-an-r-script" class="section level2">
<h2>Creating an R script</h2>
<p>Let’s create an R script file:</p>
<ul>
<li>Click the <kbd>File</kbd> menu and select <kbd>New File</kbd> and
then <kbd>R Script</kbd>.</li>
<li>Before we go any further, save your script by clicking the save/disk
icon that is in the bar above the first line in the script editor, or
click the <kbd>File</kbd> menu and select <kbd>Save</kbd>.</li>
<li>In the “Save File” window that opens, select <kbd>New Folder</kbd>.
Name it <strong>“scripts”</strong>.</li>
<li>Finally, name your file <strong>“diffex”</strong> in the “File name”
field.</li>
</ul>
<p>The new script <strong>diffex.R</strong> is now in the
<code>scripts</code> folder. You can see that by clicking the
<code>scripts</code> folder in the “Files” pane. And you can go back up
to the main project folder by clicking the <code>..</code> to the right
of the up arrow in the “Files” pane. By convention, R scripts end with
the file extension <strong>.R</strong>.</p>
</div>
<div id="file-organization-reproducible-research"
class="section level2">
<h2>File organization &amp; reproducible research</h2>
<p>Part of any successful bioinformatics analysis is documentation and
data stewardship, which are important aspects of “reproducible research”
(see: <a
href="https://journals.plos.org/ploscompbiol/article?id=10.1371/journal.pcbi.1000424">Nobel,
2009</a>).</p>
<p>To follow best practices for file organization for reproducible
research, we should make distinct locations for:</p>
<ul>
<li>Raw data</li>
<li>Code</li>
<li>Output</li>
</ul>
<p>To organize our files , we’ll create some more directories in the
<code>RSD_R</code> folder which is now our project root and current
working directory. We can already see that there a <code>data/</code>
folder and we just created the <code>scripts/</code> folder with the
RStudio GUI. Let’s create an <code>outputs/</code> folder using the
<code>dir.create()</code> function in R.</p>
<pre class="r"><code>dir.create(&#39;outputs&#39;)
dir.create(&#39;outputs/figures&#39;)
dir.create(&#39;outputs/tables&#39;)</code></pre>
<p><strong>Checkpoint</strong>: <em>Please use the green ‘check’ if you
have saved your code file and see the <code>data</code>,
<code>scripts</code>, and <code>outputs</code> directories in the
<code>RSD_R</code> folder and the red ‘x’ if you do not.</em></p>
<blockquote>
<h2 id="reminder-rstudio-code-execution"
class="unlisted unnumbered">Reminder: RStudio code execution</h2>
<p><kbd>Ctrl</kbd>+<kbd>Enter</kbd> is a standard shortcut in RStudio to
send the current line (or selected lines) to the console. If you see
<code>&gt;</code> in the Console, then R has executed the command. If
you see a <code>+</code>, this means that the command is not complete, R
thinks there is more to your command. You can use the <kmd>esc</kmd> to
get out of this state.</p>
</blockquote>
<blockquote>
<h2 id="reminder-object-naming-conventions"
class="unlisted unnumbered">Reminder: Object naming conventions</h2>
<ul>
<li>Cannot start with numbers</li>
<li>Cannot include dashes</li>
<li>Cannot have spaces</li>
<li>Should not be identical to a named function</li>
<li>Dots and underscores can separate parts of names, alternatively
CamelCase accomplishes this</li>
</ul>
</blockquote>
<hr />
</div>
</div>
<div id="tools-for-differential-gene-expression-analysis"
class="section level1">
<h1>Tools for Differential Gene Expression analysis</h1>
<p>As discussed during the webinar, bulk RNA-seq can be used to test for
differential expression between conditions or treatments, using
statistical approaches that are appropriate for biological data.</p>
<p>While there are several tools that can be used for differential
expression comparisons, we will use <a
href="https://bioconductor.org/packages/release/bioc/html/DESeq2.html">DESeq2</a>
in our analysis today.</p>
<p>DESeq2 is one of two tools, along with <a
href="https://bioconductor.org/packages/release/bioc/html/edgeR.html">edgeR</a>,
considered <a
href="https://bmcbioinformatics.biomedcentral.com/articles/10.1186/1471-2105-14-91">‘best
practice’</a> for differential expression comparisons. Both tools apply
similar methods that account for the “shape” of data expected for
RNA-seq and are fairly stringent in calling differentially expressed
genes, lowering the risk of investigating “false positive” genes
(e.g. genes that don’t actually have different expression between
treatment groups and therefore are not relevant to the biological
process).</p>
<p>Additionally, DESeq2 also has an <a
href="https://bioconductor.org/packages/release/bioc/vignettes/DESeq2/inst/doc/DESeq2.html">excellent
vignette</a> from Love, Anders, and Huber, from which our workflow is
partially adapted, and is a good resource when analyzing your own data
(see also: <a href="https://doi.org/10.1186/s13059-014-0550-8">Love,
Anders, and Huber. <em>Genome Biology</em>. 2014.</a>).</p>
<div id="deseq2-assumptions-and-requirements" class="section level2">
<h2>DESeq2 assumptions and requirements</h2>
<p>A key assumption of DESeq2 is that <em>biological variance is much
greater than technical variance</em> (which should be true if <a
href="https://www.txgen.tamu.edu/faq/rna-isolation-best-practices/">best
practices</a> for <a
href="https://www.biocompare.com/Bench-Tips/128790-Four-Tips-for-Perfecting-RNA-Isolation/">quality
RNA isolation</a> are followed, including DNase treatment!).</p>
<p>Since calculating variance is key to the statistical approach used
for DESeq2, if we tried to compare treatment groups with only one sample
each, we would get an error (as shown in <a
href="https://support.bioconductor.org/p/89746/">this blog post</a>).
Without replicates, there can’t be statistical significance
(e.g. p-values), but <em>qualitative</em> approaches are an option, like
looking at the top expressed genes after normalization.</p>
<details>
<summary>
<em>Click for additional resources regarding statistical testing and
tool comparison for RNA-seq data</em>
</summary>
To learn more about statistical testing and what distributions best
model the behavior of RNA-seq data, a good resource is this <a
href="https://www.youtube.com/watch?v=HK7WKsL3c2w&amp;feature=youtu.be">EdX
lecture by Rafael Irizarry</a> or this <a
href="https://www.youtube.com/watch?v=C8RNvWu7pAw">lecture by Kasper
Hansen</a>. Another helpful guide is this <a
href="https://journals.plos.org/plosone/article?id=10.1371/journal.pone.0103207">Comparative
Study for Differential Expression Analysis by Zhang et al.</a> from
2014.
</details>
<p><br></p>
<blockquote>
<h2 id="exercise">Exercise</h2>
<ul>
<li>Post a question that was NOT addressed but that you hope we will
address in the later modules <strong>OR</strong></li>
<li>Add a thumbs up to your favorite comment(s) to upvote it</li>
</ul>
</blockquote>
<hr />
</div>
</div>
<div id="getting-started" class="section level1">
<h1>Getting started</h1>
<div id="load-packages" class="section level2">
<h2>Load Packages</h2>
<p>Several packages have already been installed on the server, so we can
load them into our R session now. To do that we’ll use the
<code>library</code> function to load the required packages.</p>
<pre class="r"><code>library(DESeq2)
library(ggplot2)
library(tidyr)
library(dplyr)
library(matrixStats)
library(ggrepel)
library(pheatmap)
library(RColorBrewer)
library(data.table)</code></pre>
<p><em>Note: We expect to see some red messages in your console while
these packages are loading</em></p>
<p>R/RStudio [AKA Posit] has great resources for getting help, including
<a href="https://posit.co/resources/cheatsheets/">code ‘cheatsheets’</a>
and package vignettes, like for <a
href="https://cran.r-project.org/web/packages/tidyr/vignettes/tidy-data.html">tidyr</a>.</p>
<p>Since we loaded the libraries into our R session, we can see
documentation out using the <code>?</code> operator.</p>
<pre class="r"><code>?`DESeq2-package`</code></pre>
<p><strong>Checkpoint</strong>: <em>If you see the R documentation for
DESeq2 pop up in your ‘help’ panel on the right, please indicate with
the green ‘check’ button. If not please use the red ‘x’ button.</em></p>
</div>
<div id="read-counts" class="section level2">
<h2>Read Counts</h2>
<p>Another key assumption for DESeq2 is that the analysis will start
with <a
href="http://bioconductor.org/packages/devel/bioc/vignettes/DESeq2/inst/doc/DESeq2.html#why-un-normalized-counts">un-normalized
counts</a>.</p>
<p>To begin we’ll read in a <strong>raw</strong> count data file,
<code>gene_expected_count.txt</code> which is similar to what would be
generated in the alignment steps (and what you would receive from AGC).
We’ll discuss a few normalizations that can be helpful for understanding
broad patterns in our expression data, but raw data <strong>should
always</strong> be used as an input for DESeq2.</p>
<pre class="r"><code>count_table = read.table(&quot;data/gene_expected_count.txt&quot;, header = TRUE, row.names = 1)</code></pre>
<pre class="r"><code>head(count_table, n=2) # look at the top of the table</code></pre>
<pre><code>                   sample_A sample_B sample_C sample_D sample_E sample_F
ENSMUSG00000000001     1041      905     1296     3481     1283     1921
ENSMUSG00000000003        0        0        0        0        0        0</code></pre>
<p>Now that the file is read into R, note that we’ve created a data
frame that includes ‘gene ids’ in ENSEMBL format as rownames and count
data from twelve different samples.</p>
<p>How many genes were included in our reference? We can get a sense of
the size of our data with the <code>str</code> function.</p>
<pre class="r"><code>str(count_table) # look at the size &amp; structure of the table</code></pre>
<pre><code>&#39;data.frame&#39;:   55492 obs. of  6 variables:
 $ sample_A: int  1041 0 1043 1819 19 18 14972 1 888 402 ...
 $ sample_B: int  905 0 1232 914 11 1 14768 0 607 483 ...
 $ sample_C: int  1296 0 1664 1618 18 4 21026 0 911 744 ...
 $ sample_D: int  3481 0 2690 8618 48 24 22962 0 1689 898 ...
 $ sample_E: int  1283 0 1825 1350 37 1 22263 0 738 811 ...
 $ sample_F: int  1921 0 2720 1222 29 1 23622 0 1180 1261 ...</code></pre>
<p>One important consideration for RNA-seq data is how many genes we
measured in our data and how well we measured them. Both are impacted by
sequencing depth or the number of reads generated per sample.</p>
<p>Generally, for human and mouse experiments the recommendation is
30-40 million reads per sample to capture both highly expressed
(abundant) and lowly expressed (rarer) transcripts in the sample,
assuming ~25,000 protein-coding genes would be expected for a polyA
(mRNA targeting) library prep. However, as the image from Illumina below
shows, sequencing depth has less of an impact than number of replicates
in detecting differentially expressed genes (DEGs).</p>
<div class="float">
<img src="images/de_replicates_img.png" style="width:50.0%"
alt="Illumina’s differential expression recovery across replicate number and sequencing depth" />
<div class="figcaption">Illumina’s differential expression recovery
across replicate number and sequencing depth</div>
</div>
<p><br></p>
<p>How many reads were generated per sample in our data? We can use the
<code>summarize_all</code> function from the dplyr package to determine
the number of reads or “depth” per sample.</p>
<pre class="r"><code>count_table %&gt;% summarize_all(sum)</code></pre>
<pre><code>  sample_A sample_B sample_C sample_D sample_E sample_F
1 30149372 27849704 39153147 49383769 33895207 39253330</code></pre>
</div>
</div>
<div id="getting-help" class="section level1">
<h1>Getting help</h1>
<p>How did we know to use the <code>summarize_all</code> function to sum
each column?</p>
<p>R and RStudio have a strong community component so, as mentioned in
Computational Foundations, if you are getting an error or wondering how
to make a command work or how to perform a specific task, there is
likely already a solution out there. Search engines are your friend,
which is exactly how we found <a
href="https://stackoverflow.com/questions/28873057/sum-across-multiple-columns-with-dplyr">that
approach</a></p>
<p>It can sometimes be a challenge to figure out <em>what to search
for</em>, so some key parts of a successful search are:</p>
<ul>
<li>Package or command run</li>
<li><code>R</code> or <code>Bioconductor</code></li>
<li>The error message if there is one</li>
<li>Version information</li>
</ul>
<p>How to get session information to aid in a search:</p>
<pre class="r"><code>sessionInfo()</code></pre>
<p>Highly recommend using resources like <a
href="https://support.bioconductor.org/">Bioconductor Support</a>, <a
href="https://www.biostars.org/">Biostars</a>, and <a
href="https://stackoverflow.com/questions/tagged/r">Stack Overflow</a>,
including threads on specific packages or common bioinformatic
tasks.</p>
<p>I personally use one or more of these resources <strong>every
day</strong>.</p>
<p>10x Genomics also has a helpful <a
href="https://www.10xgenomics.com/resources/analysis-guides/10-tips-for-biologists-learning-bioinformatics">10
tips for biologists learning bioinformatics</a> included in their
resources.</p>
</div>
<div id="summary" class="section level1">
<h1>Summary</h1>
<p>In this section, we:</p>
<ul>
<li>Set up our compute environment</li>
<li>Learned about the DESeq2 package</li>
<li>Read in a raw count table and saved it as a data frame</li>
</ul>
<p>Now that we have our count data processed, we can move on to
“unblinding” our data, as the sample names are unique identifiers
generated by a sequencing center and not very informative as far as our
experimental conditions.</p>
<hr />
</div>
<div id="optional-content" class="section level1">
<h1>Optional content</h1>
<p>An important note is that there are several bonus content sections on
the instruction pages, like the two below that we will not be covering
in this workshop, but that may have useful context or be helpful when
you review this material.</p>
<details>
<summary>
<em>Click for alternative DESeq2 input options for RSEM outputs</em>
</summary>
The package <code>tximport</code> is another option<a
href="https://support.bioconductor.org/p/90672/">recommended the DESeq2
authors</a> to read in the RSEM expected_counts, as this package allows
for the average transcript length per gene to be used in the DE analysis
and, as <a href="https://support.bioconductor.org/p/88763/">described by
the author</a>, the <code>tximport-to-DESeqDataSet</code> constructor
function round the non-integer data generated by RSEM to whole numbers.
</details>
<br>
<details>
<summary>
<em>Click for comparison of RNA-seq data and microarray data</em>
</summary>
<p>With <a
href="https://www.illumina.com/science/technology/next-generation-sequencing/microarray-rna-seq-comparison.html">higher
sensitivity, greater flexiblity, and decreasing cost</a>, sequencing has
largely replaced microarray assays for measuring gene expression. A key
difference between the platforms is that microarrays measure intensities
and are therefore <em>continous</em> data while the count data from
sequencing is <em>discrete</em>. A more detailed comparison between
microarrays and sequencing technologies/analysis is outlined in <a
href="https://online.stat.psu.edu/stat555/node/30/">the online materials
for Penn State’s STAT555 course</a></p>
</details>
<p><br></p>
</div>
<div id="sources" class="section level1">
<h1>Sources</h1>
<p>Training resources used to develop materials:</p>
<ul>
<li>HBC DGE setup: <a
href="https://hbctraining.github.io/DGE_workshop/lessons/01_DGE_setup_and_overview.html"
class="uri">https://hbctraining.github.io/DGE_workshop/lessons/01_DGE_setup_and_overview.html</a></li>
<li>HBC Count Normalization: <a
href="https://hbctraining.github.io/DGE_workshop/lessons/02_DGE_count_normalization.html"
class="uri">https://hbctraining.github.io/DGE_workshop/lessons/02_DGE_count_normalization.html</a></li>
<li>DESeq2 standard vignette: <a
href="http://bioconductor.org/packages/devel/bioc/vignettes/DESeq2/inst/doc/DESeq2.html"
class="uri">http://bioconductor.org/packages/devel/bioc/vignettes/DESeq2/inst/doc/DESeq2.html</a></li>
<li>DESeq2 beginners vignette: <a
href="https://bioc.ism.ac.jp/packages/2.14/bioc/vignettes/DESeq2/inst/doc/beginner.pdf"
class="uri">https://bioc.ism.ac.jp/packages/2.14/bioc/vignettes/DESeq2/inst/doc/beginner.pdf</a></li>
<li>Bioconductor RNA-seq Workflows: <a
href="https://www.bioconductor.org/help/course-materials/2015/LearnBioconductorFeb2015/B02.1_RNASeq.html"
class="uri">https://www.bioconductor.org/help/course-materials/2015/LearnBioconductorFeb2015/B02.1_RNASeq.html</a></li>
<li>CCDL Gastric cancer training materials: <a
href="https://alexslemonade.github.io/training-modules/RNA-seq/03-gastric_cancer_exploratory.nb.html"
class="uri">https://alexslemonade.github.io/training-modules/RNA-seq/03-gastric_cancer_exploratory.nb.html</a></li>
<li>CCDL Neuroblastoma training materials: <a
href="https://alexslemonade.github.io/training-modules/RNA-seq/05-nb_cell_line_DESeq2.nb.html"
class="uri">https://alexslemonade.github.io/training-modules/RNA-seq/05-nb_cell_line_DESeq2.nb.html</a></li>
</ul>
<hr />
<p>These materials have been adapted and extended from materials listed
above. These are open access materials distributed under the terms of
the <a href="http://creativecommons.org/licenses/by/4.0/">Creative
Commons Attribution license (CC BY 4.0)</a>, which permits unrestricted
use, distribution, and reproduction in any medium, provided the original
author and source are credited.</p>
<br/> <br/>
<hr/>
<table style="width:100%;">
<colgroup>
<col width="28%" />
<col width="42%" />
<col width="28%" />
</colgroup>
<thead>
<tr class="header">
<th align="left"><a href="Module05_Additional_Details.html">Previous
lesson</a></th>
<th align="center"><a href="#top">Top of this lesson</a></th>
<th align="right"><a href="Module07_DESeq2Init.html">Next
lesson</a></th>
</tr>
</thead>
<tbody>
</tbody>
</table>
</div>

<div id="rmd-source-code">---
title: "Differential Expression (DE) Introduction"
author: "UM Bioinformatics Core"
date: "`r Sys.Date()`"
output:
        html_document:
            includes:
                in_header: header.html
            theme: paper
            toc: true
            toc_depth: 4
            toc_float: true
            number_sections: false
            fig_caption: true
            markdown: GFM
            code_download: true
---

<style type="text/css">
body, td {
   font-size: 18px;
}
code.r{
  font-size: 12px;
}
pre {
  font-size: 12px
}
</style>

```{r, include = FALSE}
source("../bin/chunk-options.R")
knitr_fig_path("06-")
```

In this module, we will learn:

* How to set up a differential expression analysis, using reproducible research principles  
* What is DESeq2 and why it is widely used for differential expression comparisons
* How to load and review gene count data

<br>

# Differential Expression Workflow {.unlisted .unnumbered}

Today we will proceed through key steps in a differential expression analysis, starting from a count table that's similar to what you generated in the first half of the workshop and one of the outputs included in the [data that the Advanced Genomics Core delivers](https://brcf.medicine.umich.edu/cores/advanced-genomics/data-delivery/) for RNA-seq libraries.

<img src="images/wayfinder/wayfinder-DEAnalysisSetup.png" alt="wayfinder" style="width: 1000px;"/>

# Getting Started in RStudio

## Log on to RStudio Server

Open a web browser to the following URL: [http://bfx-workshop01.med.umich.edu](http://bfx-workshop01.med.umich.edu)

You should now be looking at a page that will allow you to login to the RStudio server:

<img src="images/rstudio_login_screen.png" alt="rstudio default session" style="width: 1000px;"/>

Enter your user credentials and click <kbd>Sign In</kbd>. The credentials are the same used earlier in the workshop and were provided via an email entitled "UM BioinfCore Workshop Login".

However, if you forget yours, a helper can retrieve it for you if you ask in Slack. You should now see the RStudio interface:

<img src="images/rstudio_session_default.png" alt="rstudio default session" style="width:1000px;"/>

<br>

## What are we working towards?

To refresh ourselves on the interface of RStudio, look in the lower right pane, that has a list of files and folders. Click on the `RSD_R` folder and then click on the `WelcomeToRSD.R` script to open it.

There should now be four panes in your RStudio window, with the `WelcomeToRSD` script in the upper left pane, this is the "Source" or "Scripts" pane. It displays the code that we will write to perform our analysis.

<img src="images/WelcomeScriptOpened.png" alt="rstudio default session" style="width:1000px;"/>

In the Scripts pane, there is a line of icons, towards the right side of the pane there is a <kbd>Run</kbd> button. First highlight all the code in the Source pane, and then click <kbd>Run</kbd>.

<br>

**Checkpoint**

What just happened?

1. We see in that our code has been run in the Console panel, including some warning messages.

1. We see that a number of objects were created in our Environment panel including an object named `DE_results`.

1. We see that the Plot tab now has a volcano plot displayed - Note: if the plot is not displayed, you may need to expand the right panel boundary before re-running the last line of the script

**What does this volcano plot show?**

*Now that we have an idea of what we are working towards, any questions?*

<br>
<br>

## Create an RStudio Project

Just like in Computational Foundation workshop, let's start our analysis by creating a project where we can store our data, scripts and outputs.

1. To create a project, go to the <kbd>File</kbd> menu, and click <kbd>New Project...</kbd>. The following window will appear:

<img src="images/new_project_window.png" alt="new project window" style="width: 600px;"/>

2. In this window, select <kbd>Existing Directory</kbd>. For "Project working directory", click <kbd>Browse...</kbd>, select the "RSD_R" folder, and click <kbd>Choose</kbd>. This will use the `/home/workshop/user/RSD_R` folder as the project directory.

3. Finally click <kbd>Create Project</kbd>. In the "Files" tab of your output pane (more about the RStudio layout in a moment), you should see an RStudio project file, **RSD_R.Rproj**. All RStudio projects end with the "**.Rproj**" file extension.

Note that there is already a `data/` folder which contains the data we will use for these lessons.

## Creating an R script

Let's create an R script file:

- Click the <kbd>File</kbd> menu and select <kbd>New File</kbd> and then <kbd>R Script</kbd>.
- Before we go any further, save your script by clicking the save/disk icon that is in the bar above the first line in the script editor, or click the <kbd>File</kbd> menu and select <kbd>Save</kbd>.
- In the "Save File" window that opens, select <kbd>New Folder</kbd>. Name it **"scripts"**.
- Finally, name your file **"diffex"** in the "File name" field.

The new script **diffex.R** is now in the `scripts` folder. You can see that by clicking the `scripts` folder in the "Files" pane. And you can go back up to the main project folder by clicking the `..` to the right of the up arrow in the "Files" pane. By convention, R scripts end with the file extension **.R**.


## File organization & reproducible research

Part of any successful bioinformatics analysis is documentation and data stewardship, which are important aspects of "reproducible research" (see: [Nobel, 2009](https://journals.plos.org/ploscompbiol/article?id=10.1371/journal.pcbi.1000424)).

To follow best practices for file organization for reproducible research, we should make distinct locations for:

* Raw data
* Code
* Output


To organize our files , we'll create some more directories in the `RSD_R` folder which is now our project root and current working directory. We can already see that there a `data/` folder and we just created the `scripts/` folder with the RStudio GUI. Let's create an `outputs/` folder using the `dir.create()` function in R.

```{r create_dirs, eval = FALSE}
dir.create('outputs')
dir.create('outputs/figures')
dir.create('outputs/tables')
```

**Checkpoint**: *Please use the green 'check' if you have saved your code file and see the `data`, `scripts`, and `outputs` directories in the `RSD_R` folder and the red 'x' if you do not.*

> ## Reminder: RStudio code execution {.unlisted .unnumbered}
>
> <kbd>Ctrl</kbd>+<kbd>Enter</kbd> is a standard shortcut in RStudio to send the current line (or selected lines) to the console. If you see `>` in the Console, then R has executed the command. If you see a `+`, this means that the command is not complete, R thinks there is more to your command. You can use the <kmd>esc</kmd> to get out of this state.

> ## Reminder: Object naming conventions {.unlisted .unnumbered}
> * Cannot start with numbers
> * Cannot include dashes
> * Cannot have spaces
> * Should not be identical to a named function
> * Dots and underscores can separate parts of names, alternatively CamelCase accomplishes this

----

# Tools for Differential Gene Expression analysis

As discussed during the webinar, bulk RNA-seq can be used to test for differential expression between conditions or treatments, using statistical approaches that are appropriate for biological data.

While there are several tools that can be used for differential expression comparisons, we will use [DESeq2](https://bioconductor.org/packages/release/bioc/html/DESeq2.html) in our analysis today.

DESeq2 is one of two tools, along with [edgeR](https://bioconductor.org/packages/release/bioc/html/edgeR.html), considered ['best practice'](https://bmcbioinformatics.biomedcentral.com/articles/10.1186/1471-2105-14-91) for differential expression comparisons. Both tools apply similar methods that account for the "shape" of data expected for RNA-seq and are fairly stringent in calling differentially expressed genes, lowering the risk of investigating "false positive" genes (e.g. genes that don't actually have different expression between treatment groups and therefore are not relevant to the biological process).

Additionally, DESeq2 also has an [excellent vignette](https://bioconductor.org/packages/release/bioc/vignettes/DESeq2/inst/doc/DESeq2.html)
from Love, Anders, and Huber, from which our workflow is partially adapted, and is a good resource when analyzing your own data
(see also: [Love, Anders, and Huber. _Genome Biology_. 2014.](https://doi.org/10.1186/s13059-014-0550-8)).


## DESeq2 assumptions and requirements

A key assumption of DESeq2 is that *biological variance is much greater than technical variance* (which should be true if [best practices](https://www.txgen.tamu.edu/faq/rna-isolation-best-practices/) for [quality RNA isolation](https://www.biocompare.com/Bench-Tips/128790-Four-Tips-for-Perfecting-RNA-Isolation/) are followed, including DNase treatment!).

Since calculating variance is key to the statistical approach used for DESeq2, if we tried to compare treatment groups with only one sample each, we would get an error (as shown in [this blog post](https://support.bioconductor.org/p/89746/)). Without replicates, there can't be statistical significance (e.g. p-values), but *qualitative* approaches are an option, like looking at the top expressed genes after normalization.

<details>
    <summary>*Click for additional resources regarding statistical testing and tool comparison for RNA-seq data*</summary>
    To learn more about statistical testing and what distributions best model the behavior of RNA-seq data, a good resource is this [EdX lecture by Rafael Irizarry](https://www.youtube.com/watch?v=HK7WKsL3c2w&feature=youtu.be) or this [lecture by Kasper Hansen](https://www.youtube.com/watch?v=C8RNvWu7pAw). Another helpful guide is this [Comparative Study for Differential Expression Analysis by Zhang et al.](https://journals.plos.org/plosone/article?id=10.1371/journal.pone.0103207) from 2014.
</details>
<br>

> ## Exercise
>
> - Post a question that was NOT addressed but that you hope we will address in the later modules **OR**
> - Add a thumbs up to your favorite comment(s) to upvote it

----

# Getting started

## Load Packages

Several packages have already been installed on the server, so we can load them into our R session now. To do that we'll use the `library` function to load the required packages.

```{r Modules, message=FALSE, warning= FALSE, eval=TRUE}
library(DESeq2)
library(ggplot2)
library(tidyr)
library(dplyr)
library(matrixStats)
library(ggrepel)
library(pheatmap)
library(RColorBrewer)
library(data.table)
```

*Note: We expect to see some red messages in your console while these packages are loading*

R/RStudio [AKA Posit] has great resources for getting help, including [code 'cheatsheets'](https://posit.co/resources/cheatsheets/) and package vignettes, like for [tidyr](https://cran.r-project.org/web/packages/tidyr/vignettes/tidy-data.html).

Since we loaded the libraries into our R session, we can see documentation out using the `?` operator.
```{r CheckDocumentaion}
?`DESeq2-package`
```

**Checkpoint**: *If you see the R documentation for DESeq2 pop up in your 'help' panel on the right, please indicate with the green 'check' button. If not please use the red 'x' button.*

## Read Counts

Another key assumption for DESeq2 is that the analysis will start with [un-normalized counts](http://bioconductor.org/packages/devel/bioc/vignettes/DESeq2/inst/doc/DESeq2.html#why-un-normalized-counts).

To begin we'll read in a **raw** count data file, `gene_expected_count.txt` which is similar to what would be generated in the alignment steps (and what you would receive from AGC). We'll discuss a few normalizations that can be helpful for understanding broad patterns in our expression data, but raw data **should always** be used as an input for DESeq2.

```{r DataTable, eval=TRUE, echo=FALSE}
## need to include paths relative to site build locations but keep code block hidden
count_table = read.table("../data/R_data/gene_expected_count.txt", header = TRUE, row.names = 1)
```

```{r DataTable2, eval=FALSE}
count_table = read.table("data/gene_expected_count.txt", header = TRUE, row.names = 1)
```

```{r DataTableCheck, eval=TRUE}
head(count_table, n=2) # look at the top of the table
```

Now that the file is read into R, note that we've created a data frame that includes 'gene ids' in ENSEMBL format as rownames and count data from twelve different samples.

How many genes were included in our reference? We can get a sense of the size of our data with the `str` function.
```{r DataTableCheck2, eval=TRUE}
str(count_table) # look at the size & structure of the table
```


One important consideration for RNA-seq data is how many genes we measured in our data and how well we measured them. Both are impacted by sequencing depth or the number of reads generated per sample.

Generally, for human and mouse experiments the recommendation is 30-40 million reads per sample to capture both highly expressed (abundant) and lowly expressed (rarer) transcripts in the sample, assuming ~25,000 protein-coding genes would be expected for a polyA (mRNA targeting) library prep. However, as the image from Illumina below shows, sequencing depth has less of an impact than number of replicates in detecting differentially expressed genes (DEGs).

![Illumina's differential expression recovery across replicate number and sequencing depth](images/de_replicates_img.png){width=50%}

<br>

How many reads were generated per sample in our data? We can use the `summarize_all` function from the dplyr package to determine the number of reads or "depth" per sample.

```{r SummarizeReads, eval=TRUE}
count_table %>% summarize_all(sum)
```

# Getting help

How did we know to use the `summarize_all` function to sum each column?

R and RStudio have a strong community component so, as mentioned in Computational Foundations, if you are getting an error or wondering how to make a command work or how to perform a specific task, there is likely already a solution out there. Search engines are your friend, which is exactly how we found [that approach](https://stackoverflow.com/questions/28873057/sum-across-multiple-columns-with-dplyr)

It can sometimes be a challenge to figure out *what to search for*, so some key parts of a successful search are:

* Package or command run
* `R` or `Bioconductor`
* The error message if there is one
* Version information

How to get session information to aid in a search:
```{r Session info, eval = FALSE, message = FALSE}
sessionInfo()
```

Highly recommend using resources like [Bioconductor Support](https://support.bioconductor.org/), [Biostars](https://www.biostars.org/), and [Stack Overflow](https://stackoverflow.com/questions/tagged/r), including threads on specific packages or common bioinformatic tasks.

I personally use one or more of these resources **every day**.

10x Genomics also has a helpful [10 tips for biologists learning  bioinformatics](https://www.10xgenomics.com/resources/analysis-guides/10-tips-for-biologists-learning-bioinformatics) included in their resources.

# Summary

In this section, we:

* Set up our compute environment
* Learned about the DESeq2 package
* Read in a raw count table and saved it as a data frame

Now that we have our count data processed, we can move on to "unblinding" our data, as the sample names are unique identifiers generated by a sequencing center and not very informative as far as our experimental conditions.

---

# Optional content

An important note is that there are several bonus content sections on the instruction pages, like the two below that we will not be covering in this workshop, but that may have useful context or be helpful when you review this material.

<details>
    <summary>*Click for alternative DESeq2 input options for RSEM outputs*</summary>
    The package `tximport` is another option[recommended the DESeq2  authors](https://support.bioconductor.org/p/90672/) to read in the RSEM expected_counts, as this  package allows for the average transcript length per gene to be used in the DE analysis and, as [described by the author](https://support.bioconductor.org/p/88763/), the `tximport-to-DESeqDataSet` constructor function round the non-integer data generated by RSEM to whole numbers.
</details>
<br>
<details>
    <summary>*Click for comparison of RNA-seq data and microarray data*</summary>
    With [higher sensitivity, greater flexiblity, and decreasing cost](https://www.illumina.com/science/technology/next-generation-sequencing/microarray-rna-seq-comparison.html), sequencing has largely replaced microarray assays for measuring gene expression. A key difference between the platforms is that microarrays measure intensities and are therefore *continous* data while the count data from sequencing is *discrete*. A more detailed comparison between microarrays and sequencing technologies/analysis is outlined in [the online materials for Penn State's STAT555 course](https://online.stat.psu.edu/stat555/node/30/)

</details>
<br>

# Sources

Training resources used to develop materials:

* HBC DGE setup: https://hbctraining.github.io/DGE_workshop/lessons/01_DGE_setup_and_overview.html
* HBC Count Normalization: https://hbctraining.github.io/DGE_workshop/lessons/02_DGE_count_normalization.html
* DESeq2 standard vignette: http://bioconductor.org/packages/devel/bioc/vignettes/DESeq2/inst/doc/DESeq2.html
* DESeq2 beginners vignette: https://bioc.ism.ac.jp/packages/2.14/bioc/vignettes/DESeq2/inst/doc/beginner.pdf
* Bioconductor RNA-seq Workflows: https://www.bioconductor.org/help/course-materials/2015/LearnBioconductorFeb2015/B02.1_RNASeq.html
* CCDL Gastric cancer training materials: https://alexslemonade.github.io/training-modules/RNA-seq/03-gastric_cancer_exploratory.nb.html
* CCDL Neuroblastoma training materials: https://alexslemonade.github.io/training-modules/RNA-seq/05-nb_cell_line_DESeq2.nb.html


```{r WriteOut.RData, eval=TRUE, echo=FALSE, message=FALSE}
#Hidden code block to write out data for knitting

# dir.create("rdata", showWarnings=FALSE)
# save.image(file = "rdata/RunningData.RData")
# How to load for next segment
# load("rdata/RunningData.RData")
```

---

These materials have been adapted and extended from materials listed above. These are open access materials distributed under the terms of the [Creative Commons Attribution license (CC BY 4.0)](http://creativecommons.org/licenses/by/4.0/), which permits unrestricted use, distribution, and reproduction in any medium, provided the original author and source are credited.

<br/>
<br/>
<hr/>
| [Previous lesson](Module05_Additional_Details.html) | [Top of this lesson](#top) | [Next lesson](Module07_DESeq2Init.html) |
| :--- | :----: | ---: |</div>


</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->
<script>
$(document).ready(function () {
  window.initializeSourceEmbed("Module06_DEAnalysisSetup.Rmd");
});
</script>

<script>
$(document).ready(function ()  {

    // temporarily add toc-ignore selector to headers for the consistency with Pandoc
    $('.unlisted.unnumbered').addClass('toc-ignore')

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3,h4",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
