<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />


<meta name="author" content="UM Bioinformatics Core" />

<meta name="date" content="2024-07-26" />

<title>DE Testing</title>

<script src="site_libs/header-attrs-2.25/header-attrs.js"></script>
<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/paper.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/jqueryui-1.13.2/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<script src="site_libs/navigation-1.1/codefolding.js"></script>
<script src="site_libs/navigation-1.1/sourceembed.js"></script>
<link href="site_libs/highlightjs-9.12.0/default.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link rel="shortcut icon" href="favicon-16x16.png" />
<!--
Favicon from
https://twemoji.twitter.com/
https://favicon.io/emoji-favicons/dna/
-->

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>




<style type="text/css">
#rmd-source-code {
  display: none;
}
</style>





<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
details > summary > p:only-child {
  display: inline;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark the anchor link active (and if it's in a dropdown, also mark that active)
  var dropdown = menuAnchor.closest('li.dropdown');
  if (window.bootstrap) { // Bootstrap 4+
    menuAnchor.addClass('active');
    dropdown.find('> .dropdown-toggle').addClass('active');
  } else { // Bootstrap 3
    menuAnchor.parent().addClass('active');
    dropdown.addClass('active');
  }

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before, .tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "\e259";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "\e258";
  font-family: 'Glyphicons Halflings';
  border: none;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->
<style type="text/css">
.code-folding-btn { margin-bottom: 4px; }
</style>



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-inverse  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-bs-toggle="collapse" data-target="#navbar" data-bs-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">RNA-Seq Demystified</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="Module00_Introduction.html">Intro</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Sequence QC
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="Module01_Warming_Up.html">Warming up</a>
    </li>
    <li>
      <a href="Module02a_Sequence_QC.html">Sequence QC</a>
    </li>
    <li>
      <a href="Module02b_Trimming.html">Trimming</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Alignment
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="Module03a_Reference_Genomes.html">Reference Genomes</a>
    </li>
    <li>
      <a href="Module03b_Alignment.html">Alignment</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Quantification
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="Module04_Alignment_QC_and_Quantification.html">Alignment QC and Quantification</a>
    </li>
    <li>
      <a href="Module05_Additional_Details.html">Optional: Additional details</a>
    </li>
  </ul>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Groups/Model
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="Module06_DEAnalysisSetup.html">DE Introduction</a>
    </li>
    <li>
      <a href="Module07_DESeq2Init.html">DE Initialization</a>
    </li>
    <li>
      <a href="Module08_DESeq2DE.html">DE Normalization and Modeling</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Expression plots
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="Module09_SampleQCViz.html">Sample QC</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Diff Expression
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="Module10_DEComparisons.html">DE Testing</a>
    </li>
    <li>
      <a href="Module10a_batchAndCovars.html">Batches and Covariates</a>
    </li>
    <li>
      <a href="Module11_DEVisualizations.html">DE Visualization</a>
    </li>
    <li>
      <a href="Module12_DEAnnotations.html">DE Annotation</a>
    </li>
    <li>
      <a href="R_bonus_content.html">Bonus content</a>
    </li>
  </ul>
</li>
<li>
  <a href="Module99_Wrap_up.html">Wrap up</a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">

<div class="btn-group pull-right float-right">
<button type="button" class="btn btn-default btn-xs btn-secondary btn-sm dropdown-toggle" data-toggle="dropdown" data-bs-toggle="dropdown" aria-haspopup="true" aria-expanded="false"><span>Code</span> <span class="caret"></span></button>
<ul class="dropdown-menu dropdown-menu-right" style="min-width: 50px;">
<li><a id="rmd-download-source" href="#">Download Rmd</a></li>
</ul>
</div>



<h1 class="title toc-ignore">DE Testing</h1>
<h4 class="author">UM Bioinformatics Core</h4>
<h4 class="date">2024-07-26</h4>

</div>


<style type="text/css">
body, td {
   font-size: 18px;
}
code.r{
  font-size: 12px;
}
pre {
  font-size: 12px
}
</style>
<p>In this module, we will learn:</p>
<ul>
<li>What a p-value does (and does not!) represent</li>
<li>How to modify and fit our DE model</li>
<li>How to generate tables of DE statistics</li>
<li>The importance multiple hypothesis corrections</li>
</ul>
<p><br></p>
<div id="differential-expression-workflow"
class="section level1 unlisted unnumbered">
<h1 class="unlisted unnumbered">Differential Expression Workflow</h1>
<p>Now we test for differential expression between our groups of
interest and return a table of results.</p>
<p><img src="images/wayfinder/wayfinder-DEComparisons.png"
style="width:75.0%" /></p>
<hr />
</div>
<div id="testing-for-de" class="section level1">
<h1>Testing for DE</h1>
<p>Before showing the code for generating differential expression
results, let’s walk through some toy data and discuss our intuition, its
limitations, and what we can do to formalize our thought around
determining if a gene is differentially expressed.</p>
<p>Let’s start by looking at expression boxplots for a few “genes” where
we’ve made up the data. In each plot, we are comparing the expression
levels (on the y-axis) for samples (each point) representing a WT (red)
and KO (teal) condition. The boxplot shows the 25% - 75% distribution,
along with outliers, with a bar representing the median value, and a
black-crossed-point representing the mean.</p>
<p>For the first example, we ask: Does the KO affect the expression of
the gene?</p>
<p><img src="images/Module10_stat_plot_Gene_3.png"
style="width:75.0%" /></p>
<p>It seems pretty clear that there is a large difference between the
means of the two groups, and within-group spread is quite low. So the
answer to the question is likely “Yes”.</p>
<p>Consider a second example:</p>
<p><img src="images/Module10_stat_plot_Gene_1.png"
style="width:75.0%" /></p>
<p>Here there isn’t a large difference between the means of the two
groups, and there is quite a bit of spread within-group. The answer here
is likely “No”.</p>
<p>Finally, consider a third example:</p>
<p><img src="images/Module10_stat_plot_Gene_2.png"
style="width:75.0%" /></p>
<p>The means are not so close to each other in this example, though
there is still quite a bit of spread. This is example is perhaps less
clear.</p>
<p>Looking at all three of the hypothetical genes together, we see them
in relation to one another and how they span the range from No, to
Maybe?, to Definitely.</p>
<p><img src="images/Module10_stat_plot.png" /></p>
<p>Now, consider needing to make this decision for 20,000 genes. Even if
all the genes had clear separation, that would take a lot of time.
However, it’s reasonable to expect to see a lot of genes with more
ambiguous expression like Gene 2.</p>
<p><strong>We need a formal, reproducible, way to make this
decision!</strong></p>
<p>Thankfully statistics provides the formality we want, but for any
statistical test we need to clearly state what we are testing. When
testing for differential expression we assume that for any particular
gene, there is no difference in expression between conditions.
Statisticians would call this the <strong>“null
hypothesis”</strong>.</p>
<p>For each gene, DESeq2 computes a “Wald statistic” which is a single
number encapsulating the difference in the means and the spread of the
groups. However, this number alone doesn’t allow us to determine if a
gene is differentially expressed. We need a <strong>second</strong>
number to give us an idea of how extreme that statistic is among the
distribution of possible statistics, which is the
<strong>“p-value”</strong>. The <strong>“p-value”</strong> along with
choosing a threshold will inform us if we can <strong>reject the null
hypothesis</strong> and classify a gene as differentially expressed.</p>
<p>Imagine shuffling the group labels and recomputing the Wald statistic
over and over again. You’d get a distribution of statistics that would
look similar to a normal curve. The p-value essentially tells you how
likely you are to have seen the statistic by chance. So when we set a
p-value = 0.05 as a threshold, we’re saying, “there is a 5% chance I’d
see something this extreme when there was actually no effect”. So the
evidence is <strong>strong</strong>, but not
<strong>ironclad</strong>.</p>
</div>
<div id="deseq2-model-fitting" class="section level1">
<h1>DESeq2 Model Fitting</h1>
<p>Based on our PCA results, we’ll copy our DESeq object and revise our
original model design to add a covariate for “batch”. Then, we’ll fit
our model using the <code>DESeq</code> function and take a look at the
objects we generate. This command applies the model to our data, using
the sample information supplied and model <code>dds_batch</code> object
so can take some time to run.</p>
<pre class="r"><code>#  access model slot from original `dds_filtered` DESeq2 object
design(dds_filtered)</code></pre>
<pre><code>~condition</code></pre>
<pre class="r"><code># copy original object and then modify model
dds_batch_filtered = dds_filtered
design(dds_batch_filtered) &lt;- ~ condition + batch

# after updating model, fit model with `DESeq` function
?DESeq
dds_batch_fitted = DESeq(dds_batch_filtered)
dds_batch_fitted</code></pre>
<pre><code>class: DESeqDataSet 
dim: 16249 6 
metadata(1): version
assays(4): counts mu H cooks
rownames(16249): ENSMUSG00000000001 ENSMUSG00000000028 ... ENSMUSG00000118651 ENSMUSG00000118653
rowData names(26): baseMean baseVar ... deviance maxCooks
colnames(6): sample_A sample_B ... sample_E sample_F
colData names(4): genotype condition batch sizeFactor</code></pre>
<!-- add as bonus: Accessing functions available for specific object types https://rpubs.com/mikelove/pkgmethods 
# see what functions are available for accessing/modifying our DESeqDataSet objects
showMethods(classes="DESeqDataSet", where=getNamespace("DESeq2"))-->
<p>While we won’t be able to delve into model options in detail, more
complex model designs including adding “interaction terms” between
multiple group labels, are helpfully described in <a
href="https://support.bioconductor.org/p/98628/">this support thread</a>
as well as in the <a
href="https://bioconductor.org/packages/release/bioc/vignettes/DESeq2/inst/doc/DESeq2.html#interactions">DESeq2
vignette</a>.</p>
<details>
<summary>
<em>Click for example of model design check</em>
</summary>
<p>When including multiple terms in our model it’s helpful to check the
corresponding design matrix to ensure that our batches are not
confounded, which would cause <code>DESeq2</code> to return a
<code>model not full rank</code> error when attempting to fit the model.
We can do that using the <code>model.matrix</code> function, providing
our intended model and our sample information.</p>
<pre class="r"><code>model.matrix(~ condition + batch, samplesheet_batch)</code></pre>
<pre><code>         (Intercept) conditiondeficient batchDay2
sample_A           1                  0         0
sample_B           1                  0         1
sample_C           1                  0         1
sample_D           1                  1         0
sample_E           1                  1         1
sample_F           1                  1         1
attr(,&quot;assign&quot;)
[1] 0 1 2
attr(,&quot;contrasts&quot;)
attr(,&quot;contrasts&quot;)$condition
[1] &quot;contr.treatment&quot;

attr(,&quot;contrasts&quot;)$batch
[1] &quot;contr.treatment&quot;</code></pre>
When we look at the outputs, we can see that all of the returned columns
have values (<code>1</code>) included. If our model was not full rank,
then we would see a columns with no values (all <code>0</code>)
returned.
</details>
<p><br></p>
<details>
<summary>
<em>Dispersion models and possible warning messages</em>
</summary>
<p>Depending on the data set you are analyzing, you may see a warning
that the default ‘parametric’ dispersion model so a local regression was
substituted. When seeing this warning, we recommend looking at a
dispersion plot with the <code>plotDispEsts(dds)</code> function, but as
<a href="https://support.bioconductor.org/p/107937/">this bioconductor
thread</a> discusses, other visualizations of our data might be more
helpful and/or easier to interpret why the data doesn’t fit the default
parametric model.</p>
</details>
<p><br> <br></p>
<details>
<summary>
<em>Click for code for a plot of dispersion estimates</em>
</summary>
<p>We can visualize the <strong>dispersion estimates</strong> with the
<code>plotDispEsts</code> function. This plot shows the the DESeq2
normalization results for our data, which centers on shrinking the
variance across all genes to better fit the expected spread at a given
expression level.</p>
<pre class="r"><code>plotDispEsts(dds_batch_fitted)</code></pre>
<p><img src="fig/10-CheckDispersions-1.png" width="816" style="display: block; margin: auto;" /></p>
<p>Above is the raw data plotted in black, the fitted (or expected)
dispersion in red, and the normalized data with scaled variance in blue.
Since we have fairly small sample sizes for each condition, we see
shrinkage for many genes but a reasonable correlation between the
expression level and dispersions.</p>
This <a
href="https://hbctraining.github.io/DGE_workshop/lessons/04_DGE_DESeq2_analysis.html">HBC
tutorial</a> has a more detailed overview of estimating size factors,
estimating gene dispersion, and the shrinkage procedure, as well as
examples of concerning dispersion plots that may suggest reassessing
quality of the experimental data.
</details>
<p><br></p>
<p>Notice that there is now more information in the
<code>DESeqDataSet</code> object than there was prior to our
normalization. There is information about the model fit and about the
library size normalization. DESeq2 will use this information when we
perform the test for differential expression.</p>
<p>The <code>DESeq()</code> function is actually doing three things
automatically for us. It calculates:</p>
<ol style="list-style-type: decimal">
<li>The size factors to normalize for library size with
<code>estimateSizeFactors(dds_filtered)</code>,</li>
<li>Dispersion estimates to shrink the dispersions with
<code>estimateDispersions(dds_filtered)</code>, and</li>
<li>The Wald test statistics with
<code>nbinomWaldTest(dds_filtered)</code>.</li>
</ol>
<p>The <code>resultsNames()</code> function returns the names of the
estimated effects of the model. Note that the results are stored based
on the levels we set in the sample sheet and that the convention is to
always have “case” first versus “control” second.</p>
<pre class="r"><code>resultsNames(dds_batch_fitted)</code></pre>
<pre><code>[1] &quot;Intercept&quot;                      &quot;condition_deficient_vs_control&quot; &quot;batch_Day2_vs_Day1&quot;            </code></pre>
<p>The results include the single comparison representing the two levels
of <code>condition</code>. If there were more levels in the
<code>condition</code> column, there would be more results listed here
because DESeq2 would implicitly compare all other levels to the
reference level.</p>
<p><strong>Checkpoint</strong>: <em>If you see the same results when you
execute <code>resultsNames(dds_batch_fitted)</code>, please indicate
with the green ‘yes’ button. Otherwise, please use the red ‘x’ button to
get help</em></p>
</div>
<div id="generating-de-results" class="section level1">
<h1>Generating DE Results</h1>
<p>There is only the one comparison in the results, so we will refer to
it in the <code>name</code> parameter of the <code>results()</code>
function, and assign the result as an object.</p>
<pre class="r"><code>results_deficient_vs_control = results(dds_batch_fitted, name = &#39;condition_deficient_vs_control&#39;)
head(results_deficient_vs_control)</code></pre>
<pre><code>log2 fold change (MLE): condition deficient vs control 
Wald test p-value: condition deficient vs control 
DataFrame with 6 rows and 6 columns
                      baseMean log2FoldChange     lfcSE      stat    pvalue      padj
                     &lt;numeric&gt;      &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt;
ENSMUSG00000000001  1489.83039       0.278343  0.148278  1.877165 0.0604955  0.324515
ENSMUSG00000000028  1748.93544       0.221671  0.128835  1.720588 0.0853256  0.385578
ENSMUSG00000000031  2151.87715       0.135823  0.284313  0.477724 0.6328467  0.866812
ENSMUSG00000000037    24.91672       0.599542  0.562067  1.066673 0.2861195        NA
ENSMUSG00000000049     7.78377      -1.227258  1.148327 -1.068735 0.2851890        NA
ENSMUSG00000000056 19653.54030      -0.201183  0.167344 -1.202215 0.2292803  0.596345</code></pre>
<p>If we look at the results table, we see that the row names are gene
identifiers (in this case ENSEMBL IDs because that’s what the GTF we
used in the call to RSEM+STAR) and then we see the following
columns:</p>
<ol style="list-style-type: decimal">
<li><code>baseMean</code> is the average of the normalized count values,
divided by size factors and taken over all samples, and can be
interpreted as the relative expression level of that gene across all
samples.</li>
<li><code>log2FoldChange</code> is the log2 transformed ratio of the
expression of the numerator group (first group) over the denominator
group (second group after “vs”). Note that in our comparison, the
<code>log2FoldChange</code> column compares the expression of
<code>deficient</code> over the denominator group, <code>control</code>.
If the value is positive, that means the expression of that gene is
greater across the <code>deficient</code> samples than across the
<code>control</code> samples. If the value is negative, that means the
expression of that gene is greater across the <code>deficient</code>
samples.</li>
<li><code>lfcSE</code> is the standard error for the log2 fold change
estimate.</li>
</ol>
<blockquote>
<h1 id="note-results-defaults" class="unlisted unnumbered">Note:
<code>results()</code> defaults</h1>
<p>If no arguments are passed to <code>results()</code>, then the log2
fold changes and Wald test p-value will be for the <strong>last
variable</strong> in the design formula, and if this is a factor, the
comparison will be the <strong>last level</strong> over the
<strong>reference level</strong>. If you specify <code>name</code>, as
we did above, then the behavior is given by the name used from
<code>resultsNames()</code>.</p>
</blockquote>
<p>There are multiple ways to specify the test to be done using the
<code>results()</code> function. It is especially helpful to know this
when fitting more complex models and testing more complex contrasts. To
demonstrate this, consider this description from the help for
<code>results()</code>:</p>
<blockquote>
<p><code>contrast</code>: a character vector with exactly three
elements: the name of a factor in the design formula, the name of the
numerator level for the fold change, and the name of the denominator
level for the fold change</p>
</blockquote>
<p>So an alternative way to test the same contrast as above
(i.e. <code>deficient</code> / <code>control</code>) is:</p>
<pre class="r"><code>alt_results_deficient_vs_control = results(dds_batch_fitted, contrast = c(&#39;condition&#39;, &#39;deficient&#39;, &#39;control&#39;))
head(alt_results_deficient_vs_control)</code></pre>
<pre><code>log2 fold change (MLE): condition deficient vs control 
Wald test p-value: condition deficient vs control 
DataFrame with 6 rows and 6 columns
                      baseMean log2FoldChange     lfcSE      stat    pvalue      padj
                     &lt;numeric&gt;      &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt;
ENSMUSG00000000001  1489.83039       0.278343  0.148278  1.877165 0.0604955  0.324515
ENSMUSG00000000028  1748.93544       0.221671  0.128835  1.720588 0.0853256  0.385578
ENSMUSG00000000031  2151.87715       0.135823  0.284313  0.477724 0.6328467  0.866812
ENSMUSG00000000037    24.91672       0.599542  0.562067  1.066673 0.2861195        NA
ENSMUSG00000000049     7.78377      -1.227258  1.148327 -1.068735 0.2851890        NA
ENSMUSG00000000056 19653.54030      -0.201183  0.167344 -1.202215 0.2292803  0.596345</code></pre>
<p>This way of calling <code>results()</code> is especially helpful when
the levels of the column of interest contain more than two levels
because you can specify exactly which levels to test with little
confusion.</p>
<p>If we continue to look at the results table, after the
<code>lfcSE</code> column we see:</p>
<ol start="4" style="list-style-type: decimal">
<li><code>stat</code> is the calculated Wald statistic for that
gene.</li>
<li><code>pvalue</code> is the <em>nominal</em> significance that we
described earlier.</li>
<li><code>padj</code> is the <em>adjusted p-value</em> (also known as a
“q-value”) and is what we use for determining significantly differently
expressed genes.</li>
</ol>
<blockquote>
<h1 id="question" class="unlisted unnumbered">Question</h1>
<p>Why should we use values from <code>padj</code> instead of the
<code>pvalue</code>? Post in the Slack thread.</p>
</blockquote>
<p><br></p>
<div id="multiple-hypothesis-testing-and-fdr-correction"
class="section level2">
<h2>Multiple hypothesis testing and FDR correction</h2>
<p>Each p-value reported in the table is the result of a single test for
a single gene. As stated earlier - with a significance cut-off of
p-value &lt; 0.05, we’re expecting a 5% chance it is a false positive.
However the more genes we test, the greater chance we have of seeing a
significant results by chance.</p>
<div id="the-multiple-hypothesis-problem" class="section level3">
<h3>The multiple hypothesis problem</h3>
<p>For example, if we consider running the Wald test for 20 genes
instead of just three, we might see something like this where
<code>Gene N</code> has a p-value below our threshold of 0.05. If we see
this pattern, do we expect that <code>Gene N</code> is actually
differentially expressed (e.g. a true positive)?</p>
<p><br> <img src="images/FDR-1.png" style="width:75.0%" /> <br></p>
<p>What if we tested 100 genes? How many genes would we expect to have
p-values &lt; 0.05, just by chance?</p>
<p><br> <img src="images/FDR-2.png" style="width:75.0%" /> <br></p>
<p>What if we tested 10,000 genes? We would expect to see 500 genes with
p-values &lt; 0.05, just by chance!</p>
<p><br> <img src="images/FDR-3.png" style="width:75.0%" /> <br></p>
<p><strong>So if we are testing 20,000 genes for differential
expression, we would expect to see ~1,000 genes (5%) with statistical
significance, just by chance.</strong> <br></p>
</div>
<div id="the-fdr-solution" class="section level3">
<h3>The FDR solution</h3>
<p>To address this <a
href="https://multithreaded.stitchfix.com/blog/2015/10/15/multiple-hypothesis-testing/">multiple
hypothesis correction</a> can be performed. While there are a few
approaches, the default method in DESeq2 is the False Discovery Rate
(FDR) (<a
href="https://rss.onlinelibrary.wiley.com/doi/10.1111/j.2517-6161.1995.tb02031.x">Benjamini
and Hochberg (1995)</a>).</p>
<p>The default FDR rate cutoff for our analyses is 0.05, meaning the
proportion of false positives amongst our <em>differentially expressed
genes</em> is controlled to 5%. So if we find 500 genes that are
differentially expressed with this FDR cutoff, we expect only 25 of them
to be false positives. DESeq2 vignette’s includes a <a
href="http://bioconductor.org/packages/devel/bioc/vignettes/DESeq2/inst/doc/DESeq2.html#independent-filtering-and-multiple-testing">further
discussion of filtering and multiple testing correction</a>.</p>
<blockquote>
<h1 id="note-on-padj-values-set-to-na" class="unlisted unnumbered">Note
on <code>padj</code> values set to NA</h1>
<p>As discussed in the <a
href="https://hbctraining.github.io/DGE_workshop/lessons/05_DGE_DESeq2_analysis2.html">HBC
tutorial</a> as well as the <a
href="http://bioconductor.org/packages/devel/bioc/vignettes/DESeq2/inst/doc/DESeq2.html#i-want-to-benchmark-deseq2-comparing-to-other-de-tools.">DESeq2
vignette</a>, DESeq2 reduces the number of genes that will be tested by
removing genes with low number of counts and outlier samples.</p>
<ul>
<li>If within a row, all samples have zero counts, the baseMean column
will be zero, and the log2 fold change estimates, p-value and adjusted
p-value will all be set to NA.</li>
<li>If a row contains a sample with an extreme count outlier then the
p-value and adjusted p-value will be set to NA. These outlier counts are
detected by <a
href="https://en.wikipedia.org/wiki/Cook%27s_distance">Cook’s
distance</a>.</li>
<li>If a row is filtered by automatic independent filtering, e.g. for
having a low mean normalized count, then only the adjusted p-value will
be set to NA.</li>
</ul>
</blockquote>
</div>
</div>
<div id="save-fitted-model-and-data-as-.rds" class="section level2">
<h2>Save fitted model and data as .rds</h2>
<p>It can be useful to save key R objects to file as we proceed through
our analysis - before we do that, let’s look at the documentation for
the <code>save</code> function to see if it does what we want.</p>
<pre class="r"><code>?save</code></pre>
<p>As we can see from the documentation the function <code>save</code>
writes an “external representation” of R objects that can be read back
from the file at a later date by using the function <code>load</code> or
<code>attach</code> in most cases. We’ll proceed with saving our
<code>dds_batch_fitted</code> object, creating a subdirectory first.</p>
<pre class="r"><code>dir.create(&quot;outputs/Robjs&quot;, recursive=TRUE)</code></pre>
<pre><code>Warning in dir.create(&quot;outputs/Robjs&quot;, recursive = TRUE): &#39;outputs/Robjs&#39; already exists</code></pre>
<pre class="r"><code>save(dds_batch_fitted,
          file=&quot;outputs/Robjs/dds_batch_fitted.rds&quot;)</code></pre>
<hr />
</div>
</div>
<div id="summary" class="section level1">
<h1>Summary</h1>
<p>In this section, we:</p>
<ul>
<li>Fitted our DESeq2 model</li>
<li>Performed statistical tests for comparisons of interest</li>
<li>Generated tables of differential expression results - i.e. fold
changes and adjusted pvalues for each gene in dataset</li>
<li>Discussed importance and application of multiple hypothesis
correction</li>
<li>Saved an Robject from our environment to file</li>
</ul>
<p>Now that we’ve generated our differential comparisons and have an
understanding of our results, including multiple hypothesis correction,
we can determine how many genes are differentially expressed between our
conditions and how to visualize our results.</p>
<hr />
</div>
<div id="sources" class="section level1">
<h1>Sources</h1>
<ul>
<li>HBC DGE training module, part 1: <a
href="https://hbctraining.github.io/DGE_workshop/lessons/04_DGE_DESeq2_analysis.html"
class="uri">https://hbctraining.github.io/DGE_workshop/lessons/04_DGE_DESeq2_analysis.html</a></li>
<li>HBC DGE training module, part 2: <a
href="https://hbctraining.github.io/DGE_workshop/lessons/05_DGE_DESeq2_analysis2.html"
class="uri">https://hbctraining.github.io/DGE_workshop/lessons/05_DGE_DESeq2_analysis2.html</a></li>
<li>DESeq2 vignette: <a
href="http://bioconductor.org/packages/devel/bioc/vignettes/DESeq2/inst/doc/DESeq2.html#differential-expression-analysis"
class="uri">http://bioconductor.org/packages/devel/bioc/vignettes/DESeq2/inst/doc/DESeq2.html#differential-expression-analysis</a></li>
</ul>
<hr />
<p>These materials have been adapted and extended from materials listed
above. These are open access materials distributed under the terms of
the <a href="http://creativecommons.org/licenses/by/4.0/">Creative
Commons Attribution license (CC BY 4.0)</a>, which permits unrestricted
use, distribution, and reproduction in any medium, provided the original
author and source are credited.</p>
<br/> <br/>
<hr/>
<table style="width:100%;">
<colgroup>
<col width="28%" />
<col width="42%" />
<col width="28%" />
</colgroup>
<thead>
<tr class="header">
<th align="left"><a href="Module09_SampleQCViz.html">Previous
lesson</a></th>
<th align="center"><a href="#top">Top of this lesson</a></th>
<th align="right"><a href="Module11_DEVisualizations.html">Next
lesson</a></th>
</tr>
</thead>
<tbody>
</tbody>
</table>
</div>

<div id="rmd-source-code">---
title: "DE Testing"
author: "UM Bioinformatics Core"
date: "`r Sys.Date()`"
output:
        html_document:
            includes:
                in_header: header.html
            theme: paper
            toc: true
            toc_depth: 4
            toc_float: true
            number_sections: false
            fig_caption: true
            markdown: GFM
            code_download: true
---

<style type="text/css">
body, td {
   font-size: 18px;
}
code.r{
  font-size: 12px;
}
pre {
  font-size: 12px
}
</style>

```{r, include = FALSE}
source("../bin/chunk-options.R")
knitr_fig_path("10-")
```

In this module, we will learn:

* What a p-value does (and does not!) represent
* How to modify and fit our DE model
* How to generate tables of DE statistics
* The importance multiple hypothesis corrections

<br>

```{r Modules, eval=TRUE, echo=FALSE, message=FALSE, warning=FALSE}
library(DESeq2)
library(tidyr)
library(dplyr)
# load("rdata/RunningData.RData")
```

# Differential Expression Workflow {.unlisted .unnumbered}

Now we test for differential expression between our groups of interest and return a table of results.

![](./images/wayfinder/wayfinder-DEComparisons.png){width=75%}

---

# Testing for DE

Before showing the code for generating differential expression results, let's walk through some toy data and discuss our intuition, its limitations, and what we can do to formalize our thought around determining if a gene is differentially expressed.

Let's start by looking at expression boxplots for a few "genes" where we've made up the data. In each plot, we are comparing the expression levels (on the y-axis) for samples (each point) representing a WT (red) and KO (teal) condition. The boxplot shows the 25% - 75% distribution, along with outliers, with a bar representing the median value, and a black-crossed-point representing the mean.

For the first example, we ask: Does the KO affect the expression of the gene?

![](./images/Module10_stat_plot_Gene_3.png){width=75%}

It seems pretty clear that there is a large difference between the means of the two groups, and within-group spread is quite low. So the answer to the question is likely "Yes".

Consider a second example:

![](./images/Module10_stat_plot_Gene_1.png){width=75%}

Here there isn't a large difference between the means of the two groups, and there is quite a bit of spread within-group. The answer here is likely "No".

Finally, consider a third example:

![](./images/Module10_stat_plot_Gene_2.png){width=75%}

The means are not so close to each other in this example, though there is still quite a bit of spread. This is example is perhaps less clear.

Looking at all three of the hypothetical genes together, we see them in relation to one another and how they span the range from No, to Maybe?, to Definitely.

![](./images/Module10_stat_plot.png)

Now, consider needing to make this decision for 20,000 genes. Even if all the genes had clear separation, that would take a lot of time. However, it's reasonable to expect to see a lot of genes with more ambiguous expression like Gene 2.

**We need a formal, reproducible, way to make this decision!**

Thankfully statistics provides the formality we want, but for any statistical test we need to clearly state what we are testing.  When testing for differential expression we assume that for any particular gene, there is no difference in expression between conditions. Statisticians would call this the **"null hypothesis"**.

For each gene, DESeq2 computes a "Wald statistic" which is a single number encapsulating the difference in the means and the spread of the groups. However, this number alone doesn't allow us to determine if a gene is differentially expressed. We need a **second** number to give us an idea of how extreme that statistic is among the distribution of possible statistics, which is the **"p-value"**. The **"p-value"** along with choosing a threshold will inform us if we can **reject the null hypothesis** and classify a gene as differentially expressed.

Imagine shuffling the group labels and recomputing the Wald statistic over and over again. You'd get a distribution of statistics that would look similar to a normal curve. The p-value essentially tells you how likely you are to have seen the statistic by chance. So when we set a p-value = 0.05 as a threshold, we're saying, "there is a 5% chance I'd see something this extreme when there was actually no effect". So the evidence is **strong**, but not **ironclad**.



# DESeq2 Model Fitting

Based on our PCA results, we'll copy our DESeq object and revise our original model design to add a covariate for "batch". Then, we'll fit our model using the `DESeq` function and take a look at the objects we generate. This command applies the model to our data, using the sample information supplied and model  `dds_batch` object so can take some time to run.

```{r FitModelBatch, message=FALSE}

#  access model slot from original `dds_filtered` DESeq2 object
design(dds_filtered)

# copy original object and then modify model
dds_batch_filtered = dds_filtered
design(dds_batch_filtered) <- ~ condition + batch

# after updating model, fit model with `DESeq` function
?DESeq
dds_batch_fitted = DESeq(dds_batch_filtered)
dds_batch_fitted
```

<!-- add as bonus: Accessing functions available for specific object types https://rpubs.com/mikelove/pkgmethods 
# see what functions are available for accessing/modifying our DESeqDataSet objects
showMethods(classes="DESeqDataSet", where=getNamespace("DESeq2"))--> 


While we won't be able to delve into model options in detail, more complex model designs including adding "interaction terms" between multiple group labels, are helpfully described in [this support thread](https://support.bioconductor.org/p/98628/) as well as in the [DESeq2 vignette](https://bioconductor.org/packages/release/bioc/vignettes/DESeq2/inst/doc/DESeq2.html#interactions).

<details>
    <summary>*Click for example of model design check*</summary>
    When including multiple terms in our model it's helpful to check the corresponding design matrix to ensure that our batches are not confounded, which would cause `DESeq2` to return a `model not full rank` error when attempting to fit the model. We can do that using the `model.matrix` function, providing our intended model and our sample information. 

```{r TestInteractions}
model.matrix(~ condition + batch, samplesheet_batch)
```
  When we look at the outputs, we can see that all of the returned columns have values (`1`) included. If our model was not full rank, then we would see a columns with no values (all `0`) returned.
</details>
<br>

<details>
    <summary>*Dispersion models and possible warning messages*</summary>

   Depending on the data set you are analyzing, you may see a warning that the default 'parametric' dispersion model so a local regression was substituted. When seeing this warning, we recommend looking at a dispersion plot with the `plotDispEsts(dds)` function, but as [this bioconductor thread](https://support.bioconductor.org/p/107937/) discusses, other visualizations of our data might be more helpful and/or easier to interpret why the data doesn't fit the default parametric model.

</details>
<br>
<br>

<details>
    <summary>*Click for code for a plot of dispersion estimates*</summary>
    We can visualize the **dispersion estimates** with the `plotDispEsts` function. This plot shows the the DESeq2 normalization results for our data, which centers on shrinking the variance across all genes to better fit the expected spread at a given expression level.
```{r CheckDispersions}
plotDispEsts(dds_batch_fitted)
```
    
   Above is the raw data plotted in black, the fitted (or expected) dispersion in red, and the normalized data with scaled variance in blue. Since we have fairly small sample sizes for each condition, we see shrinkage for many genes but a reasonable correlation between the expression level and dispersions.

   This [HBC tutorial](https://hbctraining.github.io/DGE_workshop/lessons/04_DGE_DESeq2_analysis.html) has a more detailed overview of estimating size factors, estimating gene dispersion, and the shrinkage procedure, as well as examples of concerning dispersion plots that may suggest reassessing quality of the experimental data.
</details>
<br>

Notice that there is now more information in the `DESeqDataSet` object than there was prior to our normalization. There is information about the model fit and about the library size normalization. DESeq2 will use this information when we perform the test for differential expression.

The `DESeq()` function is actually doing three things automatically for us. It calculates:

1. The size factors to normalize for library size with `estimateSizeFactors(dds_filtered)`,
2. Dispersion estimates to shrink the dispersions with `estimateDispersions(dds_filtered)`, and
3. The Wald test statistics with `nbinomWaldTest(dds_filtered)`.

The `resultsNames()` function returns the names of the estimated effects of the model. Note that  the results are stored based on the levels we set in the sample sheet and that the convention is to always have "case" first versus "control" second.

```{r FitModelStandardCheck1, eval=TRUE}
resultsNames(dds_batch_fitted)
```

The results include the single comparison representing the two levels of `condition`. If there were more levels in the `condition` column, there would be more results listed here because DESeq2 would implicitly compare all other levels to the reference level.


**Checkpoint**: *If you see the same results when you execute `resultsNames(dds_batch_fitted)`, please indicate with the green 'yes' button. Otherwise, please use the red 'x' button to get help*


# Generating DE Results

There is only the one comparison in the results, so we will refer to it in the `name` parameter of the `results()` function, and assign the result as an object.

```{r SetsResult}
results_deficient_vs_control = results(dds_batch_fitted, name = 'condition_deficient_vs_control')
head(results_deficient_vs_control)
```

If we look at the results table, we see that the row names are gene identifiers (in this case ENSEMBL IDs because that's what the GTF we used in the call to RSEM+STAR) and then we see the following columns:

1. `baseMean` is the average of the normalized count values, divided by size factors and taken over all samples, and can be interpreted as the relative expression level of that gene across all samples.
2. `log2FoldChange` is the log2 transformed ratio of the expression of the numerator group (first group) over the denominator group (second group after "vs"). Note that in our comparison, the `log2FoldChange` column compares the expression of `deficient` over the denominator group, `control`. If the value is positive, that means the expression of that gene is greater across the `deficient` samples than across the `control` samples. If the value is negative, that means the expression of that gene is greater across the `deficient` samples.
3. `lfcSE` is the standard error for the log2 fold change estimate.

> # Note: `results()` defaults {.unlisted .unnumbered}
> If no arguments are passed to `results()`, then the log2 fold changes and Wald test p-value will be for the **last variable** in the design formula, and if this is a factor, the comparison will be the **last level** over the **reference level**. If you specify `name`, as we did above, then the behavior is given by the name used from `resultsNames()`.

There are multiple ways to specify the test to be done using the `results()` function. It is especially helpful to know this when fitting more complex models and testing more complex contrasts. To demonstrate this, consider this description from the help for `results()`:

> `contrast`: a character vector with exactly three elements: the name of a factor in the design formula, the name of the numerator level for the fold change, and the name of the denominator level for the fold change

So an alternative way to test the same contrast as above (i.e.  `deficient` / `control`) is:

```{r SetsResultsAlt}
alt_results_deficient_vs_control = results(dds_batch_fitted, contrast = c('condition', 'deficient', 'control'))
head(alt_results_deficient_vs_control)
```

This way of calling `results()` is especially helpful when the levels of the column of interest contain more than two levels because you can specify exactly which levels to test with little confusion.

If we continue to look at the results table, after the `lfcSE` column we see:

4. `stat` is the calculated Wald statistic for that gene.
5. `pvalue` is the *nominal* significance that we described earlier.
6. `padj` is the *adjusted p-value* (also known as a "q-value") and is what we use for determining significantly differently expressed genes.


> # Question {.unlisted .unnumbered}
> Why should we use values from `padj` instead of the `pvalue`? Post in the Slack thread.

<br>

## Multiple hypothesis testing and FDR correction


Each p-value reported in the table is the result of a single test for a single gene. As stated earlier - with a significance cut-off of p-value < 0.05, we're expecting a 5% chance it is a false positive. However the more genes we test, the greater chance we have of seeing a significant results by chance. 

### The multiple hypothesis problem

For example, if we consider running the Wald test for 20 genes instead of just three, we might see something like this where `Gene N` has a p-value below our threshold of 0.05. If we see this pattern, do we expect that `Gene N` is actually differentially expressed (e.g. a true positive)?

<br>
![](./images/FDR-1.png){width=75%}
<br>

What if we tested 100 genes? How many genes would we expect to have p-values < 0.05, just by chance?

<br>
![](./images/FDR-2.png){width=75%}
<br>

What if we tested 10,000 genes? We would expect to see 500 genes with p-values < 0.05, just by chance!

<br>
![](./images/FDR-3.png){width=75%}
<br>

**So if we are testing 20,000 genes for differential expression, we would expect to see ~1,000 genes (5%) with statistical significance, just by chance.**
<br>

### The FDR solution

To address this [multiple hypothesis correction](https://multithreaded.stitchfix.com/blog/2015/10/15/multiple-hypothesis-testing/) can be performed. While there are a few approaches, the default method in DESeq2 is the False Discovery Rate (FDR) ([Benjamini and Hochberg (1995)](https://rss.onlinelibrary.wiley.com/doi/10.1111/j.2517-6161.1995.tb02031.x)).

The default FDR rate cutoff for our analyses is 0.05, meaning the proportion of false positives amongst our *differentially expressed genes* is controlled to 5%. So if we find 500 genes that are differentially expressed with this FDR cutoff, we expect only 25 of them to be false positives. DESeq2 vignette's includes a [further discussion of filtering and multiple testing correction](http://bioconductor.org/packages/devel/bioc/vignettes/DESeq2/inst/doc/DESeq2.html#independent-filtering-and-multiple-testing).

> # Note on `padj` values set to NA {.unlisted .unnumbered}
>
> As discussed in the [HBC tutorial](https://hbctraining.github.io/DGE_workshop/lessons/05_DGE_DESeq2_analysis2.html) as well as the [DESeq2 vignette](http://bioconductor.org/packages/devel/bioc/vignettes/DESeq2/inst/doc/DESeq2.html#i-want-to-benchmark-deseq2-comparing-to-other-de-tools.), DESeq2 reduces the number of genes that will be tested by removing genes with low number of counts and outlier samples.
>
> * If within a row, all samples have zero counts, the baseMean column will be zero, and the log2 fold change estimates, p-value and adjusted p-value will all be set to NA.
> * If a row contains a sample with an extreme count outlier then the p-value and adjusted p-value will be set to NA. These outlier counts are detected by [Cook’s distance](https://en.wikipedia.org/wiki/Cook%27s_distance).
> * If a row is filtered by automatic independent filtering, e.g. for having a low mean normalized count, then only the adjusted p-value will be set to NA.



## Save fitted model and data as .rds

It can be useful to save key R objects to file as we proceed through our analysis - before we do that, let's look at the documentation for the `save` function to see if it does what we want.

```{r SaveDocumentation}
?save
```

As we can see from the documentation the function `save` writes an "external representation" of R objects that can be read back from the file at a later date by using the function `load` or `attach` in most cases. We'll proceed with saving our `dds_batch_fitted` object, creating a subdirectory first.

```{r WriteFitModelObject}
dir.create("outputs/Robjs", recursive=TRUE)
save(dds_batch_fitted,
          file="outputs/Robjs/dds_batch_fitted.rds")
```



---

# Summary

In this section, we:

* Fitted our DESeq2 model
* Performed statistical tests for comparisons of interest
* Generated tables of differential expression results - i.e. fold changes and adjusted pvalues for each gene in dataset
* Discussed importance and application of multiple hypothesis correction
* Saved an Robject from our environment to file

Now that we've generated our differential comparisons and have an understanding of our results, including multiple hypothesis correction, we can determine how many genes are differentially expressed between our conditions and how to visualize our results.

---

# Sources

* HBC DGE training module, part 1: https://hbctraining.github.io/DGE_workshop/lessons/04_DGE_DESeq2_analysis.html
* HBC DGE training module, part 2: https://hbctraining.github.io/DGE_workshop/lessons/05_DGE_DESeq2_analysis2.html
* DESeq2 vignette: http://bioconductor.org/packages/devel/bioc/vignettes/DESeq2/inst/doc/DESeq2.html#differential-expression-analysis



```{r WriteOut.RData, eval=TRUE, echo=FALSE, message=FALSE, warning=FALSE}
# Hidden code block to write out data for knitting
# save.image(file = "rdata/RunningData.RData")
```

---

These materials have been adapted and extended from materials listed above. These are open access materials distributed under the terms of the [Creative Commons Attribution license (CC BY 4.0)](http://creativecommons.org/licenses/by/4.0/), which permits unrestricted use, distribution, and reproduction in any medium, provided the original author and source are credited.

<br/>
<br/>
<hr/>
| [Previous lesson](Module09_SampleQCViz.html) | [Top of this lesson](#top) | [Next lesson](Module11_DEVisualizations.html) |
| :--- | :----: | ---: |
</div>


</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->
<script>
$(document).ready(function () {
  window.initializeSourceEmbed("Module10_DEComparisons.Rmd");
});
</script>

<script>
$(document).ready(function ()  {

    // temporarily add toc-ignore selector to headers for the consistency with Pandoc
    $('.unlisted.unnumbered').addClass('toc-ignore')

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3,h4",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
