<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />


<meta name="author" content="UM Bioinformatics Core" />

<meta name="date" content="2024-07-05" />

<title>Covariates and Batch correction</title>

<script src="site_libs/header-attrs-2.27/header-attrs.js"></script>
<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/paper.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/jqueryui-1.13.2/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<script src="site_libs/navigation-1.1/codefolding.js"></script>
<script src="site_libs/navigation-1.1/sourceembed.js"></script>
<link href="site_libs/highlightjs-9.12.0/default.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link rel="shortcut icon" href="favicon-16x16.png" />
<!--
Favicon from
https://twemoji.twitter.com/
https://favicon.io/emoji-favicons/dna/
-->

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>




<style type="text/css">
#rmd-source-code {
  display: none;
}
</style>





<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
details > summary > p:only-child {
  display: inline;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark the anchor link active (and if it's in a dropdown, also mark that active)
  var dropdown = menuAnchor.closest('li.dropdown');
  if (window.bootstrap) { // Bootstrap 4+
    menuAnchor.addClass('active');
    dropdown.find('> .dropdown-toggle').addClass('active');
  } else { // Bootstrap 3
    menuAnchor.parent().addClass('active');
    dropdown.addClass('active');
  }

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before, .tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "\e259";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "\e258";
  font-family: 'Glyphicons Halflings';
  border: none;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->
<style type="text/css">
.code-folding-btn { margin-bottom: 4px; }
</style>



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-inverse  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-bs-toggle="collapse" data-target="#navbar" data-bs-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">RNA-Seq Demystified</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="Module00_Introduction.html">Intro</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Sequence QC
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="Module01_Warming_Up.html">Warming up</a>
    </li>
    <li>
      <a href="Module02a_Sequence_QC.html">Sequence QC</a>
    </li>
    <li>
      <a href="Module02b_Trimming.html">Trimming</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Alignment
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="Module03a_Reference_Genomes.html">Reference Genomes</a>
    </li>
    <li>
      <a href="Module03b_Alignment.html">Alignment</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Quantification
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="Module04_Alignment_QC_and_Quantification.html">Alignment QC and Quantification</a>
    </li>
    <li>
      <a href="Module05_Additional_Details.html">Optional: Additional details</a>
    </li>
  </ul>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Groups/Model
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="Module06_DEAnalysisSetup.html">DE Introduction</a>
    </li>
    <li>
      <a href="Module07_DESeq2Init.html">DE Initialization</a>
    </li>
    <li>
      <a href="Module08_DESeq2DE.html">DE Normalization and Modeling</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Expression plots
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="Module09_SampleQCViz.html">Sample QC</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Diff Expression
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="Module10_DEComparisons.html">DE Testing</a>
    </li>
    <li>
      <a href="Module10a_batchAndCovars.html">Batches and Covariates</a>
    </li>
    <li>
      <a href="Module11_DEVisualizations.html">DE Visualization</a>
    </li>
    <li>
      <a href="Module12_DEAnnotations.html">DE Annotation</a>
    </li>
    <li>
      <a href="R_bonus_content.html">Bonus content</a>
    </li>
  </ul>
</li>
<li>
  <a href="Module99_Wrap_up.html">Wrap up</a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">

<div class="btn-group pull-right float-right">
<button type="button" class="btn btn-default btn-xs btn-secondary btn-sm dropdown-toggle" data-toggle="dropdown" data-bs-toggle="dropdown" aria-haspopup="true" aria-expanded="false"><span>Code</span> <span class="caret"></span></button>
<ul class="dropdown-menu dropdown-menu-right" style="min-width: 50px;">
<li><a id="rmd-download-source" href="#">Download Rmd</a></li>
</ul>
</div>



<h1 class="title toc-ignore">Covariates and Batch correction</h1>
<h4 class="author">UM Bioinformatics Core</h4>
<h4 class="date">2024-07-05</h4>

</div>


<style type="text/css">
body, td {
   font-size: 18px;
}
code.r{
  font-size: 12px;
}
pre {
  font-size: 12px
}
</style>
<p><Have section follow the fitting a DE of standard model></p>
<p>In this module, we will learn:</p>
<ul>
<li>How to identify possible “confounding” factors</li>
<li>How batches or other covariates may impact your data</li>
</ul>
<p><br></p>
<div id="differential-expression-workflow"
class="section level1 unlisted unnumbered">
<h1 class="unlisted unnumbered">Differential Expression Workflow</h1>
<p>Although we’ve successfully generated DE results, there are instances
where patterns in our PCA plots or additional discussions with our
collaborators cause us to revise our DE model(s).</p>
<p><img src="images/wayfinder/wayfinder-batchAndCovars.png"
style="width:75.0%" /></p>
<p>In this case, we weren’t aware of any <a
href="https://methods-sagepub-com.proxy.lib.umich.edu/reference/encyc-of-research-design/n85.xml">covariates</a>
that should be considered in our comparisons based on the <a
href="https://www.ncbi.nlm.nih.gov/Traces/study/?acc=PRJNA488955&amp;o=acc_s%3Aa">information
available</a> <a
href="https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE119365">publicly
for these data</a>. However, let’s look back at our initial PCA plot and
see if there are any concerning patterns:</p>
<pre class="r"><code>pca_plot = plotPCA(rld, intgroup = c(&#39;condition&#39;), ntop = 500)</code></pre>
<pre><code>using ntop=500 top features by variance</code></pre>
<pre class="r"><code>pca_plot</code></pre>
<p><img src="fig/10a-PCArlog3-1.png" width="816" style="display: block; margin: auto;" /></p>
<p>We do see a high amount of variance (58%) along PC1 that doesn’t seem
to correspond to the experimental treatment. If we saw this pattern for
a “real” dataset, we’d request additional information about the
samples</p>
</div>
<div id="how-to-model-batch-effects-with-deseq2" class="section level1">
<h1>How to model batch effects with DESeq2</h1>
<p>Differences between samples can be due to biological
<strong>covariates</strong> such as sex or patient of origin.
Differences between samples can also be due to technical reasons, such
as collection on different days or different sequencing runs. Groups of
samples with different technical handling are called
<strong>batches</strong> and differences due to the handling of samples
are called <strong>batch effects</strong>.</p>
<p>Any relevant technical batches or biological covariates should be
added as <a href="https://support.bioconductor.org/p/75309/">additional
columns</a> in the sample information table and <a
href="https://support.bioconductor.org/p/98700/">added to a design
formula</a>. We can include batch effects and covariates in the same way
in our design model, as long as the groups do not overlap, or
<strong>confound</strong>, the biological treatment groups.</p>
<p>Let’s assume that after reviewing the above PCA plot, we reached out
to our collaborators and they indicated that the samples were collected
on three different days. We can add that information to our samplesheet
and then create another <code>DESeq2</code> object that includes those
labels and a model with an <a
href="https://bioconductor.org/packages/devel/bioc/vignettes/DESeq2/inst/doc/DESeq2.html#multi-factor-designs">additional
term</a> to account for possible batch effects due to date of
collection.</p>
<pre class="r"><code>samplesheet_batch = samplesheet_ready
samplesheet_batch$batch = factor(c(rep(c(&quot;Day1&quot;, &quot;Day2&quot;, &quot;Day3&quot;), 2)), 
                                 levels = c(&quot;Day1&quot;, &quot;Day2&quot;, &quot;Day3&quot;))

dds_batch = DESeqDataSetFromMatrix(countData = count_table,
                      colData = samplesheet_batch,
                      design = ~ batch + condition)</code></pre>
<p>Note that we first created a new column called “batch” and added date
of collection labels to our sample sheet and then initialized the
<code>DESeq2</code> object with that sample information and a model that
includes an additional term that matches the “batch” column name.</p>
<p>While we don’t have time to delve into model options in detail, more
complex model designs including adding “interaction terms” between
multiple group labels, are helpfully described in <a
href="https://support.bioconductor.org/p/98628/">this support thread</a>
as well as in the <a
href="https://bioconductor.org/packages/release/bioc/vignettes/DESeq2/inst/doc/DESeq2.html#interactions">DESeq2
vignette</a>.</p>
<details>
<summary>
<em>Click for example of model design check</em>
</summary>
<p>When including multiple terms in our model it’s helpful to check the
corresponding design matrix to ensure that our batches are not
confounded, which would cause <code>DESeq2</code> to return a
<code>model not full rank</code> error when attempting to fit the model.
We can do that using the <code>model.matrix</code> function, providing
our intended model and our sample information.</p>
<pre class="r"><code>model.matrix(~ condition + batch, samplesheet_batch)</code></pre>
<pre><code>         (Intercept) conditiondeficient batchDay2 batchDay3
sample_A           1                  0         0         0
sample_B           1                  0         1         0
sample_C           1                  0         0         1
sample_D           1                  1         0         0
sample_E           1                  1         1         0
sample_F           1                  1         0         1
attr(,&quot;assign&quot;)
[1] 0 1 2 2
attr(,&quot;contrasts&quot;)
attr(,&quot;contrasts&quot;)$condition
[1] &quot;contr.treatment&quot;

attr(,&quot;contrasts&quot;)$batch
[1] &quot;contr.treatment&quot;</code></pre>
When we look at the outputs, we can see that all of the returned columns
have values (<code>1</code>) included. If our model was not full rank,
then we would see a columns with no values (all <code>0</code>)
returned.
</details>
<p><br></p>
<p>Just like in our initial analysis, we’ll also want to filter our
data.</p>
<pre class="r"><code>keep = rowSums(counts(dds_batch)) &gt;= 10
dds_batch_filtered = dds_batch[keep,]</code></pre>
<p>Now that we’ve initialized a <code>DESeq2</code> object that includes
our batch labels and filtered the data, we can generate the
<code>rlog</code> normalized count data and have that meta-data
included:</p>
<pre class="r"><code>rld_batch = rlog(dds_batch_filtered, blind = TRUE)
head(rld_batch)</code></pre>
<pre><code>class: DESeqTransform 
dim: 6 6 
metadata(1): version
assays(1): &#39;&#39;
rownames(6): ENSMUSG00000000001 ENSMUSG00000000028 ... ENSMUSG00000000049
  ENSMUSG00000000056
rowData names(7): baseMean baseVar ... dispFit rlogIntercept
colnames(6): sample_A sample_B ... sample_E sample_F
colData names(4): genotype condition batch sizeFactor</code></pre>
<p>Now that we have the additional labels added and normalized our data,
we can generate a new PCA plot to see if the the batches explain any of
the variance along PC1:</p>
<pre class="r"><code>pca_plot_batch = plotPCA(rld_batch, intgroup = c(&#39;batch&#39;), ntop = 500) + ggtitle(&quot;Batch labeled - iron deficiency data&quot;)</code></pre>
<pre><code>using ntop=500 top features by variance</code></pre>
<pre class="r"><code>pca_plot_batch</code></pre>
<p><img src="fig/10a-BatchPCA-1.png" width="816" style="display: block; margin: auto;" /></p>
<p>From this plot, we can see that the dates of collection are primarily
separated along PC1, suggesting that this variation might be batch
effects. For data with multiple technical or biological covariates, we
might need to generate PCA plots for each of the additional labels to
determine what labels might be relevant to that dataset. However, based
on the PCA plot for these data, we can proceed with fitting our model
and generating DE results that account for these batches.</p>
</div>
<div id="fit-a-model-that-includes-a-covariate" class="section level1">
<h1>Fit a model that includes a covariate</h1>
<p>To fit our updated model, just like for our initial analysis, we use
the <code>DESeq</code> function:</p>
<pre class="r"><code>dds_batch_fitted = DESeq(dds_batch_filtered)
resultsNames(dds_batch_fitted) </code></pre>
<pre><code>[1] &quot;Intercept&quot;                      &quot;batch_Day2_vs_Day1&quot;            
[3] &quot;batch_Day3_vs_Day1&quot;             &quot;condition_deficient_vs_control&quot;</code></pre>
<p>Notice that we have two additional sets of results that include the
<code>batch</code> labels. While this indicates that the term we
included for <code>batch</code> was fit, these results would not be
biologically interesting.</p>
<p>Let’s look at these results from our comparison of interest:</p>
<pre class="r"><code>resultsBatch_deficient_vs_control = results(dds_batch_fitted, name = &#39;condition_deficient_vs_control&#39;)
head(resultsBatch_deficient_vs_control) # results from model with term for batch</code></pre>
<pre><code>log2 fold change (MLE): condition deficient vs control 
Wald test p-value: condition deficient vs control 
DataFrame with 6 rows and 6 columns
                      baseMean log2FoldChange     lfcSE      stat    pvalue      padj
                     &lt;numeric&gt;      &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt;
ENSMUSG00000000001  1489.83039       0.274966  0.165703  1.659394 0.0970364  0.538526
ENSMUSG00000000028  1748.93544       0.217805  0.152329  1.429833 0.1527651  0.647029
ENSMUSG00000000031  2151.87715       0.138522  0.289999  0.477665 0.6328885  0.936072
ENSMUSG00000000037    24.91672       0.580793  0.709198  0.818944 0.4128186        NA
ENSMUSG00000000049     7.78377      -1.180919  1.360812 -0.867805 0.3855013        NA
ENSMUSG00000000056 19653.54030      -0.204334  0.180015 -1.135096 0.2563350  0.761406</code></pre>
<pre class="r"><code>head(results_deficient_vs_control) # results from model with no term</code></pre>
<pre><code>log2 fold change (MLE): condition deficient vs control 
Wald test p-value: condition deficient vs control 
DataFrame with 6 rows and 6 columns
                      baseMean log2FoldChange     lfcSE      stat    pvalue      padj
                     &lt;numeric&gt;      &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt;
ENSMUSG00000000001  1489.83039       0.297760  0.210310  1.415815  0.156830  0.868573
ENSMUSG00000000028  1748.93544       0.226421  0.176795  1.280695  0.200301  0.902900
ENSMUSG00000000031  2151.87715       0.457635  0.764579  0.598545  0.549476  0.995391
ENSMUSG00000000037    24.91672       0.579130  0.561259  1.031840  0.302147  0.950613
ENSMUSG00000000049     7.78377      -0.899483  1.553063 -0.579167  0.562476  0.998043
ENSMUSG00000000056 19653.54030      -0.174048  0.203529 -0.855151  0.392468  0.982479</code></pre>
<p>We can see that while the structure of the results table is the same,
the returned statistics are slightly different. However, since we
<em>didn’t</em> actually talk to a collaborator to identify if batch
could explain the % of variance observed within our treatment groups,
we’ll save these results to file but will proceed with visualizing the
results of our initial model.</p>
<pre class="r"><code>save(dds_batch_fitted,
          file=&quot;outputs/Robjs/dds_batch_fitted.Robj&quot;)</code></pre>
<hr />
</div>
<div id="summary" class="section level1">
<h1>Summary</h1>
<p>In this section, we:</p>
<ul>
<li>Discussed technical batches and biological covariates</li>
<li>Fitted a DESeq2 model that includes batch</li>
<li>Generated tables of differential expression results for our batch
model - i.e. fold changes and adjusted pvalues for each gene in
dataset</li>
<li>Saved our fitted model and results with batch to file</li>
</ul>
<p>Now that we’ve generated multiple differential comparisons, we can
determine how many genes are differentially expressed between our
conditions and how to visualize our results.</p>
<hr />
<hr />
</div>
<div id="sources" class="section level1">
<h1>Sources</h1>
<ul>
<li>HBC DGE training module, part 1: <a
href="https://hbctraining.github.io/DGE_workshop/lessons/04_DGE_DESeq2_analysis.html"
class="uri">https://hbctraining.github.io/DGE_workshop/lessons/04_DGE_DESeq2_analysis.html</a></li>
<li><a href="https://biohpc.cornell.edu/doc/RNA-Seq-2020-exercise2.html"
class="uri">https://biohpc.cornell.edu/doc/RNA-Seq-2020-exercise2.html</a></li>
</ul>
<hr />
<p>These materials have been adapted and extended from materials listed
above. These are open access materials distributed under the terms of
the <a href="http://creativecommons.org/licenses/by/4.0/">Creative
Commons Attribution license (CC BY 4.0)</a>, which permits unrestricted
use, distribution, and reproduction in any medium, provided the original
author and source are credited.</p>
<br/> <br/>
<hr/>
<table style="width:100%;">
<colgroup>
<col width="28%" />
<col width="42%" />
<col width="28%" />
</colgroup>
<thead>
<tr class="header">
<th align="left"><a href="Module09_SampleQCViz.html">Previous
lesson</a></th>
<th align="center"><a href="#top">Top of this lesson</a></th>
<th align="right"><a href="Module11_DEVisualizations.html">Next
lesson</a></th>
</tr>
</thead>
<tbody>
</tbody>
</table>
</div>

<div id="rmd-source-code">---
title: "Covariates and Batch correction"
author: "UM Bioinformatics Core"
date: "`r Sys.Date()`"
output:
        html_document:
            includes:
                in_header: header.html
            theme: paper
            toc: true
            toc_depth: 4
            toc_float: true
            number_sections: false
            fig_caption: true
            markdown: GFM
            code_download: true
---

<style type="text/css">
body, td {
   font-size: 18px;
}
code.r{
  font-size: 12px;
}
pre {
  font-size: 12px
}
</style>

```{r, include = FALSE}
source("../bin/chunk-options.R")
knitr_fig_path("10a-")
```

<Have section follow the fitting a DE of standard model>

In this module, we will learn:

* How to identify possible "confounding" factors
* How batches or other covariates may impact your data

<br>

```{r Modules, eval=TRUE, echo=FALSE, message=FALSE, warning=FALSE}
library(DESeq2)
library(tidyr)
library(dplyr)
# load("rdata/RunningData.RData")
```


# Differential Expression Workflow {.unlisted .unnumbered}

Although we've successfully generated DE results, there are instances where patterns in our PCA plots or additional discussions with our collaborators cause us to revise our DE model(s).

![](./images/wayfinder/wayfinder-batchAndCovars.png){width=75%}

In this case, we weren't aware of any [covariates](https://methods-sagepub-com.proxy.lib.umich.edu/reference/encyc-of-research-design/n85.xml) that should be considered in our comparisons based on the [information available](https://www.ncbi.nlm.nih.gov/Traces/study/?acc=PRJNA488955&o=acc_s%3Aa) [publicly for these data](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE119365). However, let's look back at our initial PCA plot and see if there are any concerning patterns:

```{r PCArlog3}
pca_plot = plotPCA(rld, intgroup = c('condition'), ntop = 500)
pca_plot
```

We do see a high amount of variance (58%) along PC1 that doesn't seem to correspond to the experimental treatment. If we saw this pattern for a "real" dataset, we'd request additional information about the samples 

# How to model batch effects with DESeq2

Differences between samples can be due to biological **covariates** such as sex or patient of origin. Differences between samples can also be due to technical reasons, such as collection on different days or different sequencing runs. Groups of samples with different technical handling are called **batches** and differences due to the handling of samples are called **batch effects**. 

Any relevant technical batches or biological covariates should be added as [additional columns](https://support.bioconductor.org/p/75309/) in the sample information table and [added to a design formula](https://support.bioconductor.org/p/98700/). We can include batch effects and covariates in the same way in our design model, as long as the groups do not overlap, or **confound**, the biological treatment groups.

Let's assume that after reviewing the above PCA plot, we reached out to our collaborators and they indicated that the samples were collected on three different days. We can add that information to our samplesheet and then create another `DESeq2` object that includes those labels and a model with an [additional term](https://bioconductor.org/packages/devel/bioc/vignettes/DESeq2/inst/doc/DESeq2.html#multi-factor-designs) to account for possible batch effects due to date of collection.

```{r Confounders}
samplesheet_batch = samplesheet_ready
samplesheet_batch$batch = factor(c(rep(c("Day1", "Day2", "Day3"), 2)), 
                                 levels = c("Day1", "Day2", "Day3"))

dds_batch = DESeqDataSetFromMatrix(countData = count_table,
                      colData = samplesheet_batch,
                      design = ~ batch + condition)
```

Note that we first created a new column called "batch" and added date of collection labels to our sample sheet and then initialized the `DESeq2` object with that sample information and a model that includes an additional term that matches the "batch" column name.

While we don't have time to delve into model options in detail, more complex model designs including adding "interaction terms" between multiple group labels, are helpfully described in [this support thread](https://support.bioconductor.org/p/98628/) as well as in the [DESeq2 vignette](https://bioconductor.org/packages/release/bioc/vignettes/DESeq2/inst/doc/DESeq2.html#interactions).

<details>
    <summary>*Click for example of model design check*</summary>
    When including multiple terms in our model it's helpful to check the corresponding design matrix to ensure that our batches are not confounded, which would cause `DESeq2` to return a `model not full rank` error when attempting to fit the model. We can do that using the `model.matrix` function, providing our intended model and our sample information. 

```{r TestInteractions}
model.matrix(~ condition + batch, samplesheet_batch)
```
  When we look at the outputs, we can see that all of the returned columns have values (`1`) included. If our model was not full rank, then we would see a columns with no values (all `0`) returned.
</details>
<br>

Just like in our initial analysis, we'll also want to filter our data.
```{r FilterData}
keep = rowSums(counts(dds_batch)) >= 10
dds_batch_filtered = dds_batch[keep,]
```

Now that we've initialized a `DESeq2` object that includes our batch labels and filtered the data, we can generate the `rlog` normalized count data and have that meta-data included:

```{r BatchNormalization}
rld_batch = rlog(dds_batch_filtered, blind = TRUE)
head(rld_batch)
```

Now that we have the additional labels added and normalized our data, we can generate a new PCA plot to see if the the batches explain any of the variance along PC1:

```{r BatchPCA}
pca_plot_batch = plotPCA(rld_batch, intgroup = c('batch'), ntop = 500) + ggtitle("Batch labeled - iron deficiency data")
pca_plot_batch
```

From this plot, we can see that the dates of collection are primarily separated along PC1, suggesting that this variation might be batch effects. For data with multiple technical or biological covariates, we might need to generate PCA plots for each of the additional labels to determine what labels might be relevant to that dataset. However, based on the PCA plot for these data, we can proceed with fitting our model and generating DE results that account for these batches.

# Fit a model that includes a covariate 

To fit our updated model, just like for our initial analysis, we use the `DESeq` function:

```{r FitModelCovariate, message=FALSE}
dds_batch_fitted = DESeq(dds_batch_filtered)
resultsNames(dds_batch_fitted) 
```
Notice that we have two additional sets of results that include the `batch` labels. While this indicates that the term we included for `batch` was fit, these results would not be biologically interesting.

Let's look at these results from our comparison of interest:
```{r setBatchResults}
resultsBatch_deficient_vs_control = results(dds_batch_fitted, name = 'condition_deficient_vs_control')
head(resultsBatch_deficient_vs_control) # results from model with term for batch
head(results_deficient_vs_control) # results from model with no term
```
We can see that while the structure of the results table is the same, the returned statistics are slightly different. However, since we *didn't* actually talk to a collaborator to identify if batch could explain the % of variance observed within our treatment groups, we'll save these results to file but will proceed with visualizing the results of our initial model.

```{r WriteFitModelObject}
save(dds_batch_fitted,
          file="outputs/Robjs/dds_batch_fitted.Robj")
```



---

# Summary

In this section, we:

* Discussed technical batches and biological covariates
* Fitted a DESeq2 model that includes batch
* Generated tables of differential expression results for our batch model - i.e. fold changes and adjusted pvalues for each gene in dataset
* Saved our fitted model and results with batch to file

Now that we've generated multiple differential comparisons, we can determine how many genes are differentially expressed between our conditions and how to visualize our results.

---


---

# Sources

* HBC DGE training module, part 1: https://hbctraining.github.io/DGE_workshop/lessons/04_DGE_DESeq2_analysis.html
* https://biohpc.cornell.edu/doc/RNA-Seq-2020-exercise2.html

---

These materials have been adapted and extended from materials listed above. These are open access materials distributed under the terms of the [Creative Commons Attribution license (CC BY 4.0)](http://creativecommons.org/licenses/by/4.0/), which permits unrestricted use, distribution, and reproduction in any medium, provided the original author and source are credited.

<br/>
<br/>
<hr/>
| [Previous lesson](Module09_SampleQCViz.html) | [Top of this lesson](#top) | [Next lesson](Module11_DEVisualizations.html) |
| :--- | :----: | ---: |

</div>


</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->
<script>
$(document).ready(function () {
  window.initializeSourceEmbed("Module10a_batchAndCovars.Rmd");
});
</script>

<script>
$(document).ready(function ()  {

    // temporarily add toc-ignore selector to headers for the consistency with Pandoc
    $('.unlisted.unnumbered').addClass('toc-ignore')

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3,h4",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
