---
title: "Differential Expression (DE) Introduction"
author: "UM Bioinformatics Core"
date: "`r Sys.Date()`"
output:
        html_document:
            includes:
                in_header: header.html
            theme: paper
            toc: true
            toc_depth: 4
            toc_float: true
            number_sections: false
            fig_caption: true
            markdown: GFM
            code_download: true
---

<style type="text/css">
body, td {
   font-size: 18px;
}
code.r{
  font-size: 12px;
}
pre {
  font-size: 12px
}
</style>

```{r, include = FALSE}
source("../bin/chunk-options.R")
knitr_fig_path("06-")
```

In this module, we will learn:

* How to set up a differential expression analysis, using reproducible research principles  
* What is DESeq2 and why it is widely used for differential expression comparisons
* How to load and review gene count data

<br>

# Differential Expression Workflow {.unlisted .unnumbered}

Today we will proceed through key steps in a differential expression analysis, starting from a count table that's similar to what you generated in the first half of the workshop and one of the outputs included in the [data that the Advanced Genomics Core delivers](https://brcf.medicine.umich.edu/cores/advanced-genomics/data-delivery/) for RNA-seq libraries.

<img src="images/wayfinder/wayfinder-DEAnalysisSetup.png" alt="wayfinder" style="width: 1000px;"/>

# Getting Started in RStudio

## Log on to RStudio Server

Open a web browser to the following URL: [http://bfx-workshop01.med.umich.edu](http://bfx-workshop01.med.umich.edu)

You should now be looking at a page that will allow you to login to the RStudio server:

<img src="images/rstudio_login_screen.png" alt="rstudio default session" style="width: 1000px;"/>

Enter your user credentials and click <kbd>Sign In</kbd>. The credentials are the same used earlier in the workshop and were provided via an email entitled "UM BioinfCore Workshop Login".

However, if you forget yours, a helper can retrieve it for you if you ask in Slack. You should now see the RStudio interface:

<img src="images/rstudio_session_default.png" alt="rstudio default session" style="width:1000px;"/>

<br>

## What are we working towards?

To refresh ourselves on the interface of RStudio, look in the lower right pane, that has a list of files and folders. Click on the `RSD_R` folder and then click on the `WelcomeToRSD.R` script to open it.

There should now be four panes in your RStudio window, with the `WelcomeToRSD` script in the upper left pane, this is the "Source" or "Scripts" pane. It displays the code that we will write to perform our analysis.

<img src="images/WelcomeScriptOpened.png" alt="rstudio default session" style="width:1000px;"/>

In the Scripts pane, there is a line of icons, towards the right side of the pane there is a <kbd>Run</kbd> button. First highlight all the code in the Source pane, and then click <kbd>Run</kbd>.

<br>

**Checkpoint**

What just happened?

1. We see in that our code has been run in the `Console` panel, including some warning messages.

1. We see that a number of objects were created in our Environment panel including an object named `DE_results`.

1. We see that the Plot tab now has a volcano plot displayed - Note: if the plot is not displayed, you may need to expand the right panel boundary before re-running the last line of the script

**What does this volcano plot show?**

*Now that we have seen the end results of a DE analysis, any questions?*

<br>
<br>

## Create an RStudio Project

Just like in Computational Foundation workshop, let's start our analysis by creating a project where we can store our data, scripts and outputs.

1. To create a project, go to the <kbd>File</kbd> menu, and click <kbd>New Project...</kbd>. The following window will appear:

<img src="images/new_project_window.png" alt="new project window" style="width: 600px;"/>

2. In this window, select <kbd>Existing Directory</kbd>. For "Project working directory", click <kbd>Browse...</kbd>, select the "RSD_R" folder, and click <kbd>Choose</kbd>. This will use the `/home/workshop/user/RSD_R` folder as the project directory.

3. Finally click <kbd>Create Project</kbd>. In the "Files" tab of your output pane (more about the RStudio layout in a moment), you should see an RStudio project file, **RSD_R.Rproj**. All RStudio projects end with the "**.Rproj**" file extension.

Note that there is already a `data/` folder which contains the data we will use for these lessons.

## Creating an R script

Let's create an R script file:

- Click the <kbd>File</kbd> menu and select <kbd>New File</kbd> and then <kbd>R Script</kbd>.
- Before we go any further, save your script by clicking the save/disk icon that is in the bar above the first line in the script editor, or click the <kbd>File</kbd> menu and select <kbd>Save</kbd>.
- In the "Save File" window that opens, select <kbd>New Folder</kbd>. Name it **"scripts"**.
- Finally, name your file **"diffex"** in the "File name" field.

The new script **diffex.R** is now in the `scripts` folder. You can see that by clicking the `scripts` folder in the "Files" pane. And you can go back up to the main project folder by clicking the `..` to the right of the up arrow in the "Files" pane. By convention, R scripts end with the file extension **.R**.


## File organization for reproducible research

Part of any successful bioinformatics analysis is documentation and data stewardship, which are important aspects of "reproducible research" (see: [Nobel, 2009](https://journals.plos.org/ploscompbiol/article?id=10.1371/journal.pcbi.1000424)).

To follow best practices for file organization for reproducible research, we should make distinct locations for:

* Raw data
* Code
* Output


To organize our files , we'll create some more directories in the `RSD_R` folder which is now our project root and current working directory. We can already see that there a `data/` folder and we just created the `scripts/` folder with the RStudio GUI. Let's create an `outputs/` folder using the `dir.create()` function in R.

```{r create_dirs, eval = FALSE}
dir.create('outputs')
dir.create('outputs/figures')
dir.create('outputs/tables')
```

**Checkpoint**: *Please use the green 'check' if you have saved your code file and see the `data`, `scripts`, and `outputs` directories in the `RSD_R` folder and the red 'x' if you do not.*

> ## Reminder: RStudio code execution {.unlisted .unnumbered}
>
> <kbd>Ctrl</kbd>+<kbd>Enter</kbd> is a standard shortcut in RStudio to send the current line (or selected lines) to the console. If you see `>` in the Console, then R has executed the command. If you see a `+`, this means that the command is not complete, R thinks there is more to your command. You can use the <kmd>esc</kmd> to get out of this state.

> ## Reminder: Object naming conventions {.unlisted .unnumbered}
> * Cannot start with numbers
> * Cannot include dashes
> * Cannot have spaces
> * Should not be identical to a named function
> * Dots and underscores can separate parts of names, alternatively CamelCase accomplishes this

----

# Tools for Differential Gene Expression analysis

As discussed during the webinar, bulk RNA-seq can be used to test for differential expression between conditions or treatments, using statistical approaches that are appropriate for biological data.

While there are several tools that can be used for differential expression comparisons, we will use [DESeq2](https://bioconductor.org/packages/release/bioc/html/DESeq2.html) in our analysis today.

DESeq2 is one of two tools, along with [edgeR](https://bioconductor.org/packages/release/bioc/html/edgeR.html), considered ['best practice'](https://bmcbioinformatics.biomedcentral.com/articles/10.1186/1471-2105-14-91) for differential expression comparisons. Both tools apply similar methods that account for the "shape" of data expected for RNA-seq and are fairly stringent in calling differentially expressed genes, lowering the risk of investigating "false positive" genes (e.g. genes that don't actually have different expression between treatment groups and therefore are not relevant to the biological process).

Additionally, DESeq2 also has an [excellent vignette](https://bioconductor.org/packages/release/bioc/vignettes/DESeq2/inst/doc/DESeq2.html)
from Love, Anders, and Huber, from which our workflow is partially adapted, and is a good resource when analyzing your own data
(see also: [Love, Anders, and Huber. _Genome Biology_. 2014.](https://doi.org/10.1186/s13059-014-0550-8)).


## DESeq2 assumptions and requirements

A key assumption of DESeq2 is that *biological variance is much greater than technical variance* (which should be true if [best practices](https://www.txgen.tamu.edu/faq/rna-isolation-best-practices/) for [quality RNA isolation](https://www.biocompare.com/Bench-Tips/128790-Four-Tips-for-Perfecting-RNA-Isolation/) are followed, including DNase treatment!).

Since calculating variance is key to the statistical approach used for DESeq2, if we tried to compare treatment groups with only one sample each, we would get an error (as shown in [this blog post](https://support.bioconductor.org/p/89746/)). Without replicates, there can't be statistical significance (e.g. p-values), but *qualitative* approaches are an option, like looking at the top expressed genes after normalization.

<details>
    <summary>*Click for additional resources regarding statistical testing and tool comparison for RNA-seq data*</summary>
    To learn more about statistical testing and what distributions best model the behavior of RNA-seq data, a good resource is this [EdX lecture by Rafael Irizarry](https://www.youtube.com/watch?v=HK7WKsL3c2w&feature=youtu.be) or this [lecture by Kasper Hansen](https://www.youtube.com/watch?v=C8RNvWu7pAw). Another helpful guide is this [Comparative Study for Differential Expression Analysis by Zhang et al.](https://journals.plos.org/plosone/article?id=10.1371/journal.pone.0103207) from 2014.
</details>
<br>

> ## Exercise
>
> - Post a question that was NOT addressed but that you hope we will be addressed in the later modules **OR**
> - Add a thumbs up to your favorite comment(s) to upvote it

----

# Getting started

## Load Packages

Several packages have already been installed on the server, so we can load them into our R session now. To do that we'll use the `library` function to load the required packages.

```{r Modules, message=FALSE, warning= FALSE, eval=TRUE}
# load libraries
library(DESeq2)
library(tidyverse)
library(matrixStats)
library(ggrepel)
library(pheatmap)
library(RColorBrewer)
library(data.table)
```

*Note: We expect to see some red messages in your console while these packages are loading*

R/RStudio [AKA Posit] has great resources for getting help, including [code 'cheatsheets'](https://posit.co/resources/cheatsheets/) and package vignettes, like for [tidyr](https://cran.r-project.org/web/packages/tidyr/vignettes/tidy-data.html).

Since we loaded the libraries into our R session, we can see documentation out using the `?` operator.
```{r CheckDocumentaion}
?`DESeq2-package`
```

**Checkpoint**: *If you see the R documentation for DESeq2 pop up in your 'help' panel on the right, please indicate with the green 'check' button. If not please use the red 'x' button.*

## Read Counts

Another key assumption for DESeq2 is that the analysis will start with [un-normalized counts](http://bioconductor.org/packages/devel/bioc/vignettes/DESeq2/inst/doc/DESeq2.html#why-un-normalized-counts).

To begin we'll read in a **raw** count data file, `gene_expected_count.txt` which is similar to what would be generated in the alignment steps (and what you would receive from AGC). We'll discuss a few normalizations that can be helpful for understanding broad patterns in our expression data, but raw data **should always** be used as an input for DESeq2.

```{r DataTable, eval=TRUE, echo=FALSE}
## need to include paths relative to site build locations but keep code block hidden
count_table = read.table("../data/R_data/gene_expected_count.txt", header = TRUE, row.names = 1)
```

```{r DataTable2, eval=FALSE}
# read in count data
count_table = read.table("data/gene_expected_count.txt", header = TRUE, row.names = 1)
```

```{r DataTableCheck, eval=TRUE}
head(count_table, n=2) # look at the top of the table
```

By loading in the count data, we've created a data frame with 'gene ids' in ENSEMBL format as rownames and six columns of count data for our samples.

How many genes were included in our reference? We can get a sense of the size of our data with the `str` function.
```{r DataTableCheck2, eval=TRUE}
str(count_table) # look at the size & structure of the table
```


One important consideration for RNA-seq data is how many genes we measured in our data and how well we measured them. Both are impacted by sequencing depth or the number of reads generated per sample.

For human and mouse experiments, the recommendation is sequencing 30-40 million reads per sample to capture both highly expressed (abundant) and lowly expressed (rarer) transcripts in the sample, as ~25,000 protein-coding genes are expected with a polyA (mRNA targeting) library prep for these species. However, as the image below from Illumina shows, sequencing depth has less of an impact than number of replicates in detecting differentially expressed genes (DEGs).

![Illumina's differential expression recovery across replicate number and sequencing depth](images/de_replicates_img.png){width=50%}

<br>

How many reads were generated per sample in our data? We can use the `summarize_all` function from the `dplyr` package to determine the total number of reads or "depth" per sample.

```{r SummarizeReads, eval=TRUE}
count_table %>% summarize_all(sum)
```

# Getting help

How did we know to use the `summarize_all` function to sum each column?

R and RStudio have a strong community component, so if you are getting an error or wondering how to make a command work or how to perform a specific task, there is likely already a solution out there. Search engines are your friend, which is exactly how we found [that approach](https://stackoverflow.com/questions/28873057/sum-across-multiple-columns-with-dplyr)

It can sometimes be a challenge to figure out *what to search for*, so some key parts of a successful search are:

* Package or command run
* `R` or `Bioconductor`
* The error message if there is one
* Version information

How to get session information to aid in a search:
```{r Session info, eval = FALSE, message = FALSE}
sessionInfo()
```

We highly recommend using resources like [Bioconductor Support](https://support.bioconductor.org/), [Biostars](https://www.biostars.org/), and [Stack Overflow](https://stackoverflow.com/questions/tagged/r), including threads on specific packages or common bioinformatic tasks.

I personally use one or more of these resources **every day**.

10x Genomics also has a helpful [10 tips for biologists learning  bioinformatics](https://www.10xgenomics.com/resources/analysis-guides/10-tips-for-biologists-learning-bioinformatics) included in their resources.

# Summary

In this section, we:

* Set up our compute environment
* Learned about the DESeq2 package
* Read in a raw count table and saved it as a data frame

Now that we have our count data processed, we can move on to "unblinding" our data, as the sample names are unique identifiers generated by a sequencing center and not very informative as far as our experimental conditions.

---

# Optional content

An important note is that there are several bonus content sections on the instruction pages, like the two below that we will not be covering in this workshop, but that may have useful context or be helpful when you review this material.

<details>
    <summary>*Click for alternative DESeq2 input options for RSEM outputs*</summary>
    The package `tximport` is another option[recommended the DESeq2  authors](https://support.bioconductor.org/p/90672/) to read in the RSEM expected_counts, as this  package allows for the average transcript length per gene to be used in the DE analysis and, as [described by the author](https://support.bioconductor.org/p/88763/), the `tximport-to-DESeqDataSet` constructor function round the non-integer data generated by RSEM to whole numbers.
</details>
<br>
<details>
    <summary>*Click for comparison of RNA-seq data and microarray data*</summary>
    With [higher sensitivity, greater flexiblity, and decreasing cost](https://www.illumina.com/science/technology/next-generation-sequencing/microarray-rna-seq-comparison.html), sequencing has largely replaced microarray assays for measuring gene expression. A key difference between the platforms is that microarrays measure intensities and are therefore *continous* data while the count data from sequencing is *discrete*. A more detailed comparison between microarrays and sequencing technologies/analysis is outlined in [the online materials for Penn State's STAT555 course](https://online.stat.psu.edu/stat555/node/30/)

</details>
<br>

# Sources

Training resources used to develop materials:

* HBC DGE setup: https://hbctraining.github.io/DGE_workshop/lessons/01_DGE_setup_and_overview.html
* HBC Count Normalization: https://hbctraining.github.io/DGE_workshop/lessons/02_DGE_count_normalization.html
* DESeq2 standard vignette: http://bioconductor.org/packages/devel/bioc/vignettes/DESeq2/inst/doc/DESeq2.html
* DESeq2 beginners vignette: https://bioc.ism.ac.jp/packages/2.14/bioc/vignettes/DESeq2/inst/doc/beginner.pdf
* Bioconductor RNA-seq Workflows: https://www.bioconductor.org/help/course-materials/2015/LearnBioconductorFeb2015/B02.1_RNASeq.html
* CCDL Gastric cancer training materials: https://alexslemonade.github.io/training-modules/RNA-seq/03-gastric_cancer_exploratory.nb.html
* CCDL Neuroblastoma training materials: https://alexslemonade.github.io/training-modules/RNA-seq/05-nb_cell_line_DESeq2.nb.html


```{r WriteOut.RData, eval=TRUE, echo=FALSE, message=FALSE}
#Hidden code block to write out data for knitting

# dir.create("rdata", showWarnings=FALSE)
# save.image(file = "rdata/RunningData.RData")
# How to load for next segment
# load("rdata/RunningData.RData")
```

---

These materials have been adapted and extended from materials listed above. These are open access materials distributed under the terms of the [Creative Commons Attribution license (CC BY 4.0)](http://creativecommons.org/licenses/by/4.0/), which permits unrestricted use, distribution, and reproduction in any medium, provided the original author and source are credited.

<br/>
<br/>
<hr/>
| [Previous lesson](Module05_Additional_Details.html) | [Top of this lesson](#top) | [Next lesson](Module07_DESeq2Init.html) |
| :--- | :----: | ---: |