<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />


<meta name="author" content="UM Bioinformatics Core" />


<title>Alignment QC and Quantification</title>

<script src="site_libs/header-attrs-2.21/header-attrs.js"></script>
<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/paper.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<script src="site_libs/navigation-1.1/codefolding.js"></script>
<script src="site_libs/navigation-1.1/sourceembed.js"></script>
<link href="site_libs/highlightjs-9.12.0/default.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link rel="shortcut icon" href="favicon-16x16.png" />
<!--
Favicon from
https://twemoji.twitter.com/
https://favicon.io/emoji-favicons/dna/
-->

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>




<style type="text/css">
#rmd-source-code {
  display: none;
}
</style>





<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
details > summary > p:only-child {
  display: inline;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark the anchor link active (and if it's in a dropdown, also mark that active)
  var dropdown = menuAnchor.closest('li.dropdown');
  if (window.bootstrap) { // Bootstrap 4+
    menuAnchor.addClass('active');
    dropdown.find('> .dropdown-toggle').addClass('active');
  } else { // Bootstrap 3
    menuAnchor.parent().addClass('active');
    dropdown.addClass('active');
  }

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before, .tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "\e259";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "\e258";
  font-family: 'Glyphicons Halflings';
  border: none;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->
<style type="text/css">
.code-folding-btn { margin-bottom: 4px; }
</style>



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-inverse  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-bs-toggle="collapse" data-target="#navbar" data-bs-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">RNA-Seq Demystified</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="Module00_Introduction.html">Intro</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Sequence QC
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="Module01_Warming_Up.html">Warming up</a>
    </li>
    <li>
      <a href="Module02a_Sequence_QC.html">Sequence QC</a>
    </li>
    <li>
      <a href="Module02b_Trimming.html">Trimming</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Alignment
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="Module03a_Reference_Genomes.html">Reference Genomes</a>
    </li>
    <li>
      <a href="Module03b_Alignment.html">Alignment</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Quantification
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="Module04_Alignment_QC_and_Quantification.html">Alignment QC and Quantification</a>
    </li>
    <li>
      <a href="Module05_Additional_Details.html">Optional: Additional details</a>
    </li>
  </ul>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Groups/Model
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="Module06_DEAnalysisSetup.html">DE Introduction</a>
    </li>
    <li>
      <a href="Module07_DESeq2Init.html">DE Initialization</a>
    </li>
    <li>
      <a href="Module08_DESeq2DE.html">DE Normalization and Modeling</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Expression plots
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="Module09_SampleQCViz.html">Sample QC</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Diff Expression
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="Module10_DEComparisons.html">DE Testing</a>
    </li>
    <li>
      <a href="Module11_DEVisualizations.html">DE Visualization</a>
    </li>
    <li>
      <a href="Module12_DEAnnotations.html">DE Annotation</a>
    </li>
    <li>
      <a href="R_bonus_content.html">Bonus content</a>
    </li>
  </ul>
</li>
<li>
  <a href="Module99_Wrap_up.html">Wrap up</a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">

<div class="btn-group pull-right float-right">
<button type="button" class="btn btn-default btn-xs btn-secondary btn-sm dropdown-toggle" data-toggle="dropdown" data-bs-toggle="dropdown" aria-haspopup="true" aria-expanded="false"><span>Code</span> <span class="caret"></span></button>
<ul class="dropdown-menu dropdown-menu-right" style="min-width: 50px;">
<li><a id="rmd-download-source" href="#">Download Rmd</a></li>
</ul>
</div>



<h1 class="title toc-ignore">Alignment QC and Quantification</h1>
<h4 class="author">UM Bioinformatics Core</h4>

</div>


<style type="text/css">
body{ /* Normal  */
      font-size: 14pt;
  }
pre {
  font-size: 12pt
}
</style>
<p>In this module, we will learn:</p>
<ul>
<li>about the MultiQC tool and its capabilities</li>
<li>how to run multiQC on a remote system</li>
<li>how MultiQC gathers information for QC purposes</li>
<li>how MultiQC presents the results of cutadapt trimming and STAR
alignment</li>
<li>how to combine gene-level results into a count matrix</li>
</ul>
<div id="differential-expression-workflow" class="section level1">
<h1>Differential Expression Workflow</h1>
<p>Here we will take the results from the previous module and operate on
them a bit further. This will wrap up the preceding exercises, leaving
us well-poised to begin differential expression, which we will discuss
today and tomorrow.</p>
<p><img
src="images/wayfinder/wayfinder-Alignment_QC_and_Quantification.png" />
<br> <br> <br> <br></p>
<div id="alignment-statistics-with-multiqc" class="section level2">
<h2>Alignment Statistics with MultiQC</h2>
<p>After aligning reads it is often helpful to know how many reads were
uniquely aligned, mapped to multiple loci, or not mapped at all. The
<code>sample_NLog.final.out</code> file which is output alongside the
alignments in <code>sample_N.temp/</code> folder (we used the
<code>--keep-intermediate-files</code> flag), reports this
information:</p>
<pre><code>                                 Started job on |   Oct 02 13:09:23
                             Started mapping on |   Oct 02 13:09:51
                                    Finished on |   Oct 02 13:12:47
       Mapping speed, Million of reads per hour |   238.82

                          Number of input reads |   11675504
                      Average input read length |   146
                                    UNIQUE READS:
                   Uniquely mapped reads number |   10609591
                        Uniquely mapped reads % |   90.87%
                          Average mapped length |   146.37
                       Number of splices: Total |   2755543
            Number of splices: Annotated (sjdb) |   2734730
                       Number of splices: GT/AG |   2739697
                       Number of splices: GC/AG |   13204
                       Number of splices: AT/AC |   1744
               Number of splices: Non-canonical |   898
                      Mismatch rate per base, % |   0.18%
                         Deletion rate per base |   0.01%
                        Deletion average length |   1.60
                        Insertion rate per base |   0.01%
                       Insertion average length |   1.29
                             MULTI-MAPPING READS:
        Number of reads mapped to multiple loci |   599915
             % of reads mapped to multiple loci |   5.14%
        Number of reads mapped to too many loci |   12786
             % of reads mapped to too many loci |   0.11%
                                  UNMAPPED READS:
  Number of reads unmapped: too many mismatches |   20381
       % of reads unmapped: too many mismatches |   0.17%
            Number of reads unmapped: too short |   389960
                 % of reads unmapped: too short |   3.34%
                Number of reads unmapped: other |   42871
                     % of reads unmapped: other |   0.37%
                                  CHIMERIC READS:
                       Number of chimeric reads |   0
                            % of chimeric reads |   0.00%</code></pre>
</div>
</div>
<div id="multiqc" class="section level1">
<h1>MultiQC</h1>
<p>While the information is incredibly useful, it can be tedious to look
at the report for each sample separately, while keeping track of what
trends emerge. It would be much easier to look at all the data compiled
into a single report. <a href="https://multiqc.info/">MultiQC</a> is a
tool that does exactly this.</p>
<p>MultiQC is designed to interpret and aggregate reports from <a
href="https://multiqc.info/#supported-tools">various tools</a> and
output a single report as an HTML document.</p>
<div id="multiqc-details" class="section level2">
<h2>MultiQC Details</h2>
<p>MultiQC’s main output is the report file in HTML format. This can be
viewed in a web browser. Additionally, it creates a <code>data</code>
directory with text files containing the data that MultiQC gathered
during its execution - this same data is what is shown in the
report.</p>
<p>Given an output directory out_multiqc, we should see something like
the following:</p>
<pre><code># directory of multiqc data files
out_multiqc/multiqc_data/multiqc.log
out_multiqc/multiqc_data/multiqc_data.json
out_multiqc/multiqc_data/multiqc_general_stats.txt
out_multiqc/multiqc_data/multiqc_rsem.txt
out_multiqc/multiqc_data/multiqc_sources.txt
# multiqc report
out_multiqc/multiqc_report.html</code></pre>
<p>In a moment we will run <code>multiqc</code>, and it will detect
these reports from STAR and include them in the report.</p>
<p>If we then open the MultiQC report (HTML), the newly included STAR
section will look something like the following:</p>
<center>
<p><img src="images/multiqc_star.png"
alt="Example of STAR alignment statistics in MultiQC." /> Example of
STAR alignment statistics in MultiQC.</p>
<p>Source: <a
href="https://multiqc.info/examples/rna-seq/multiqc_report.html#star">MultiQC
example report</a></p>
</center>
</div>
<div id="multiqc-with-star-exercise" class="section level2">
<h2>MultiQC With STAR Exercise:</h2>
<ol style="list-style-type: decimal">
<li>Note the contents of our analysis directory, including the STAR
contents</li>
<li>Construct a MultiQC command and execute it on this directory</li>
<li>View the MultiQC report</li>
</ol>
<pre><code># Ensure that we&#39;re in our analysis directory, note the contents
cd ~/RSD_Shell/analysis
ls -l
# View MultiQC help page
multiqc --help
# Construct a MultiQC command and execute it
multiqc --outdir out_multiqc_rsem out_rsem/
# Verify that the output files are present</code></pre>
<p>We just learned how to view all of our alignment results in one
report with the help of MultiQC. MultiQC is also useful because it can
utilize multiple separate modules to create summary figures of different
steps in our pipeline.</p>
<p>What would happen if we pointed multiQC at our trimmed FastQC
directory? What about if we point it to the entire analysis directory?
Let’s try it out and see:</p>
<pre><code># A MultiQC command to analyze the trimmed read results
multiqc --outdir out_multiqc_cutadapt out_fastqc_trimmed/
# A multiQC command for a report of the entire project
multiqc --outdir out_multiqc_all  .</code></pre>
</div>
<div id="transfer-multiqc-reports-with-scp-exercise"
class="section level2">
<h2>Transfer MultiQC Reports With <code>scp</code> Exercise</h2>
<p>Before starting our file transfer exercise, we should make sure that
we are on the same page. Follow the link below:</p>
<p><a href="Module04_breakout03_ex.html">Link to report transfer
exercise</a></p>
<p><br> <br></p>
</div>
<div id="creating-the-count-matrix" class="section level2">
<h2>Creating the count matrix</h2>
<p>We have viewed some of the gene expression quantification results
individually. It can be useful to combine these expression values into a
count matrix. This is helpful when gathering expression-level QC
metrics, as well as for input into a differential gene expression
program such as DESeq2.</p>
<p>There are many ways to combine these results into a count matrix. For
this workshop, we’ll use a small python script to
<code>combine.py</code> that we’ve made for this purpose. To understand
the process a bit more, let’s review the <code>.genes.results</code>
files that we want to combine and discuss some details of the script.
Finally, we’ll end with an exercise creating a count matrix.</p>
<p>If we review the *.genes.results files, we can see various columns of
data output from RSEM that we discussed in the last module.</p>
<pre><code># Review of *.genes.results file contents
gene_id                 transcript_id(s)                        length  effective_length        expected_count  TPM     FPKM
ENSMUSG00000000001      ENSMUST00000000001                      3262.00 3116.28                 601.00          45.50   36.70
ENSMUSG00000000003      ENSMUST00000000003,ENSMUST00000114041   799.50  653.78                  0.00            0.00    0.00</code></pre>
<p>We’ll take the <code>expected_count</code> column from each sample’s
data, and combine these so that we have an aggregated data matrix with a
row for each gene and a column for each sample.</p>
<p>The input for this step will be the directory of *.genes.results
files from RSEM, and the output will be a tab-separated count matrix
file which we can use for count-level QC and differential expression
analysis.</p>
<details>
<summary>
Contents of combine.py script
</summary>
<a
href="https://gist.github.com/twsaari/12c5aa2773292c09c1809d5a3db66903">Here</a>
are the contents of the python script we’ll use,
<code>combine.py</code>:
</details>
</div>
<div id="count-matrix-exercise" class="section level2">
<h2>Count Matrix Exercise:</h2>
<ol style="list-style-type: decimal">
<li>View the help file of <code>combine.py</code></li>
<li>Construct / execute a command to combine our results into a count
matrix</li>
<li>View the resulting count matrix</li>
</ol>
<pre><code># View the help file of combine.py
combine.py --help
# Construct and execute the command to combine.py
combine.py --input_path &quot;out_rsem/*.genes.results&quot; --output_file combined_counts.txt -c expected_count --id_columns gene_id
# View the resulting count matrix
less -S combined_counts.txt</code></pre>
<p><br> <br></p>
<hr />
<p>These materials have been adapted and extended from materials created
by the <a href="http://bioinformatics.sph.harvard.edu/">Harvard Chan
Bioinformatics Core (HBC)</a>. These are open access materials
distributed under the terms of the <a
href="http://creativecommons.org/licenses/by/4.0/">Creative Commons
Attribution license (CC BY 4.0)</a>, which permits unrestricted use,
distribution, and reproduction in any medium, provided the original
author and source are credited.</p>
<br/> <br/>
<hr/>
<table style="width:100%;">
<colgroup>
<col width="28%" />
<col width="42%" />
<col width="28%" />
</colgroup>
<thead>
<tr class="header">
<th align="left"><a href="Module03b_Alignment.html">Previous
lesson</a></th>
<th align="center"><a href="#top">Top of this lesson</a></th>
<th align="right"><a href="Module05_Additional_Details.html">Next
lesson</a></th>
</tr>
</thead>
<tbody>
</tbody>
</table>
</div>
</div>

<div id="rmd-source-code">---
title: "Alignment QC and Quantification"
author: "UM Bioinformatics Core"
output:
        html_document:
            includes:
                in_header: header.html
            theme: paper
            toc: true
            toc_depth: 4
            toc_float: true
            number_sections: false
            fig_caption: true
            markdown: GFM
            code_download: true
---
<style type="text/css">
body{ /* Normal  */
      font-size: 14pt;
  }
pre {
  font-size: 12pt
}
</style>

In this module, we will learn:

* about the MultiQC tool and its capabilities
* how to run multiQC on a remote system
* how MultiQC gathers information for QC purposes
* how MultiQC presents the results of cutadapt trimming and STAR alignment
* how to combine gene-level results into a count matrix

# Differential Expression Workflow

Here we will take the results from the previous module and operate on them a bit further. This will wrap up the preceding exercises, leaving us well-poised to begin differential expression, which we will discuss today and tomorrow.

![](images/wayfinder/wayfinder-Alignment_QC_and_Quantification.png)
<br>
<br>
<br>
<br>

## Alignment Statistics with MultiQC

After aligning reads it is often helpful to know how many reads were uniquely aligned, mapped to multiple loci, or not mapped at all. The `sample_NLog.final.out` file which is output alongside the alignments in `sample_N.temp/` folder (we used the `--keep-intermediate-files` flag), reports this information:

```
                                 Started job on |	Oct 02 13:09:23
                             Started mapping on |	Oct 02 13:09:51
                                    Finished on |	Oct 02 13:12:47
       Mapping speed, Million of reads per hour |	238.82

                          Number of input reads |	11675504
                      Average input read length |	146
                                    UNIQUE READS:
                   Uniquely mapped reads number |	10609591
                        Uniquely mapped reads % |	90.87%
                          Average mapped length |	146.37
                       Number of splices: Total |	2755543
            Number of splices: Annotated (sjdb) |	2734730
                       Number of splices: GT/AG |	2739697
                       Number of splices: GC/AG |	13204
                       Number of splices: AT/AC |	1744
               Number of splices: Non-canonical |	898
                      Mismatch rate per base, % |	0.18%
                         Deletion rate per base |	0.01%
                        Deletion average length |	1.60
                        Insertion rate per base |	0.01%
                       Insertion average length |	1.29
                             MULTI-MAPPING READS:
        Number of reads mapped to multiple loci |	599915
             % of reads mapped to multiple loci |	5.14%
        Number of reads mapped to too many loci |	12786
             % of reads mapped to too many loci |	0.11%
                                  UNMAPPED READS:
  Number of reads unmapped: too many mismatches |	20381
       % of reads unmapped: too many mismatches |	0.17%
            Number of reads unmapped: too short |	389960
                 % of reads unmapped: too short |	3.34%
                Number of reads unmapped: other |	42871
                     % of reads unmapped: other |	0.37%
                                  CHIMERIC READS:
                       Number of chimeric reads |	0
                            % of chimeric reads |	0.00%
```

# MultiQC

While the information is incredibly useful, it can be tedious to look at the report for each sample separately, while keeping track of what trends emerge. It would be much easier to look at all the data compiled into a single report. [MultiQC](https://multiqc.info/) is a tool that does exactly this.

MultiQC is designed to interpret and aggregate reports from [various tools](https://multiqc.info/#supported-tools) and output a single report as an HTML document.

## MultiQC Details

MultiQC's main output is the report file in HTML format. This can be viewed in a web browser. Additionally, it creates a `data` directory with text files containing the data that MultiQC gathered during its execution - this same data is what is shown in the report.

Given an output directory out_multiqc, we should see something like the following:

    # directory of multiqc data files
    out_multiqc/multiqc_data/multiqc.log
    out_multiqc/multiqc_data/multiqc_data.json
    out_multiqc/multiqc_data/multiqc_general_stats.txt
    out_multiqc/multiqc_data/multiqc_rsem.txt
    out_multiqc/multiqc_data/multiqc_sources.txt
    # multiqc report
    out_multiqc/multiqc_report.html


In a moment we will run `multiqc`, and it will detect these reports from STAR and include them in the report.

If we then open the MultiQC report (HTML), the newly included STAR section will look something like the following:

<center>

![Example of STAR alignment statistics in MultiQC.](images/multiqc_star.png)
Example of STAR alignment statistics in MultiQC.

Source: [MultiQC example report](https://multiqc.info/examples/rna-seq/multiqc_report.html#star)

</center>

## MultiQC With STAR Exercise:

1. Note the contents of our analysis directory, including the STAR contents
2. Construct a MultiQC command and execute it on this directory
3. View the MultiQC report

```
# Ensure that we're in our analysis directory, note the contents
cd ~/RSD_Shell/analysis
ls -l
# View MultiQC help page
multiqc --help
# Construct a MultiQC command and execute it
multiqc --outdir out_multiqc_rsem out_rsem/
# Verify that the output files are present
```

We just learned how to view all of our alignment results in one report with the help of MultiQC. MultiQC is also useful because it can utilize multiple separate modules to create summary figures of different steps in our pipeline.

What would happen if we pointed multiQC at our trimmed FastQC directory? What about if we point it to the entire analysis directory? Let's try it out and see:

```
# A MultiQC command to analyze the trimmed read results
multiqc --outdir out_multiqc_cutadapt out_fastqc_trimmed/
# A multiQC command for a report of the entire project
multiqc --outdir out_multiqc_all  .
```

## Transfer MultiQC Reports With `scp` Exercise

Before starting our file transfer exercise, we should make sure that we are on the same page. Follow the link below:

[Link to report transfer exercise](Module04_breakout03_ex.html)


<br>
<br>

## Creating the count matrix

We have viewed some of the gene expression quantification results individually. It can be useful to combine these expression values into a count matrix. This is helpful when gathering expression-level QC metrics, as well as for input into a differential gene expression program such as DESeq2.

There are many ways to combine these results into a count matrix. For this workshop, we'll use a small python script to `combine.py` that we've made for this purpose. To understand the process a bit more, let's review the `.genes.results` files that we want to combine and discuss some details of the script. Finally, we'll end with an exercise creating a count matrix.


If we review the *.genes.results files, we can see various columns of data output from RSEM that we discussed in the last module.

    # Review of *.genes.results file contents
    gene_id                 transcript_id(s)                        length  effective_length        expected_count  TPM     FPKM
    ENSMUSG00000000001      ENSMUST00000000001                      3262.00 3116.28                 601.00          45.50   36.70
    ENSMUSG00000000003      ENSMUST00000000003,ENSMUST00000114041   799.50  653.78                  0.00            0.00    0.00

We'll take the `expected_count` column from each sample's data, and combine these so that we have an aggregated data matrix with a row for each gene and a column for each sample.

The input for this step will be the directory of *.genes.results files from RSEM, and the output will be a tab-separated count matrix file which we can use for count-level QC and differential expression analysis.


<details>
<summary>Contents of combine.py script</summary>

[Here](https://gist.github.com/twsaari/12c5aa2773292c09c1809d5a3db66903) are the contents of the python script we'll use, `combine.py`:
</details>

## Count Matrix Exercise:

1. View the help file of `combine.py`
2. Construct / execute a command to combine our results into a count matrix
3. View the resulting count matrix

```
# View the help file of combine.py
combine.py --help
# Construct and execute the command to combine.py
combine.py --input_path "out_rsem/*.genes.results" --output_file combined_counts.txt -c expected_count --id_columns gene_id
# View the resulting count matrix
less -S combined_counts.txt
```

<br>
<br>

---

These materials have been adapted and extended from materials created by the [Harvard Chan Bioinformatics Core (HBC)](http://bioinformatics.sph.harvard.edu/). These are open access materials distributed under the terms of the [Creative Commons Attribution license (CC BY 4.0)](http://creativecommons.org/licenses/by/4.0/), which permits unrestricted use, distribution, and reproduction in any medium, provided the original author and source are credited.

<br/>
<br/>
<hr/>
| [Previous lesson](Module03b_Alignment.html) | [Top of this lesson](#top) | [Next lesson](Module05_Additional_Details.html) |
| :--- | :----: | ---: |</div>


</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->
<script>
$(document).ready(function () {
  window.initializeSourceEmbed("Module04_Alignment_QC_and_Quantification.md");
});
</script>

<script>
$(document).ready(function ()  {

    // temporarily add toc-ignore selector to headers for the consistency with Pandoc
    $('.unlisted.unnumbered').addClass('toc-ignore')

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3,h4",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
