<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />


<meta name="author" content="UM Bioinformatics Core" />

<meta name="date" content="2023-08-07" />

<title>Module 08: DE Normalization and Modeling</title>

<script src="site_libs/header-attrs-2.21/header-attrs.js"></script>
<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/paper.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<script src="site_libs/navigation-1.1/codefolding.js"></script>
<script src="site_libs/navigation-1.1/sourceembed.js"></script>
<link href="site_libs/highlightjs-9.12.0/default.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link rel="shortcut icon" href="favicon-16x16.png" />
<!--
Favicon from
https://twemoji.twitter.com/
https://favicon.io/emoji-favicons/dna/
-->

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>




<style type="text/css">
#rmd-source-code {
  display: none;
}
</style>





<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
details > summary > p:only-child {
  display: inline;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark the anchor link active (and if it's in a dropdown, also mark that active)
  var dropdown = menuAnchor.closest('li.dropdown');
  if (window.bootstrap) { // Bootstrap 4+
    menuAnchor.addClass('active');
    dropdown.find('> .dropdown-toggle').addClass('active');
  } else { // Bootstrap 3
    menuAnchor.parent().addClass('active');
    dropdown.addClass('active');
  }

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before, .tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "\e259";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "\e258";
  font-family: 'Glyphicons Halflings';
  border: none;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->
<style type="text/css">
.code-folding-btn { margin-bottom: 4px; }
</style>



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-inverse  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-bs-toggle="collapse" data-target="#navbar" data-bs-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">RNA-Seq Demystified</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="Module00_Introduction.html">Intro</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Sequence QC
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="Module01_Warming_Up.html">Warming up</a>
    </li>
    <li>
      <a href="Module02a_Sequence_QC.html">Sequence QC</a>
    </li>
    <li>
      <a href="Module02b_Trimming.html">Trimming</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Alignment
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="Module03a_Reference_Genomes.html">Reference Genomes</a>
    </li>
    <li>
      <a href="Module03b_Alignment.html">Alignment</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Quantification
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="Module04_Alignment_QC_and_Quantification.html">Alignment QC and Quantification</a>
    </li>
    <li>
      <a href="Module05_Additional_Details.html">Optional: Additional details</a>
    </li>
  </ul>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Groups/Model
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="Module06_DEAnalysisSetup.html">DE Introduction</a>
    </li>
    <li>
      <a href="Module07_DESeq2Init.html">DE Initialization</a>
    </li>
    <li>
      <a href="Module08_DESeq2DE.html">DE Normalization and Modeling</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Expression plots
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="Module09_SampleQCViz.html">Sample QC</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Diff Expression
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="Module10_DEComparisons.html">DE Testing</a>
    </li>
    <li>
      <a href="Module11_DEVisualizations.html">DE Visualization</a>
    </li>
    <li>
      <a href="Module12_DEAnnotations.html">DE Annotation</a>
    </li>
    <li>
      <a href="R_bonus_content.html">Bonus content</a>
    </li>
  </ul>
</li>
<li>
  <a href="Module99_Wrap_up.html">Wrap up</a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">

<div class="btn-group pull-right float-right">
<button type="button" class="btn btn-default btn-xs btn-secondary btn-sm dropdown-toggle" data-toggle="dropdown" data-bs-toggle="dropdown" aria-haspopup="true" aria-expanded="false"><span>Code</span> <span class="caret"></span></button>
<ul class="dropdown-menu dropdown-menu-right" style="min-width: 50px;">
<li><a id="rmd-download-source" href="#">Download Rmd</a></li>
</ul>
</div>



<h1 class="title toc-ignore">Module 08: DE Normalization and
Modeling</h1>
<h4 class="author">UM Bioinformatics Core</h4>
<h4 class="date">2023-08-07</h4>

</div>


<style type="text/css">
body, td {
   font-size: 18px;
}
code.r{
  font-size: 12px;
}
pre {
  font-size: 12px
}
</style>
<blockquote>
<h1 id="objectives" class="unlisted unnumbered">Objectives</h1>
<ul>
<li>Discuss count normalizations</li>
<li>Execute model fitting for differential expression comparisons</li>
</ul>
</blockquote>
<div id="differential-expression-workflow"
class="section level1 unlisted unnumbered">
<h1 class="unlisted unnumbered">Differential Expression Workflow</h1>
<p>Here we will proceed with count normalizations and fit our DESeq2
model.</p>
<p><img src="images/wayfinder/wayfinder-DESeq2DE.png"
style="width:75.0%" /></p>
<hr />
</div>
<div id="count-normalizations" class="section level1">
<h1>Count normalizations</h1>
<p>Since counts of mapped reads for each gene is proportional to the
expression of RNA in addition to many “uninteresting” other factors,
normalization is the process of scaling raw count values to account for
the “uninteresting” factors and ensure expression levels are more
comparable.</p>
<div id="normalization-goals" class="section level2">
<h2>Normalization goals</h2>
<p>Two common factors that need to be accounted for during normalization
are <strong>sequencing depth</strong> and <strong>gene length</strong>.
Common normalization approaches (such as FPKM, RPKM, CPM, TPM, etc.)
account for one or both of these factors.</p>
<ul>
<li><strong>Sequencing depth</strong> normalization is necessary to
account for the proportion of reads per gene expected for more deeply
sequenced samples (like in pink below) versus a less deeply sequenced
sample (like in green below).</li>
</ul>
<div class="float">
<img src="images/normalization_methods_depth.png" style="width:75.0%"
alt="Note that each pink or green rectangle represents an aligned read, with reads spanning an intron connected by a dashed line. Figure from HBC training materials" />
<div class="figcaption"><em>Note that each pink or green rectangle
represents an aligned read, with reads spanning an intron connected by a
dashed line. Figure from <a
href="https://hbctraining.github.io/DGE_workshop/lessons/02_DGE_count_normalization.html">HBC
training materials</a></em></div>
</div>
<ul>
<li><strong>Gene length</strong> normalization may also be necessary if
comparing between different genes, since genes of different lengths have
different probabilities of generating fragments that end up in the
library.</li>
</ul>
<blockquote>
<p><strong>Note</strong>: The above figure was originally from a <a
href="https://hbctraining.github.io/DGE_workshop/lessons/02_DGE_count_normalization.html">HBC
tutorial</a> that also includes a detailed comparison of different
normalization (CPM, TPM, FPKM) approaches and their best uses.</p>
</blockquote>
<p><strong>Check-in: Questions about normalizations?</strong></p>
</div>
<div id="deseq2-normalizations" class="section level2">
<h2>DESeq2 normalizations</h2>
<p>DESeq2 has an <a
href="https://genomebiology.biomedcentral.com/articles/10.1186/gb-2010-11-10-r106">internal
normalization process</a> that accounts for <strong>RNA
composition</strong>. A few highly differentially expressed genes,
differences in the number of genes expressed between samples, or
contamination are not accounted for by depth or gene length
normalization methods. Accounting for RNA composition is particularly
important for differential expression analyses, regardless of the tool
used.</p>
<p>For data exploration and visualizations, it is helpful to generate an
object of independently normalized counts. We will use the <a
href="http://bioconductor.org/packages/devel/bioc/vignettes/DESeq2/inst/doc/DESeq2.html#count-data-transformations">rlog
transformation</a> from DESeq2 that accounts for sequencing depth for
each sample and RNA composition for the downstream quality control
visualizations.</p>
<p>The rlog transformation produces log2 scaled data that has also been
normalized to overall library size as well as variance across genes at
different mean expression levels. For larger numbers of samples, there
is an alternative transformation method, <a
href="http://www.bioconductor.org/packages/release/bioc/vignettes/DESeq2/inst/doc/DESeq2.html#count-data-transformations">vst</a>
that can be used instead for count normalizations.</p>
<p>The command to generate the normalized count object has a few parts,
including <code>dds</code> as an input and providing a value to the
option <code>blind</code>. For our purposes, we set
<code>blind = TRUE</code> because we want to compare samples in
downstream QC plots in an unbiased manner.</p>
<pre class="r"><code>rld = rlog(dds_filtered, blind = TRUE)</code></pre>
<p>Next, we’ll look at the results of the transformation by extracting
the values with the <code>assay()</code> function.</p>
<pre class="r"><code>head(assay(rld), 2)</code></pre>
<pre><code>                   sample_A sample_B sample_C sample_D sample_E sample_F
ENSMUSG00000000001 10.51481 10.36671 10.41946 10.84037 10.41045 10.57877
ENSMUSG00000000028 10.60446 10.73451 10.73503 10.68271 10.82094 10.99100</code></pre>
<p>Looking at the rld values, we can see that they are now in log scale.
Since we set <code>blind=TRUE</code>, the transformation is blind to the
sample information we specified in the design formula. The normalized
counts are helpful for visualization methods during expression-level
quality assessment but <strong>aren’t used in the model
fitting</strong>.</p>
<p>We’ll come back to these normalized data, but first let’s write out
both the raw and normalized count tables to file.</p>
<div id="output-count-tables" class="section level3">
<h3>Output count tables</h3>
<p>To output the raw counts, we will need to use the <code>counts</code>
function to access the count table from within its larger
<code>DESeqDataSet</code> object.</p>
<pre class="r"><code>write.csv(counts(dds, normalized = FALSE), file=&quot;outputs/tables/DESeq2_raw_counts.csv&quot;)</code></pre>
<p>Then we’ll output the rlog count table, using the <code>assay</code>
function to access the normalized count table from within its larger
<code>DESeqDataSet</code> object.</p>
<pre class="r"><code>write.csv(assay(rld), file=&quot;outputs/tables/DESeq2_rlog_normalized_counts.csv&quot;)</code></pre>
</div>
</div>
</div>
<div id="deseq2-model-fitting" class="section level1">
<h1>DESeq2 Model Fitting</h1>
<p>Next, we’ll fit our standard model using the <code>DESeq</code>
function and take a look at the objects we generate. This command
applies the model to our data, using the sample information supplied
when generating the initial <code>dds</code> object so can take some
time to run.</p>
<pre class="r"><code>dds_fitted = DESeq(dds_filtered)
dds_fitted</code></pre>
<pre><code>class: DESeqDataSet 
dim: 16249 6 
metadata(1): version
assays(4): counts mu H cooks
rownames(16249): ENSMUSG00000000001 ENSMUSG00000000028 ...
  ENSMUSG00000118651 ENSMUSG00000118653
rowData names(22): baseMean baseVar ... deviance maxCooks
colnames(6): sample_A sample_B ... sample_E sample_F
colData names(3): genotype condition sizeFactor</code></pre>
<details>
<summary>
<em>Dispersion models and possible warning messages</em>
</summary>
<p>Depending on the data set you are analyzing, you may see a warning
that the default ‘parametric’ dispersion model so a local regression was
substituted. When seeing this warning, we recommend looking at a
dispersion plot with the <code>plotDispEsts(dds)</code> function, but as
<a href="https://support.bioconductor.org/p/107937/">this bioconductor
thread</a> discusses, other visualizations of our data might be more
helpful and/or easier to interpret why the data doesn’t fit the default
parametric model.</p>
</details>
<p><br></p>
<p>Notice that there is now more information in the
<code>DESeqDataSet</code> object than there was prior to our
normalization. There is information about the model fit and about the
library size normalization. DESeq2 will use this information when we
perform the test for differential expression.</p>
<p>The <code>DESeq()</code> function is actually doing three things
automatically for us. It calculates:</p>
<ol style="list-style-type: decimal">
<li>The size factors to normalize for library size with
<code>estimateSizeFactors(dds_filtered)</code>,</li>
<li>Dispersion estimates to shrink the dispersions with
<code>estimateDispersions(dds_filtered)</code>, and</li>
<li>The Wald test statistics with
<code>nbinomWaldTest(dds_filtered)</code>.</li>
</ol>
<p>The <code>resultsNames()</code> function returns the names of the
estimated effects of the model.</p>
<pre class="r"><code>resultsNames(dds_fitted)</code></pre>
<pre><code>[1] &quot;Intercept&quot;               &quot;condition_minus_vs_plus&quot;</code></pre>
<p>The results include the single comparison representing the two levels
of <code>condition</code>. If there were more levels in the
<code>condition</code> column, there would be more results listed here
because DESeq2 would implicitly compare all other levels to the
reference level.</p>
<p><strong>Checkpoint</strong>: <em>If you see the same results when you
execute <code>resultsNames(dds_filtered)</code>, please indicate with
the green ‘yes’ button. Otherwise, please use the red ‘x’ button to get
help before the break</em></p>
<div id="save-fitted-model-and-data-in-robj" class="section level2">
<h2>Save fitted model and data in Robj</h2>
<p>It can be useful to save key R objects to file as we proceed through
our analysis - before we do that, let’s look at the documentation for
the <code>save</code> function to see if it does what we want.</p>
<pre class="r"><code>?save</code></pre>
<p>As we can see from the documentation the function <code>save</code>
writes an “external representation” of R objects that can be read back
from the file at a later date by using the function <code>load</code> or
<code>attach</code> in most cases. We’ll proceed with saving our
<code>dds_fitted</code> object, creating a subdirectory first.</p>
<pre class="r"><code>dir.create(&quot;outputs/Robjs&quot;, recursive=TRUE)</code></pre>
<pre><code>Warning in dir.create(&quot;outputs/Robjs&quot;, recursive = TRUE): &#39;outputs/Robjs&#39; already
exists</code></pre>
<pre class="r"><code>save(dds_fitted,
          file=&quot;outputs/Robjs/dds_fitted.Robj&quot;)</code></pre>
</div>
</div>
<div id="optional-content" class="section level1">
<h1>Optional content</h1>
<details>
<summary>
<em>Click to fit a model that includes a covariate</em>
</summary>
<p>If you executed the commands in the additional content section from
Module 07, you can fit a separate DESeq2 model for the batch
example.</p>
<pre class="r"><code>dds_batch_fitted = DESeq(dds_batch)</code></pre>
<pre><code>estimating size factors</code></pre>
<pre><code>estimating dispersions</code></pre>
<pre><code>gene-wise dispersion estimates</code></pre>
<pre><code>mean-dispersion relationship</code></pre>
<pre><code>final dispersion estimates</code></pre>
<pre><code>fitting model and testing</code></pre>
<pre class="r"><code>resultsNames(dds_batch_fitted)</code></pre>
<pre><code>[1] &quot;Intercept&quot;               &quot;batch_Day2_vs_Day1&quot;      &quot;batch_Day3_vs_Day1&quot;     
[4] &quot;condition_minus_vs_plus&quot;</code></pre>
</details>
<p><br></p>
</div>
<div id="summary" class="section level1">
<h1>Summary</h1>
<p>In this section, we:</p>
<ul>
<li>Learned about count normalizations and uses</li>
<li>Generated a normalized count table</li>
<li>Fit two DESeq2 models for our data</li>
<li>(Optional) - saw the impact of including a covariate in our
model</li>
<li>Wrote intermediate data to file</li>
</ul>
<hr />
</div>
<div id="sources" class="section level1">
<h1>Sources</h1>
<p>Training resources used to develop materials</p>
<ul>
<li>HBC DGE setup: <a
href="https://hbctraining.github.io/DGE_workshop/lessons/01_DGE_setup_and_overview.html"
class="uri">https://hbctraining.github.io/DGE_workshop/lessons/01_DGE_setup_and_overview.html</a></li>
<li>HBC Count Normalization: <a
href="https://hbctraining.github.io/DGE_workshop/lessons/02_DGE_count_normalization.html"
class="uri">https://hbctraining.github.io/DGE_workshop/lessons/02_DGE_count_normalization.html</a></li>
<li>DESeq2 standard vignette: <a
href="http://bioconductor.org/packages/devel/bioc/vignettes/DESeq2/inst/doc/DESeq2.html"
class="uri">http://bioconductor.org/packages/devel/bioc/vignettes/DESeq2/inst/doc/DESeq2.html</a></li>
<li>DESeq2 beginners vignette: <a
href="https://bioc.ism.ac.jp/packages/2.14/bioc/vignettes/DESeq2/inst/doc/beginner.pdf"
class="uri">https://bioc.ism.ac.jp/packages/2.14/bioc/vignettes/DESeq2/inst/doc/beginner.pdf</a></li>
<li>Bioconductor RNA-seq Workflows: <a
href="https://www.bioconductor.org/help/course-materials/2015/LearnBioconductorFeb2015/B02.1_RNASeq.html"
class="uri">https://www.bioconductor.org/help/course-materials/2015/LearnBioconductorFeb2015/B02.1_RNASeq.html</a></li>
</ul>
<hr />
<p>These materials have been adapted and extended from materials listed
above. These are open access materials distributed under the terms of
the <a href="http://creativecommons.org/licenses/by/4.0/">Creative
Commons Attribution license (CC BY 4.0)</a>, which permits unrestricted
use, distribution, and reproduction in any medium, provided the original
author and source are credited.</p>
<br/> <br/>
<hr/>
<table style="width:100%;">
<colgroup>
<col width="28%" />
<col width="42%" />
<col width="28%" />
</colgroup>
<thead>
<tr class="header">
<th align="left"><a href="Module07_DESeq2Init.html">Previous
lesson</a></th>
<th align="center"><a href="#top">Top of this lesson</a></th>
<th align="right"><a href="Module09_SampleQCViz.html">Next
lesson</a></th>
</tr>
</thead>
<tbody>
</tbody>
</table>
</div>

<div id="rmd-source-code">---
title: "Module 08: DE Normalization and Modeling"
author: "UM Bioinformatics Core"
date: "`r Sys.Date()`"
output:
        html_document:
            includes:
                in_header: header.html
            theme: paper
            toc: true
            toc_depth: 4
            toc_float: true
            number_sections: false
            fig_caption: true
            markdown: GFM
            code_download: true
---

<style type="text/css">
body, td {
   font-size: 18px;
}
code.r{
  font-size: 12px;
}
pre {
  font-size: 12px
}
</style>

```{r, include = FALSE}
source("../bin/chunk-options.R")
knitr_fig_path("08-")
```

> # Objectives {.unlisted .unnumbered}
> * Discuss count normalizations
> * Execute model fitting for differential expression comparisons

```{r LoadRunningData, eval=TRUE, echo=FALSE, message=FALSE, warning=FALSE}
library(DESeq2)
# load("rdata/RunningData.RData")
```

# Differential Expression Workflow {.unlisted .unnumbered}

Here we will proceed with count normalizations and fit our DESeq2 model.

![](./images/wayfinder/wayfinder-DESeq2DE.png){width=75%}

---

# Count normalizations

Since counts of mapped reads for each gene is proportional to the expression of RNA in addition to many “uninteresting” other factors, normalization is the process of scaling raw count values to account for the “uninteresting” factors and ensure expression levels are more comparable.

## Normalization goals

Two common factors that need to be accounted for during normalization are **sequencing depth** and **gene length**. Common normalization approaches (such as FPKM, RPKM, CPM, TPM, etc.) account for one or both of these factors.

* **Sequencing depth** normalization is necessary to account for the proportion of reads per gene expected for more deeply sequenced samples (like in pink below) versus a less deeply sequenced sample (like in green below).

![*Note that each pink or green rectangle represents an aligned read, with reads spanning an intron connected by a dashed line. Figure from [HBC  training materials](https://hbctraining.github.io/DGE_workshop/lessons/02_DGE_count_normalization.html)*](./images/normalization_methods_depth.png){width=75%}

* **Gene length** normalization may also be necessary if comparing between different genes, since genes of different lengths have different probabilities of generating fragments that end up in the library.

> **Note**: The above figure was originally from a [HBC  tutorial](https://hbctraining.github.io/DGE_workshop/lessons/02_DGE_count_normalization.html) that also includes a detailed comparison of different normalization (CPM, TPM, FPKM) approaches and their best uses.

**Check-in: Questions about normalizations?**

## DESeq2 normalizations

DESeq2 has an [internal normalization process](https://genomebiology.biomedcentral.com/articles/10.1186/gb-2010-11-10-r106) that accounts for **RNA composition**. A few highly differentially expressed genes, differences in the number of genes expressed between samples, or contamination are not accounted for by depth or gene length normalization methods. Accounting for RNA composition is particularly important for differential expression analyses, regardless of the tool used.

For data exploration and visualizations, it is helpful to generate an object of independently normalized counts. We will use the [rlog transformation](http://bioconductor.org/packages/devel/bioc/vignettes/DESeq2/inst/doc/DESeq2.html#count-data-transformations) from DESeq2 that accounts for sequencing depth for each sample and RNA composition for the downstream quality control visualizations.

The rlog transformation produces log2 scaled data that has also been normalized to overall library size as well as variance across genes at different mean expression levels. For larger numbers of samples, there is an alternative transformation method, [vst](http://www.bioconductor.org/packages/release/bioc/vignettes/DESeq2/inst/doc/DESeq2.html#count-data-transformations) that can be used instead for count normalizations.

The command to generate the normalized count object has a few parts, including `dds` as an input and providing a value to the option `blind`. For our purposes, we set `blind = TRUE` because we want to compare samples in downstream QC plots in an unbiased manner.

```{r CountNorm}
rld = rlog(dds_filtered, blind = TRUE)
```

Next, we'll look at the results of the transformation by extracting the values with the `assay()` function.

```{r CountNormCheck}
head(assay(rld), 2)
```

Looking at the rld values, we can see that they are now in log scale. Since we set `blind=TRUE`, the transformation is blind to the sample information we specified in the design formula. The normalized counts are helpful for visualization methods during expression-level quality assessment but **aren't used in the model fitting**.

We'll come back to these normalized data, but first let's write out both the raw and normalized count tables to file.

### Output count tables

To output the raw counts, we will need to use the `counts` function to access the count table from within its larger `DESeqDataSet` object.

```{r OutputCountsRaw, eval = FALSE}
write.csv(counts(dds, normalized = FALSE), file="outputs/tables/DESeq2_raw_counts.csv")
```

Then we'll output the rlog count table, using the `assay` function to access the normalized count table from within its larger `DESeqDataSet` object.

```{r OutputCountsRlog, eval = FALSE}
write.csv(assay(rld), file="outputs/tables/DESeq2_rlog_normalized_counts.csv")
```


# DESeq2 Model Fitting

Next, we'll fit our standard model using the `DESeq` function and take a look at the objects we generate. This command applies the model to our data, using the sample information supplied when generating the initial `dds` object so can take some time to run.

```{r FitModelStandard, message=FALSE}
dds_fitted = DESeq(dds_filtered)
dds_fitted
```

<details>
    <summary>*Dispersion models and possible warning messages*</summary>

    Depending on the data set you are analyzing, you may see a warning that the default 'parametric' dispersion model so a local regression was substituted. When seeing this warning, we recommend looking at a dispersion plot with the `plotDispEsts(dds)` function, but as [this bioconductor thread](https://support.bioconductor.org/p/107937/) discusses, other visualizations of our data might be more helpful and/or easier to interpret why the data doesn't fit the default parametric model.

</details>
<br>


Notice that there is now more information in the `DESeqDataSet` object than there was prior to our normalization. There is information about the model fit and about the library size normalization. DESeq2 will use this information when we perform the test for differential expression.

The `DESeq()` function is actually doing three things automatically for us. It calculates:

1. The size factors to normalize for library size with `estimateSizeFactors(dds_filtered)`,
2. Dispersion estimates to shrink the dispersions with `estimateDispersions(dds_filtered)`, and
3. The Wald test statistics with `nbinomWaldTest(dds_filtered)`.

The `resultsNames()` function returns the names of the estimated effects of the model.

```{r FitModelStandardCheck1, eval=TRUE}
resultsNames(dds_fitted)
```

The results include the single comparison representing the two levels of `condition`. If there were more levels in the `condition` column, there would be more results listed here because DESeq2 would implicitly compare all other levels to the reference level.


**Checkpoint**: *If you see the same results when you execute `resultsNames(dds_filtered)`, please indicate with the green 'yes' button. Otherwise, please use the red 'x' button to get help before the break*

## Save fitted model and data in Robj

It can be useful to save key R objects to file as we proceed through our analysis - before we do that, let's look at the documentation for the `save` function to see if it does what we want.

```{r SaveDocumentation}
?save
```

As we can see from the documentation the function `save` writes an "external representation" of R objects that can be read back from the file at a later date by using the function `load` or `attach` in most cases. We'll proceed with saving our `dds_fitted` object, creating a subdirectory first.

```{r WriteFitModelObject}
dir.create("outputs/Robjs", recursive=TRUE)
save(dds_fitted,
          file="outputs/Robjs/dds_fitted.Robj")
```


# Optional content

<details>
<summary>*Click to fit a model that includes a covariate*</summary>

If you executed the commands in the additional content section from Module 07, you can fit a separate DESeq2 model for the batch example.

```{r FitModelCovariate}
dds_batch_fitted = DESeq(dds_batch)
resultsNames(dds_batch_fitted)
```

</details>

<br>

# Summary

In this section, we:

* Learned about count normalizations and uses
* Generated a normalized count table
* Fit two DESeq2 models for our data
* (Optional) - saw the impact of including a covariate in our model
* Wrote intermediate data to file

---

# Sources

Training resources used to develop materials

* HBC DGE setup: https://hbctraining.github.io/DGE_workshop/lessons/01_DGE_setup_and_overview.html
* HBC Count Normalization: https://hbctraining.github.io/DGE_workshop/lessons/02_DGE_count_normalization.html
* DESeq2 standard vignette: http://bioconductor.org/packages/devel/bioc/vignettes/DESeq2/inst/doc/DESeq2.html
* DESeq2 beginners vignette: https://bioc.ism.ac.jp/packages/2.14/bioc/vignettes/DESeq2/inst/doc/beginner.pdf
* Bioconductor RNA-seq Workflows: https://www.bioconductor.org/help/course-materials/2015/LearnBioconductorFeb2015/B02.1_RNASeq.html

```{r WriteOut.RData, eval=TRUE, echo=FALSE, message=FALSE, warning=FALSE}
# Hidden code block to write out data for knitting
# save.image(file = "rdata/RunningData.RData")
```

---

These materials have been adapted and extended from materials listed above. These are open access materials distributed under the terms of the [Creative Commons Attribution license (CC BY 4.0)](http://creativecommons.org/licenses/by/4.0/), which permits unrestricted use, distribution, and reproduction in any medium, provided the original author and source are credited.

<br/>
<br/>
<hr/>
| [Previous lesson](Module07_DESeq2Init.html) | [Top of this lesson](#top) | [Next lesson](Module09_SampleQCViz.html) |
| :--- | :----: | ---: |</div>


</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->
<script>
$(document).ready(function () {
  window.initializeSourceEmbed("Module08_DESeq2DE.Rmd");
});
</script>

<script>
$(document).ready(function ()  {

    // temporarily add toc-ignore selector to headers for the consistency with Pandoc
    $('.unlisted.unnumbered').addClass('toc-ignore')

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3,h4",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
