<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />


<meta name="author" content="UM Bioinformatics Core" />

<meta name="date" content="2024-07-05" />

<title>DE Visualization</title>

<script src="site_libs/header-attrs-2.27/header-attrs.js"></script>
<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/paper.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/jqueryui-1.13.2/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<script src="site_libs/navigation-1.1/codefolding.js"></script>
<script src="site_libs/navigation-1.1/sourceembed.js"></script>
<link href="site_libs/highlightjs-9.12.0/default.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link rel="shortcut icon" href="favicon-16x16.png" />
<!--
Favicon from
https://twemoji.twitter.com/
https://favicon.io/emoji-favicons/dna/
-->

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>




<style type="text/css">
#rmd-source-code {
  display: none;
}
</style>





<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
details > summary > p:only-child {
  display: inline;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark the anchor link active (and if it's in a dropdown, also mark that active)
  var dropdown = menuAnchor.closest('li.dropdown');
  if (window.bootstrap) { // Bootstrap 4+
    menuAnchor.addClass('active');
    dropdown.find('> .dropdown-toggle').addClass('active');
  } else { // Bootstrap 3
    menuAnchor.parent().addClass('active');
    dropdown.addClass('active');
  }

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before, .tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "\e259";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "\e258";
  font-family: 'Glyphicons Halflings';
  border: none;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->
<style type="text/css">
.code-folding-btn { margin-bottom: 4px; }
</style>



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-inverse  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-bs-toggle="collapse" data-target="#navbar" data-bs-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">RNA-Seq Demystified</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="Module00_Introduction.html">Intro</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Sequence QC
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="Module01_Warming_Up.html">Warming up</a>
    </li>
    <li>
      <a href="Module02a_Sequence_QC.html">Sequence QC</a>
    </li>
    <li>
      <a href="Module02b_Trimming.html">Trimming</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Alignment
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="Module03a_Reference_Genomes.html">Reference Genomes</a>
    </li>
    <li>
      <a href="Module03b_Alignment.html">Alignment</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Quantification
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="Module04_Alignment_QC_and_Quantification.html">Alignment QC and Quantification</a>
    </li>
    <li>
      <a href="Module05_Additional_Details.html">Optional: Additional details</a>
    </li>
  </ul>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Groups/Model
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="Module06_DEAnalysisSetup.html">DE Introduction</a>
    </li>
    <li>
      <a href="Module07_DESeq2Init.html">DE Initialization</a>
    </li>
    <li>
      <a href="Module08_DESeq2DE.html">DE Normalization and Modeling</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Expression plots
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="Module09_SampleQCViz.html">Sample QC</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Diff Expression
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="Module10_DEComparisons.html">DE Testing</a>
    </li>
    <li>
      <a href="Module10a_batchAndCovars.html">Batches and Covariates</a>
    </li>
    <li>
      <a href="Module11_DEVisualizations.html">DE Visualization</a>
    </li>
    <li>
      <a href="Module12_DEAnnotations.html">DE Annotation</a>
    </li>
    <li>
      <a href="R_bonus_content.html">Bonus content</a>
    </li>
  </ul>
</li>
<li>
  <a href="Module99_Wrap_up.html">Wrap up</a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">

<div class="btn-group pull-right float-right">
<button type="button" class="btn btn-default btn-xs btn-secondary btn-sm dropdown-toggle" data-toggle="dropdown" data-bs-toggle="dropdown" aria-haspopup="true" aria-expanded="false"><span>Code</span> <span class="caret"></span></button>
<ul class="dropdown-menu dropdown-menu-right" style="min-width: 50px;">
<li><a id="rmd-download-source" href="#">Download Rmd</a></li>
</ul>
</div>



<h1 class="title toc-ignore">DE Visualization</h1>
<h4 class="author">UM Bioinformatics Core</h4>
<h4 class="date">2024-07-05</h4>

</div>


<style type="text/css">
body, td {
   font-size: 18px;
}
code.r{
  font-size: 12px;
}
pre {
  font-size: 12px
}
</style>
<p>In this module, we will learn:</p>
<ul>
<li>How to generate common differential expression visualizations</li>
<li>How to choose thresholds for calling differentially expressed
genes</li>
<li>Options for functional enrichments and other downstream
analyses</li>
</ul>
<p><br></p>
<div id="differential-expression-workflow"
class="section level1 unlisted unnumbered">
<h1 class="unlisted unnumbered">Differential Expression Workflow</h1>
<p>Here we will generate summary figures for our results and annotate
our DE tables.</p>
<p><img src="images/wayfinder/wayfinder-DEVisualizations.png"
style="width:75.0%" /></p>
<hr />
</div>
<div id="summarizing-de-results" class="section level1">
<h1>Summarizing DE results</h1>
<p>Part of differential expression analysis is generating visualizations
and summary tables to share our results. While the DESeq2 vignette
provides <a
href="http://bioconductor.org/packages/devel/bioc/vignettes/DESeq2/inst/doc/DESeq2.html#exploring-and-exporting-results">examples
of other visualizations</a>, a common visualization to summarize DE
comparisons are <a
href="http://resources.qiagenbioinformatics.com/manuals/clcgenomicsworkbench/752/index.php?manual=Volcano_plots_inspecting_result_statistical_analysis.html">volcano
plots</a>.</p>
<div id="tabular-de-summary" class="section level2">
<h2>Tabular DE summary</h2>
<p>To generate a general summary of the DE results, we can use the
<code>summary</code> function to generate a basic summary by DESeq2.</p>
<pre class="r"><code>summary(results_deficient_vs_control)</code></pre>
<pre><code>
out of 16249 with nonzero total read count
adjusted p-value &lt; 0.1
LFC &gt; 0 (up)       : 53, 0.33%
LFC &lt; 0 (down)     : 38, 0.23%
outliers [1]       : 72, 0.44%
low counts [2]     : 3781, 23%
(mean count &lt; 7)
[1] see &#39;cooksCutoff&#39; argument of ?results
[2] see &#39;independentFiltering&#39; argument of ?results</code></pre>
<p>However, this summary is simply a text output that we are unable to
manipulate. Moreover, notice that the thresholds are not quite as we
would like them.</p>
<p>To create our own summary, we first need to choose thresholds. A
standard threshold for the adjusted p-value is less than 0.05. A
reasonable threshold for linear fold-change is less than -1.5 or greater
than 1.5 (which corresponds to log2 fold-change -0.585 and 0.585,
respectively). Including a fold-change threshold ensures that the DE
genes have a reasonable effect size as well as statistical
significance.</p>
<p>Let’s set these thresholds as variables to reuse. This is generally
good practice because if you want to change those thresholds later then
you only have to change them in one location of your script, which is
faster and can reduce the risk of errors from missing some instances in
your code.</p>
<pre class="r"><code>fc = 1.5
fdr = 0.05</code></pre>
<blockquote>
<h1 id="note-choosing-thresholds" class="unlisted unnumbered">Note:
Choosing thresholds</h1>
<p>Thresholding on adjusted p-values &lt; 0.05 is a standard threshold,
but depending on the research question and/or how the results will be
used, other thresholds might be reasonable.</p>
<p>There is a nice <a href="https://www.biostars.org/p/209118/">Biostar
post that discusses choosing adjusted p-value thresholds</a>, including
cases where a more relaxed threshold might be appropriate and (some
heated) discussion of the dangers of adjusting the choosen threshold
after running an analysis. Additionally, there is a <a
href="https://bmcbioinformatics.biomedcentral.com/articles/10.1186/1471-2105-13-S2-S11">Dalmon
et al 2012</a> paper about p-value and fold-change thresholds for
microarray data that may help provide some context.</p>
</blockquote>
<p>If we think back to Computational Foundations, conditional statements
could allow us to determine the number of genes that pass our
thresholds, which would be useful for annotating our results tables and
plots.</p>
<blockquote>
<h1 id="exercise" class="unlisted unnumbered">Exercise</h1>
<p>How would we identify the number of genes with adjusted p-values &lt;
0.05 and a fold-change above 1.5 (or below -1.5) in our comparison?</p>
</blockquote>
<details>
<summary>
Solution
</summary>
<p>Here is one possible answer:</p>
<pre class="r"><code>sum(results_deficient_vs_control$padj &lt; fdr &amp; abs(results_deficient_vs_control$log2FoldChange) &gt;= log2(fc), na.rm = TRUE)</code></pre>
<pre><code>[1] 56</code></pre>
<p><strong>Checkpoint</strong>: <em>If you see the same number of DE
ganes with our choosen thresholds, please indicate with a green check.
Otherwise use a red x.</em></p>
</details>
<p><br></p>
<p>Let’s now create a new column in
<code>results_deficient_vs_control</code> to record the significance
“call” based on these thresholds. And let’s separate the call by “Up” or
“Down”, noting that these are relative to our “Case” condition. There
are many ways to accomplish this, but the following will work:</p>
<p>First define all values as “NS” or “not significant”:</p>
<pre class="r"><code>results_deficient_vs_control$call = &#39;NS&#39;
head(results_deficient_vs_control)</code></pre>
<pre><code>log2 fold change (MLE): condition deficient vs control 
Wald test p-value: condition deficient vs control 
DataFrame with 6 rows and 7 columns
                      baseMean log2FoldChange     lfcSE      stat    pvalue      padj
                     &lt;numeric&gt;      &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt;
ENSMUSG00000000001  1489.83039       0.297760  0.210310  1.415815  0.156830  0.868573
ENSMUSG00000000028  1748.93544       0.226421  0.176795  1.280695  0.200301  0.902900
ENSMUSG00000000031  2151.87715       0.457635  0.764579  0.598545  0.549476  0.995391
ENSMUSG00000000037    24.91672       0.579130  0.561259  1.031840  0.302147  0.950613
ENSMUSG00000000049     7.78377      -0.899483  1.553063 -0.579167  0.562476  0.998043
ENSMUSG00000000056 19653.54030      -0.174048  0.203529 -0.855151  0.392468  0.982479
                          call
                   &lt;character&gt;
ENSMUSG00000000001          NS
ENSMUSG00000000028          NS
ENSMUSG00000000031          NS
ENSMUSG00000000037          NS
ENSMUSG00000000049          NS
ENSMUSG00000000056          NS</code></pre>
<p>Next, determine the “Up” and “Down” indices:</p>
<pre class="r"><code>up_idx = results_deficient_vs_control$padj &lt; fdr &amp; results_deficient_vs_control$log2FoldChange &gt; log2(fc)
down_idx = results_deficient_vs_control$padj &lt; fdr &amp; results_deficient_vs_control$log2FoldChange &lt; -log2(fc)</code></pre>
<p>Last, use those indices to assign the correct “Up” or “Down” values
to the correct indices, and look at the head of the result:</p>
<pre class="r"><code>results_deficient_vs_control$call[up_idx] = &#39;Up&#39;
results_deficient_vs_control$call[down_idx] = &#39;Down&#39;
head(results_deficient_vs_control)</code></pre>
<pre><code>log2 fold change (MLE): condition deficient vs control 
Wald test p-value: condition deficient vs control 
DataFrame with 6 rows and 7 columns
                      baseMean log2FoldChange     lfcSE      stat    pvalue      padj
                     &lt;numeric&gt;      &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt;
ENSMUSG00000000001  1489.83039       0.297760  0.210310  1.415815  0.156830  0.868573
ENSMUSG00000000028  1748.93544       0.226421  0.176795  1.280695  0.200301  0.902900
ENSMUSG00000000031  2151.87715       0.457635  0.764579  0.598545  0.549476  0.995391
ENSMUSG00000000037    24.91672       0.579130  0.561259  1.031840  0.302147  0.950613
ENSMUSG00000000049     7.78377      -0.899483  1.553063 -0.579167  0.562476  0.998043
ENSMUSG00000000056 19653.54030      -0.174048  0.203529 -0.855151  0.392468  0.982479
                          call
                   &lt;character&gt;
ENSMUSG00000000001          NS
ENSMUSG00000000028          NS
ENSMUSG00000000031          NS
ENSMUSG00000000037          NS
ENSMUSG00000000049          NS
ENSMUSG00000000056          NS</code></pre>
<p>Finally, since plotting functions often require categorical
groupings, let’s make this <code>call</code> column a factor and specify
the level ordering:</p>
<pre class="r"><code>results_deficient_vs_control$call = factor(results_deficient_vs_control$call, levels = c(&#39;Up&#39;, &#39;Down&#39;, &#39;NS&#39;))
head(results_deficient_vs_control)</code></pre>
<pre><code>log2 fold change (MLE): condition deficient vs control 
Wald test p-value: condition deficient vs control 
DataFrame with 6 rows and 7 columns
                      baseMean log2FoldChange     lfcSE      stat    pvalue      padj
                     &lt;numeric&gt;      &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt;
ENSMUSG00000000001  1489.83039       0.297760  0.210310  1.415815  0.156830  0.868573
ENSMUSG00000000028  1748.93544       0.226421  0.176795  1.280695  0.200301  0.902900
ENSMUSG00000000031  2151.87715       0.457635  0.764579  0.598545  0.549476  0.995391
ENSMUSG00000000037    24.91672       0.579130  0.561259  1.031840  0.302147  0.950613
ENSMUSG00000000049     7.78377      -0.899483  1.553063 -0.579167  0.562476  0.998043
ENSMUSG00000000056 19653.54030      -0.174048  0.203529 -0.855151  0.392468  0.982479
                       call
                   &lt;factor&gt;
ENSMUSG00000000001       NS
ENSMUSG00000000028       NS
ENSMUSG00000000031       NS
ENSMUSG00000000037       NS
ENSMUSG00000000049       NS
ENSMUSG00000000056       NS</code></pre>
<blockquote>
<h1 id="tip" class="unlisted unnumbered">Tip</h1>
<p>It is often helpful to include code like this in differential
expression analyses so there is a clearly labelled column that makes
subsetting and summarizing the results easier.</p>
</blockquote>
<p>Now we are in a position to quickly summarize our differential
expression results:</p>
<pre class="r"><code>table(results_deficient_vs_control$call)</code></pre>
<pre><code>
   Up  Down    NS 
   41    15 16193 </code></pre>
<p>We see quickly how many genes were “Up” in iron replete, how many
were “Down” in iron replete, and how many were not significant.</p>
<p><strong>Checkpoint</strong>: <em>If you successfully added the
<code>call</code> column and got the same table result as above, please
indicate with a green check. Otherwise use a red x.</em></p>
</div>
<div id="visual-de-summary" class="section level2">
<h2>Visual DE summary</h2>
<p>As described by the <a
href="https://galaxyproject.github.io/training-material/topics/transcriptomics/tutorials/rna-seq-viz-with-volcanoplot/tutorial.html">Galaxy
project</a>, a volcano plot is a type of scatterplot that shows
statistical significance (adjusted p-value) versus effect size (fold
change). In a volcano plot, the most upregulated genes are towards the
right, the most downregulated genes are towards the left, and the most
statistically significant genes are towards the top.</p>
<p>Let’s coerce the <code>DataFrame</code> which was returned by
<code>DESeq2::results()</code> into a <code>tibble</code> in
anticipation of using the <code>ggplot2</code> library to plot. We’re
also going to modify our results table so that the row names become a
separate column, and so that it’s ordered by adjusted p-value.</p>
<pre class="r"><code># Use the rownames argument to create a new column of gene IDs
# Also arrange by adjusted p-value
results_forPlot = as_tibble(results_deficient_vs_control, rownames = &#39;id&#39;) %&gt;% arrange(padj)</code></pre>
<p>Let’s start with a very simple volcano plot that plots the
<code>log2FoldChange</code> on the x-axis, and <code>-log10(padj)</code>
on the y-axis.</p>
<pre class="r"><code># Initialize the plot, specifying the plot type as &#39;geom_point&#39;
ggplot(results_forPlot, aes(x = log2FoldChange, y = -log10(padj))) +
  geom_point()</code></pre>
<pre><code>Warning: Removed 3853 rows containing missing values or values outside the scale range
(`geom_point()`).</code></pre>
<p><img src="fig/11-VolcanoPlot-1.png" width="816" style="display: block; margin: auto;" /></p>
<p>This is a good start, but it’s good practice to add better labels to
the plot with the <code>labs()</code> function:</p>
<pre class="r"><code># Add plot labels and change the theme - save the plot as object `p`
p = ggplot(results_forPlot, aes(x = log2FoldChange, y = -log10(padj))) +
    geom_point() +
    theme_bw() +
    labs(
        title = &#39;Volcano Plot&#39;,
        subtitle = &#39;control vs deficient&#39;,
        x = &#39;log2 fold-change&#39;,
        y = &#39;-log10 FDR&#39;
    )
p</code></pre>
<pre><code>Warning: Removed 3853 rows containing missing values or values outside the scale range
(`geom_point()`).</code></pre>
<p><img src="fig/11-VolcanoPlot2-1.png" width="816" style="display: block; margin: auto;" /></p>
<p>What if we now added some visual guides to indicate where the
significant genes are? We can use the <code>geom_vline()</code> and
<code>geom_hline()</code> functions to accomplish this:</p>
<pre class="r"><code># Add threshold lines
p1 = p +
    geom_vline(
        xintercept = c(0, -log2(fc), log2(fc)),
        linetype = c(1, 2, 2)) +
    geom_hline(
        yintercept = -log10(fdr),
        linetype = 2)
p1</code></pre>
<pre><code>Warning: Removed 3853 rows containing missing values or values outside the scale range
(`geom_point()`).</code></pre>
<p><img src="fig/11-VolcanoPlot3-1.png" width="816" style="display: block; margin: auto;" /></p>
<p>Finally, why not color the points by their significance status? We
already created the <code>call</code> column that has the correct
values. In this case we can get away with adding
<code>geom_point()</code> to our existing plot and specifying the
correct aesthetic:</p>
<pre class="r"><code>p2 = p1 + geom_point(aes(color = call))
p2</code></pre>
<pre><code>Warning: Removed 3853 rows containing missing values or values outside the scale range
(`geom_point()`).
Removed 3853 rows containing missing values or values outside the scale range
(`geom_point()`).</code></pre>
<p><img src="fig/11-VolcanoPlot4-1.png" width="816" style="display: block; margin: auto;" /></p>
<p>Finally, we can adjust our plot to have a nicer color scheme:</p>
<pre class="r"><code>p3 = p2 + scale_color_manual(name = &#39;&#39;, values=c(&#39;#B31B21&#39;,  &#39;#1465AC&#39;, &#39;darkgray&#39;))
p3</code></pre>
<pre><code>Warning: Removed 3853 rows containing missing values or values outside the scale range
(`geom_point()`).
Removed 3853 rows containing missing values or values outside the scale range
(`geom_point()`).</code></pre>
<p><img src="fig/11-unnamed-chunk-2-1.png" width="816" style="display: block; margin: auto;" /></p>
<p>For additional visualizations for our DE results, we included some
example code in the <a
href="https://umich-brcf-bioinf.github.io/2023-03-13-umich-rnaseq-demystified/html/R_bonus_content.html">Bonus
Content module</a> and this <a
href="https://hbctraining.github.io/DGE_workshop/lessons/06_DGE_visualizing_results.html">HBC
tutorial</a> also includes some nice examples.</p>
<div id="subsetting-significant-genes" class="section level3">
<h3>Subsetting significant genes</h3>
<p>You may be interested in creating a table of only the genes that pass
your significance thresholds. A useful way to do this is to
conditionally subset your results. Again, we already created the
<code>call</code> column, which makes this relatively simple to do.</p>
<p><em>Note: The tidyverse functions you learned in Software Carpentry
could also be alternatively used here.</em></p>
<pre class="r"><code># tidyr (requires table reformatting)
res_sig &lt;- as_tibble(results_deficient_vs_control, rownames = &quot;gene_ids&quot;) %&gt;% filter(call != &#39;NS&#39;)

# base
res_sig &lt;- results_deficient_vs_control[results_deficient_vs_control$call != &#39;NS&#39;, ]

head(res_sig)</code></pre>
<pre><code>log2 fold change (MLE): condition deficient vs control 
Wald test p-value: condition deficient vs control 
DataFrame with 6 rows and 7 columns
                    baseMean log2FoldChange     lfcSE      stat      pvalue        padj
                   &lt;numeric&gt;      &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt;   &lt;numeric&gt;   &lt;numeric&gt;
ENSMUSG00000001281   532.398       1.110493  0.282999   3.92402 8.70820e-05 2.45334e-02
ENSMUSG00000004562   499.728       0.750326  0.197826   3.79286 1.48922e-04 3.76743e-02
ENSMUSG00000019916   352.536       1.305994  0.302853   4.31231 1.61561e-05 7.41746e-03
ENSMUSG00000020142   624.633       1.720608  0.287145   5.99212 2.07127e-09 4.27924e-06
ENSMUSG00000020176  3086.303       1.276265  0.262113   4.86914 1.12087e-06 9.26290e-04
ENSMUSG00000020272  1749.268      -0.640126  0.168531  -3.79827 1.45712e-04 3.76300e-02
                       call
                   &lt;factor&gt;
ENSMUSG00000001281     Up  
ENSMUSG00000004562     Up  
ENSMUSG00000019916     Up  
ENSMUSG00000020142     Up  
ENSMUSG00000020176     Up  
ENSMUSG00000020272     Down</code></pre>
<pre class="r"><code>dim(res_sig)</code></pre>
<pre><code>[1] 56  7</code></pre>
</div>
</div>
</div>
<div id="summary" class="section level1">
<h1>Summary</h1>
<p>In this section, we:</p>
<ul>
<li>Generated a volcano plot for our differential expression
results</li>
<li>Summarized our differential expression results</li>
<li>Discussed choosing thresholds</li>
</ul>
<hr />
</div>
<div id="sources" class="section level1">
<h1>Sources</h1>
<ul>
<li>HBC DGE training module, part 1: <a
href="https://hbctraining.github.io/DGE_workshop/lessons/04_DGE_DESeq2_analysis.html"
class="uri">https://hbctraining.github.io/DGE_workshop/lessons/04_DGE_DESeq2_analysis.html</a></li>
<li>HBC DGE training module, part 2: <a
href="https://hbctraining.github.io/DGE_workshop/lessons/05_DGE_DESeq2_analysis2.html"
class="uri">https://hbctraining.github.io/DGE_workshop/lessons/05_DGE_DESeq2_analysis2.html</a></li>
<li>DESeq2 vignette: <a
href="http://bioconductor.org/packages/devel/bioc/vignettes/DESeq2/inst/doc/DESeq2.html#differential-expression-analysis"
class="uri">http://bioconductor.org/packages/devel/bioc/vignettes/DESeq2/inst/doc/DESeq2.html#differential-expression-analysis</a></li>
</ul>
<hr />
<hr />
<p>These materials have been adapted and extended from materials listed
above. These are open access materials distributed under the terms of
the <a href="http://creativecommons.org/licenses/by/4.0/">Creative
Commons Attribution license (CC BY 4.0)</a>, which permits unrestricted
use, distribution, and reproduction in any medium, provided the original
author and source are credited.</p>
<br/> <br/>
<hr/>
<table style="width:100%;">
<colgroup>
<col width="28%" />
<col width="42%" />
<col width="28%" />
</colgroup>
<thead>
<tr class="header">
<th align="left"><a href="Module10_DEComparisons.html">Previous
lesson</a></th>
<th align="center"><a href="#top">Top of this lesson</a></th>
<th align="right"><a href="Module12_DEAnnotations.html">Next
lesson</a></th>
</tr>
</thead>
<tbody>
</tbody>
</table>
</div>

<div id="rmd-source-code">---
title: "DE Visualization"
author: "UM Bioinformatics Core"
date: "`r Sys.Date()`"
output:
        html_document:
            includes:
                in_header: header.html
            theme: paper
            toc: true
            toc_depth: 4
            toc_float: true
            number_sections: false
            fig_caption: true
            markdown: GFM
            code_download: true
---

<style type="text/css">
body, td {
   font-size: 18px;
}
code.r{
  font-size: 12px;
}
pre {
  font-size: 12px
}
</style>

```{r, include = FALSE}
source("../bin/chunk-options.R")
knitr_fig_path("11-")
```

In this module, we will learn:

* How to generate common differential expression visualizations
* How to choose thresholds for calling differentially expressed genes
* Options for functional enrichments and other downstream analyses

<br>

```{r Modules, eval=TRUE, echo=FALSE, message=FALSE, warning=FALSE}
library(DESeq2)
library(ggplot2)
library(tidyr)
library(dplyr)
library(matrixStats)
library(ggrepel)
library(pheatmap)
library(RColorBrewer)
# load("rdata/RunningData.RData")

results_deficient_vs_control = results(dds_fitted, name = 'condition_deficient_vs_control')
```

# Differential Expression Workflow {.unlisted .unnumbered}

Here we will generate summary figures for our results and annotate our DE tables.

![](./images/wayfinder/wayfinder-DEVisualizations.png){width=75%}

---

# Summarizing DE results

Part of differential expression analysis is generating visualizations and summary tables to share our results. While the DESeq2 vignette provides [examples of other visualizations](http://bioconductor.org/packages/devel/bioc/vignettes/DESeq2/inst/doc/DESeq2.html#exploring-and-exporting-results), a common visualization to summarize DE comparisons are [volcano plots](http://resources.qiagenbioinformatics.com/manuals/clcgenomicsworkbench/752/index.php?manual=Volcano_plots_inspecting_result_statistical_analysis.html).

## Tabular DE summary

To generate a general summary of the DE results, we can use the `summary` function to generate a basic summary by DESeq2.

```{r DESeq2Summary}
summary(results_deficient_vs_control)
```

However, this summary is simply a text output that we are unable to manipulate. Moreover, notice that the thresholds are not quite as we would like them.

To create our own summary, we first need to choose thresholds. A standard threshold for the adjusted p-value is less than 0.05. A reasonable threshold for linear fold-change is less than -1.5 or greater than 1.5 (which corresponds to log2 fold-change -0.585 and 0.585, respectively). Including a fold-change threshold ensures that the DE genes have a reasonable effect size as well as statistical significance.

Let's set these thresholds as variables to reuse. This is generally good practice because if you want to change those thresholds later then you only have to change them in one location of your script, which is faster and can reduce the risk of errors from missing some instances in your code.

```{r PlotSetup1}
fc = 1.5
fdr = 0.05
```

> # Note: Choosing thresholds {.unlisted .unnumbered}
>
> Thresholding on adjusted p-values < 0.05 is a standard threshold, but depending on the research question and/or how the results will be used, other thresholds might be reasonable.
>
> There is a nice [Biostar post that discusses choosing adjusted p-value thresholds](https://www.biostars.org/p/209118/), including cases where a more relaxed threshold might be appropriate and (some heated) discussion of the dangers of adjusting the choosen threshold after running an analysis. Additionally, there is a [Dalmon et al 2012](https://bmcbioinformatics.biomedcentral.com/articles/10.1186/1471-2105-13-S2-S11) paper about p-value and fold-change thresholds for microarray data that may help provide some context.

If we think back to Computational Foundations, conditional statements could allow us to determine the number of genes that pass our thresholds, which would be useful for annotating our results tables and plots.

> # Exercise {.unlisted .unnumbered}
> How would we identify the number of genes with adjusted p-values < 0.05 and a fold-change above 1.5 (or below -1.5) in our comparison?


<details>
<summary>Solution</summary>

Here is one possible answer:

```{r StatSigGenes2}
sum(results_deficient_vs_control$padj < fdr & abs(results_deficient_vs_control$log2FoldChange) >= log2(fc), na.rm = TRUE)
```

**Checkpoint**: *If you see the same number of DE ganes with our choosen thresholds, please indicate with a green check. Otherwise use a red x.*

</details>
<br>

Let's now create a new column in `results_deficient_vs_control` to record the significance "call" based on these thresholds. And let's separate the call by "Up" or "Down", noting that these are relative to our "Case" condition. There are many ways to accomplish this, but the following will work:

First define all values as "NS" or "not significant":

```{r NSColumn}
results_deficient_vs_control$call = 'NS'
head(results_deficient_vs_control)
```

Next, determine the "Up" and "Down" indices:

```{r UpDownIndices}
up_idx = results_deficient_vs_control$padj < fdr & results_deficient_vs_control$log2FoldChange > log2(fc)
down_idx = results_deficient_vs_control$padj < fdr & results_deficient_vs_control$log2FoldChange < -log2(fc)
```

Last, use those indices to assign the correct "Up" or "Down" values to the correct indices, and look at the head of the result:

```{r CallColumn}
results_deficient_vs_control$call[up_idx] = 'Up'
results_deficient_vs_control$call[down_idx] = 'Down'
head(results_deficient_vs_control)
```

Finally, since plotting functions often require categorical groupings, let's make this `call` column a factor and specify the level ordering:

```{r FactorCall}
results_deficient_vs_control$call = factor(results_deficient_vs_control$call, levels = c('Up', 'Down', 'NS'))
head(results_deficient_vs_control)
```

> # Tip {.unlisted .unnumbered}
> It is often helpful to include code like this in differential expression analyses so there is a clearly labelled column that makes subsetting and summarizing the results easier.

Now we are in a position to quickly summarize our differential expression results:

```{r TableCall}
table(results_deficient_vs_control$call)
```

We see quickly how many genes were "Up" in iron replete, how many were "Down" in iron replete, and how many were not significant.

**Checkpoint**: *If you successfully added the `call` column and got the same table result as above, please indicate with a green check. Otherwise use a red x.*

## Visual DE summary

As described by the [Galaxy project](https://galaxyproject.github.io/training-material/topics/transcriptomics/tutorials/rna-seq-viz-with-volcanoplot/tutorial.html), a volcano plot is a type of scatterplot that shows statistical significance (adjusted p-value) versus effect size (fold change). In a volcano plot, the most upregulated genes are towards the right, the most downregulated genes are towards the left, and the most statistically significant genes are towards the top.

Let's coerce the `DataFrame` which was returned by `DESeq2::results()` into a `tibble` in anticipation of using the `ggplot2` library to plot. We're also going to modify our results table so that the row names become a separate column, and so that it's ordered by adjusted p-value.

```{r PlotSetup2}
# Use the rownames argument to create a new column of gene IDs
# Also arrange by adjusted p-value
results_forPlot = as_tibble(results_deficient_vs_control, rownames = 'id') %>% arrange(padj)
```

Let's start with a very simple volcano plot that plots the `log2FoldChange` on the x-axis, and `-log10(padj)` on the y-axis.

```{r VolcanoPlot}
# Initialize the plot, specifying the plot type as 'geom_point'
ggplot(results_forPlot, aes(x = log2FoldChange, y = -log10(padj))) +
  geom_point()
```

This is a good start, but it's good practice to add better labels to the plot with the `labs()` function:

```{r VolcanoPlot2}
# Add plot labels and change the theme - save the plot as object `p`
p = ggplot(results_forPlot, aes(x = log2FoldChange, y = -log10(padj))) +
    geom_point() +
    theme_bw() +
    labs(
        title = 'Volcano Plot',
        subtitle = 'control vs deficient',
        x = 'log2 fold-change',
        y = '-log10 FDR'
    )
p
```

What if we now added some visual guides to indicate where the significant genes are? We can use the `geom_vline()` and `geom_hline()` functions to accomplish this:

```{r VolcanoPlot3}
# Add threshold lines
p1 = p +
    geom_vline(
        xintercept = c(0, -log2(fc), log2(fc)),
        linetype = c(1, 2, 2)) +
    geom_hline(
        yintercept = -log10(fdr),
        linetype = 2)
p1
```

Finally, why not color the points by their significance status? We already created the `call` column that has the correct values. In this case we can get away with adding `geom_point()` to our existing plot and specifying the correct aesthetic:

```{r VolcanoPlot4}
p2 = p1 + geom_point(aes(color = call))
p2
```

Finally, we can adjust our plot to have a nicer color scheme:
```{r}
p3 = p2 + scale_color_manual(name = '', values=c('#B31B21',  '#1465AC', 'darkgray'))
p3
```

For additional visualizations for our DE results, we included some example code in the [Bonus Content module](https://umich-brcf-bioinf.github.io/2023-03-13-umich-rnaseq-demystified/html/R_bonus_content.html) and this [HBC tutorial](https://hbctraining.github.io/DGE_workshop/lessons/06_DGE_visualizing_results.html) also includes some nice examples.

### Subsetting significant genes

You may be interested in creating a table of only the genes that pass your significance thresholds. A useful way to do this is to conditionally subset your results. Again, we already created the `call` column, which makes this relatively simple to do.

*Note: The tidyverse functions you learned in Software Carpentry could also be alternatively used here.*

```{r ConditionalSubset}
# tidyr (requires table reformatting)
res_sig <- as_tibble(results_deficient_vs_control, rownames = "gene_ids") %>% filter(call != 'NS')

# base
res_sig <- results_deficient_vs_control[results_deficient_vs_control$call != 'NS', ]

head(res_sig)
dim(res_sig)
```


# Summary

In this section, we:

* Generated a volcano plot for our differential expression results
* Summarized our differential expression results
* Discussed choosing thresholds


---

# Sources

* HBC DGE training module, part 1: https://hbctraining.github.io/DGE_workshop/lessons/04_DGE_DESeq2_analysis.html
* HBC DGE training module, part 2: https://hbctraining.github.io/DGE_workshop/lessons/05_DGE_DESeq2_analysis2.html
* DESeq2 vignette: http://bioconductor.org/packages/devel/bioc/vignettes/DESeq2/inst/doc/DESeq2.html#differential-expression-analysis


---

```{r WriteOut.RData, eval=TRUE, echo=FALSE, message=FALSE, warning=FALSE}
# Hidden code block to write out data for knitting
# save.image(file = "rdata/RunningData_Full.RData")
```


---

These materials have been adapted and extended from materials listed above. These are open access materials distributed under the terms of the [Creative Commons Attribution license (CC BY 4.0)](http://creativecommons.org/licenses/by/4.0/), which permits unrestricted use, distribution, and reproduction in any medium, provided the original author and source are credited.

<br/>
<br/>
<hr/>
| [Previous lesson](Module10_DEComparisons.html) | [Top of this lesson](#top) | [Next lesson](Module12_DEAnnotations.html) |
| :--- | :----: | ---: |</div>


</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->
<script>
$(document).ready(function () {
  window.initializeSourceEmbed("Module11_DEVisualizations.Rmd");
});
</script>

<script>
$(document).ready(function ()  {

    // temporarily add toc-ignore selector to headers for the consistency with Pandoc
    $('.unlisted.unnumbered').addClass('toc-ignore')

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3,h4",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
